// @rc-component/async-validator/es/util.js
import E from "../../../@babel/runtime/helpers/esm/objectSpread2.js";
import w from "../../../@babel/runtime/helpers/esm/typeof.js";
import C from "../../../@babel/runtime/helpers/esm/createClass.js";
import D from "../../../@babel/runtime/helpers/esm/classCallCheck.js";
import A from "../../../@babel/runtime/helpers/esm/assertThisInitialized.js";
import I from "../../../@babel/runtime/helpers/esm/inherits.js";
import M from "../../../@babel/runtime/helpers/esm/createSuper.js";
import R from "../../../@babel/runtime/helpers/esm/wrapNativeSuper.js";
import O from "../../../@babel/runtime/helpers/esm/defineProperty.js";
import F from "../../../@babel/runtime/helpers/esm/toConsumableArray.js";
var G = /%[sdj%]/g, J = function() {
};
function N(n) {
  if (!n || !n.length) return null;
  var e = {};
  return n.forEach(function(r) {
    var t = r.field;
    e[t] = e[t] || [], e[t].push(r);
  }), e;
}
function U(n) {
  for (var e = arguments.length, r = new Array(e > 1 ? e - 1 : 0), t = 1; t < e; t++) r[t - 1] = arguments[t];
  var i = 0, o = r.length;
  return typeof n == "function" ? n.apply(null, r) : typeof n == "string" ? n.replace(G, function(f) {
    if (f === "%%") return "%";
    if (i >= o) return f;
    switch (f) {
      case "%s":
        return String(r[i++]);
      case "%d":
        return Number(r[i++]);
      case "%j":
        try {
          return JSON.stringify(r[i++]);
        } catch {
          return "[Circular]";
        }
        break;
      default:
        return f;
    }
  }) : n;
}
function X(n, e) {
  return n == null || !(e !== "array" || !Array.isArray(n) || n.length) || !(!/* @__PURE__ */ function(r) {
    return r === "string" || r === "url" || r === "hex" || r === "email" || r === "date" || r === "pattern";
  }(e) || typeof n != "string" || n);
}
function V(n, e, r) {
  var t = 0, i = n.length;
  (function o(f) {
    if (f && f.length) r(f);
    else {
      var u = t;
      t += 1, u < i ? e(n[u], o) : r([]);
    }
  })([]);
}
typeof process < "u" && process.env && process.env.NODE_ENV !== "production" && typeof window < "u" && typeof document < "u" && (J = function(n, e) {
  typeof console < "u" && console.warn && typeof ASYNC_VALIDATOR_NO_WARNING > "u" && e.every(function(r) {
    return typeof r == "string";
  }) && console.warn(n, e);
});
var b = function(n) {
  I(r, n);
  var e = M(r);
  function r(t, i) {
    var o;
    return D(this, r), o = e.call(this, "Async Validation Error"), O(A(o), "errors", void 0), O(A(o), "fields", void 0), o.errors = t, o.fields = i, o;
  }
  return C(r);
}(R(Error));
function Z(n, e, r, t, i) {
  if (e.first) {
    var o = new Promise(function(l, y) {
      var d = function(a) {
        var s = [];
        return Object.keys(a).forEach(function(m) {
          s.push.apply(s, F(a[m] || []));
        }), s;
      }(n);
      V(d, r, function(a) {
        return t(a), a.length ? y(new b(a, N(a))) : l(i);
      });
    });
    return o.catch(function(l) {
      return l;
    }), o;
  }
  var f = e.firstFields === !0 ? Object.keys(n) : e.firstFields || [], u = Object.keys(n), p = u.length, j = 0, c = [], g = new Promise(function(l, y) {
    var d = function(a) {
      if (c.push.apply(c, a), ++j === p) return t(c), c.length ? y(new b(c, N(c))) : l(i);
    };
    u.length || (t(c), l(i)), u.forEach(function(a) {
      var s = n[a];
      f.indexOf(a) !== -1 ? V(s, r, d) : function(m, x, k) {
        var h = [], S = 0, _ = m.length;
        function P(v) {
          h.push.apply(h, F(v || [])), ++S === _ && k(h);
        }
        m.forEach(function(v) {
          x(v, P);
        });
      }(s, r, d);
    });
  });
  return g.catch(function(l) {
    return l;
  }), g;
}
function $(n, e) {
  return function(r) {
    var t, i;
    return t = n.fullFields ? function(o, f) {
      for (var u = o, p = 0; p < f.length; p++) {
        if (u == null) return u;
        u = u[f[p]];
      }
      return u;
    }(e, n.fullFields) : e[r.field || n.fullField], (i = r) && i.message !== void 0 ? (r.field = r.field || n.fullField, r.fieldValue = t, r) : { message: typeof r == "function" ? r() : r, fieldValue: t, field: r.field || n.fullField };
  };
}
function rr(n, e) {
  if (e) {
    for (var r in e) if (e.hasOwnProperty(r)) {
      var t = e[r];
      w(t) === "object" && w(n[r]) === "object" ? n[r] = E(E({}, n[r]), t) : n[r] = t;
    }
  }
  return n;
}
export {
  b as AsyncValidationError,
  Z as asyncMap,
  $ as complementError,
  N as convertFieldsError,
  rr as deepMerge,
  U as format,
  X as isEmptyValue,
  J as warning
};
