// @secondcloset/types/src/fulfillment/shipment/shipment.ts
import { ShipmentItem } from "./shipmentItem";
import {
  BaseUser,
  BaseNote,
  BaseOrganization,
  Facility,
  Version,
} from "../../_common";
import { Package } from "./package";
import {
  SecondClosetShippingMethod,
  ExternalCarrierShippingMethod,
  CustomerPickupShippingMethod,
} from "./shippingMethod";
import { Carrier } from "./carrier";
import { AppointmentType as FulfillmentAppointmentType } from "../appointment/appointmentType";
import { AppointmentType as StorageAppointmentType } from "../../storage/appointment/appointmentType";

export type ShipmentActionType =
  | "setup"
  | "customer_pickup"
  | StorageAppointmentType
  | FulfillmentAppointmentType;

export type ServiceCode =
  | "standard"
  | "white_glove"
  | "to_the_door"
  | "room_of_choice"
  | "standard_1_person"
  | "room_of_choice_1_person"
  | "to_the_door_1_person";

export type ScServiceCode =
  | "second_closet_standard"
  | "second_closet_white_glove"
  | "second_closet_to_the_door"
  | "second_closet_room_of_choice"
  | "second_closet_standard_1_person"
  | "second_closet_room_of_choice_1_person"
  | "second_closet_to_the_door_1_person"
  | ServiceCode;

export type Shipment<V extends Version = "V1"> = V extends "V1"
  ? {
      id: string;
      organization_id: string;
      organization?: BaseOrganization;
      billed: boolean;
      user: BaseUser;
      notes: BaseNote[] | string;
      shipment_items: ShipmentItem<V>[];
      shipment_number: string;
      packages: Package[];
      external_order_id: string | null;
      external_order_type: string;
      external_shipment_id: string | null;
      carrier: Carrier | null;
      created_at: string;
      updated_at: string;
      action_summary: ShipmentActionType[];
      freight: boolean;
      shipping_method:
        | SecondClosetShippingMethod
        | ExternalCarrierShippingMethod
        | CustomerPickupShippingMethod;
      shipping_method_type:
        | "external_carrier_shipment"
        | "appointment"
        | "untracked_shipment"
        | "customer_pickup_shipment";
      external_order?: {
        id: string;
        external_order_number: string;
        external_order_id: string | null;
        created_at: string;
        updated_at: string;
      };
      facility?: Partial<Facility>;
    }
  : Omit<Shipment<"V1">, "shipment_items">;
