// @secondcloset/types/src/fulfillment/shipment/shipmentItem.ts
import { Version } from "../../_common";
import { BaseProduct } from "../product";
import { ShipmentActionType } from "./shipment";
export type ShipmentItem<V extends Version = "V1"> = V extends "V1"
  ? {
      id: string;
      order_item_id: string;
      description: string;
      item_identifier: string;
      item_identifier_type: string;
      item_identifier_value: string;
      product_id?: string;
      weight_value: string;
      weight_unit: string;
      length_value: string;
      length_unit: string;
      height_value: string;
      height_unit: string;
      width_value: string;
      width_unit: string;
      shipment_actions: ShipmentActionType[];
      shipment_actions_completion: Partial<{
        [key in ShipmentActionType]: boolean;
      }>;
      shipment_actions_status?: {
        [key in ShipmentActionType]: {
          completed: boolean;
          failed: boolean;
          failed_reason: string | null;
        };
      };
      shipment_item_group: {
        id: string;
        name: string;
        description: string;
        external_group_id: string;
        group_identifier_type: string;
        group_identifier_value: string;
        created_at: string;
        updated_at: string;
        group_identifier: string;
      } | null;
    }
  : {
      product: BaseProduct;
      shipment_item_ids: ShipmentItem<"V1">["id"][];
      shipment_actions_status: {
        [key in ShipmentActionType]: {
          completed_shipment_item_ids: string[];
          failed_shipment_items: {
            ids: string[];
            failed_reason: string | null;
          }[];
        };
      };
      quantity: number;
      attributes?: string[];
      return_reasons?: string[];
    };
