// @gobolt/genesis/dist/components/Toast/styles.cjs
"use strict";Object.defineProperty(exports,Symbol.toStringTag,{value:"Module"});const i=require("styled-components"),r=require("../../constants/index.cjs"),a=i.div`
  display: flex;
  align-items: center;
  gap: ${({theme:e})=>e.sizing.Size4}px;
  padding: ${({theme:e})=>e.sizing.Size3}px;
  background-color: ${({theme:e,state:t})=>e.colors.surface.default};
  border-left: ${({theme:e,state:t})=>`2px solid ${e.colors.status[t].default}`};
  pointer-events: none;
  cursor: default;
  border-radius: ${({theme:e})=>e.borderRadius.BorderRadiusXs}px;
  height: 56px;
  width: ${({theme:e,state:t,size:o})=>o==="standard"?"340px":"240px"};
  background-color: ${({theme:e,state:t})=>e.colors.surface.default};
  box-shadow: 0px 4px 16px 0px #00000026;
`,s=i.button`
  text-decoration: underline;
  border: none;
  background: none;
  cursor: pointer;
  pointer-events: all;
`,n=i.div`
  display: flex;
  align-items: center;
  gap: ${({theme:e})=>e.sizing.Size2}px;
  flex: 1;
  max-width: ${({theme:e,state:t,size:o})=>o==="standard"?"340px":"240px"};

  .ant-typography {
    display: -webkit-box;
    -webkit-line-clamp: 3;
    -webkit-box-orient: vertical;
    overflow: hidden;
    text-overflow: ellipsis;
    font-size: 16px;
    font-family: ${({theme:e})=>e.typography.fontFamily};
    line-height: ${({theme:e,size:t})=>t===r.SIZE.standard?e.typography.lineHeight:"16px"};
  }
`;exports.ToastCloseButton=s,exports.ToastContent=n,exports.ToastRow=a;
