// @gobolt/genesis/dist/styles/design-tokens/variables.js
const a = "#f9d4d5", e = "#f3a9ab", o = "#ee7d81", r = "#e85257", i = "#cf3237", s = "#aa292d", y = "#842023", p = "#5f1719", d = "#3a0e0f", l = "#ffe9dc", n = "#ffd3b8", S = "#ffbc95", g = "#ffa671", f = "#ff904e", z = "#d17640", t = "#a35c32", h = "#754224", F = "#472816", T = "#fff2df", C = "#ffe5be", u = "#ffd79e", b = "#ffca7d", w = "#ffbd5d", B = "#e7a13c", m = "#b27330", M = "#704819", c = "#442c0f", N = "#d3efd7", W = "#a7dfaf", G = "#7ad086", k = "#5bc16b", D = "#32863e", H = "#296e33", L = "#205628", R = "#173e1d", O = "#0e2511", v = "#deeff3", P = "#bcdfe7", Y = "#9bcfdb", x = "#79bfcf", X = "#58afc3", I = "#337e8e", _ = "#30606b", j = "#214148", q = "#00282b", A = "#d7dbfd", E = "#afb8fb", J = "#8894fa", K = "#6071f8", Q = "#384df6", U = "#2e3fca", V = "#24319d", Z = "#1a2371", $ = "#101645", aa = "#f3e0f3", ea = "#e7c1e7", oa = "#daa3da", ra = "#ce84ce", ia = "#c265c2", sa = "#9f539f", ya = "#7c417c", pa = "#592e59", da = "#361c36", la = "#f4f4f4", na = "#eeeeee", Sa = "#dfdfdf", ga = "#cbcbcb", fa = "#9a9a9a", za = "#7f7f7f", ta = "#6c6c6c", ha = "#3e3e3e", Fa = "#222222", Ta = "#101010", Ca = "#ffffff", ua = 0, ba = 1, wa = 4, Ba = 6, ma = 8, Ma = 12, ca = 16, Na = 24, Wa = 999, Ga = "'Inter', sans-serif !important", ka = 4, Da = 6, Ha = 8, La = 10, Ra = 12, Oa = 16, va = 20, Pa = 24, Ya = 28, xa = 32, Xa = 40, Ia = 48, _a = 52, ja = 56, qa = 64, Aa = 72, Ea = 80, Ja = 96, Ka = 108, Qa = 112, Ua = 120, Va = 128, Za = 136, $a = 144, ae = 156, ee = 172, oe = 200, re = 256, ie = 2, se = 6, ye = 1, pe = 10, de = 72, le = 56, ne = 40, Se = 40, ge = 32, fe = 64, ze = 52, te = 40, he = 40, Fe = 24, Te = 48, Ce = 40, ue = 32, be = 32, we = 24, Be = 16, me = 16, Me = 16, ce = 20, Ne = 16, We = 14, Ge = 14, ke = 12, De = 12, He = 12, Le = 14, Re = 12, Oe = 12, ve = 12, Pe = 16, Ye = 14, xe = 12, Xe = 16, Ie = 14, _e = 12, je = 96, qe = 48, Ae = 16, Ee = 16, Je = 20, Ke = 16, Qe = 14, Ue = 14, Ve = 12, Ze = 12, $e = 12, ao = 14, eo = 12, oo = 12, ro = 12, io = 16, so = 14, yo = 12, po = 16, lo = 14, no = 12, So = 96, go = 48, fo = 16, zo = 16, to = 20, ho = 16, Fo = 14, To = 12, Co = 14, uo = 16, bo = 18, wo = 20, Bo = 14, mo = 12, Mo = 12, co = 12, No = 16, Wo = 14, Go = 12, ko = 16, Do = 14, Ho = 12, Lo = 96, Ro = 48, Oo = 16;
export {
  ma as BorderRadiusLg,
  Ba as BorderRadiusMd,
  ua as BorderRadiusNone,
  Wa as BorderRadiusRounded,
  wa as BorderRadiusSm,
  Ma as BorderRadiusXl,
  ba as BorderRadiusXs,
  ca as BorderRadiusXxl,
  Na as BorderRadiusXxxl,
  A as ColorBlue100,
  E as ColorBlue200,
  J as ColorBlue300,
  K as ColorBlue400,
  Q as ColorBlue500,
  U as ColorBlue600,
  V as ColorBlue700,
  Z as ColorBlue800,
  $ as ColorBlue900,
  la as ColorGray100,
  na as ColorGray200,
  Sa as ColorGray300,
  ga as ColorGray400,
  fa as ColorGray500,
  za as ColorGray600,
  ta as ColorGray700,
  ha as ColorGray800,
  Fa as ColorGray900,
  Ta as ColorGrayBlack,
  Ca as ColorGrayWhite,
  N as ColorGreen100,
  W as ColorGreen200,
  G as ColorGreen300,
  k as ColorGreen400,
  D as ColorGreen500,
  H as ColorGreen600,
  L as ColorGreen700,
  R as ColorGreen800,
  O as ColorGreen900,
  l as ColorOrange100,
  n as ColorOrange200,
  S as ColorOrange300,
  g as ColorOrange400,
  f as ColorOrange500,
  z as ColorOrange600,
  t as ColorOrange700,
  h as ColorOrange800,
  F as ColorOrange900,
  aa as ColorPurple100,
  ea as ColorPurple200,
  oa as ColorPurple300,
  ra as ColorPurple400,
  ia as ColorPurple500,
  sa as ColorPurple600,
  ya as ColorPurple700,
  pa as ColorPurple800,
  da as ColorPurple900,
  a as ColorRed100,
  e as ColorRed200,
  o as ColorRed300,
  r as ColorRed400,
  i as ColorRed500,
  s as ColorRed600,
  y as ColorRed700,
  p as ColorRed800,
  d as ColorRed900,
  v as ColorSky100,
  P as ColorSky200,
  Y as ColorSky300,
  x as ColorSky400,
  X as ColorSky500,
  I as ColorSky600,
  _ as ColorSky700,
  j as ColorSky800,
  q as ColorSky900,
  T as ColorYellow100,
  C as ColorYellow200,
  u as ColorYellow300,
  b as ColorYellow400,
  w as ColorYellow500,
  B as ColorYellow600,
  m as ColorYellow700,
  M as ColorYellow800,
  c as ColorYellow900,
  ye as Size025,
  ie as Size05,
  ka as Size1,
  Xa as Size10,
  Ia as Size12,
  _a as Size13,
  ja as Size14,
  se as Size15,
  qa as Size16,
  Aa as Size18,
  Da as Size1_5,
  Ha as Size2,
  Ea as Size20,
  Ja as Size24,
  pe as Size25,
  Ka as Size27,
  Qa as Size28,
  La as Size2_5,
  Ra as Size3,
  Ua as Size30,
  Va as Size32,
  Za as Size34,
  $a as Size35,
  ae as Size39,
  Oa as Size4,
  ee as Size43,
  va as Size5,
  oe as Size50,
  Pa as Size6,
  re as Size64,
  Ya as Size7,
  xa as Size8,
  me as TypographyMediumFontSizeBody1,
  Ue as TypographyMediumFontSizeBody2,
  Ve as TypographyMediumFontSizeBody3,
  Ze as TypographyMediumFontSizeBody4,
  $e as TypographyMediumFontSizeBody5,
  So as TypographyMediumFontSizeDigits1,
  go as TypographyMediumFontSizeDigits2,
  fo as TypographyMediumFontSizeDigits3,
  fe as TypographyMediumFontSizeDisplay1,
  ze as TypographyMediumFontSizeDisplay2,
  te as TypographyMediumFontSizeDisplay3,
  he as TypographyMediumFontSizeHeading1,
  Fe as TypographyMediumFontSizeHeading2,
  Ee as TypographyMediumFontSizeHeading3,
  io as TypographyMediumFontSizeLabel1,
  so as TypographyMediumFontSizeLabel2,
  yo as TypographyMediumFontSizeLabel3,
  po as TypographyMediumFontSizeLink1,
  lo as TypographyMediumFontSizeLink2,
  no as TypographyMediumFontSizeLink3,
  ao as TypographyMediumFontSizeMessage,
  eo as TypographyMediumFontSizeOverline1,
  oo as TypographyMediumFontSizeOverline2,
  ro as TypographyMediumFontSizeOverline3,
  Je as TypographyMediumFontSizeSubHeading1,
  Ke as TypographyMediumFontSizeSubHeading2,
  Qe as TypographyMediumFontSizeSubHeading3,
  To as TypographyNarrowFontSizeBody1,
  Co as TypographyNarrowFontSizeBody2,
  uo as TypographyNarrowFontSizeBody3,
  bo as TypographyNarrowFontSizeBody4,
  wo as TypographyNarrowFontSizeBody5,
  Lo as TypographyNarrowFontSizeDigits1,
  Ro as TypographyNarrowFontSizeDigits2,
  Oo as TypographyNarrowFontSizeDigits3,
  Te as TypographyNarrowFontSizeDisplay1,
  Ce as TypographyNarrowFontSizeDisplay2,
  ue as TypographyNarrowFontSizeDisplay3,
  be as TypographyNarrowFontSizeHeading1,
  we as TypographyNarrowFontSizeHeading2,
  zo as TypographyNarrowFontSizeHeading3,
  No as TypographyNarrowFontSizeLabel1,
  Wo as TypographyNarrowFontSizeLabel2,
  Go as TypographyNarrowFontSizeLabel3,
  ko as TypographyNarrowFontSizeLink1,
  Do as TypographyNarrowFontSizeLink2,
  Ho as TypographyNarrowFontSizeLink3,
  Bo as TypographyNarrowFontSizeMessage,
  mo as TypographyNarrowFontSizeOverline1,
  Mo as TypographyNarrowFontSizeOverline2,
  co as TypographyNarrowFontSizeOverline3,
  to as TypographyNarrowFontSizeSubHeading1,
  ho as TypographyNarrowFontSizeSubHeading2,
  Fo as TypographyNarrowFontSizeSubHeading3,
  Ga as TypographyPrimaryFontFamily,
  Be as TypographyWideFontSizeBody1,
  Ge as TypographyWideFontSizeBody2,
  ke as TypographyWideFontSizeBody3,
  De as TypographyWideFontSizeBody4,
  He as TypographyWideFontSizeBody5,
  je as TypographyWideFontSizeDigits1,
  qe as TypographyWideFontSizeDigits2,
  Ae as TypographyWideFontSizeDigits3,
  de as TypographyWideFontSizeDisplay1,
  le as TypographyWideFontSizeDisplay2,
  ne as TypographyWideFontSizeDisplay3,
  Se as TypographyWideFontSizeHeading1,
  ge as TypographyWideFontSizeHeading2,
  Me as TypographyWideFontSizeHeading3,
  Pe as TypographyWideFontSizeLabel1,
  Ye as TypographyWideFontSizeLabel2,
  xe as TypographyWideFontSizeLabel3,
  Xe as TypographyWideFontSizeLink1,
  Ie as TypographyWideFontSizeLink2,
  _e as TypographyWideFontSizeLink3,
  Le as TypographyWideFontSizeMessage,
  Re as TypographyWideFontSizeOverline1,
  Oe as TypographyWideFontSizeOverline2,
  ve as TypographyWideFontSizeOverline3,
  ce as TypographyWideFontSizeSubHeading1,
  Ne as TypographyWideFontSizeSubHeading2,
  We as TypographyWideFontSizeSubHeading3
};
