// @gobolt/genesis/dist/components/Breadcrumb/styles.cjs
"use strict";Object.defineProperties(exports,{__esModule:{value:!0},[Symbol.toStringTag]:{value:"Module"}});const a=require("styled-components"),i=require("../../node_modules/antd/es/breadcrumb/Breadcrumb.cjs"),t=({colors:n,typography:e},o="wide")=>`
  &.ant-breadcrumb {
    font-size: ${e[o].body1.fontSize};
    
    .ant-breadcrumb-link,
    span[class*="ant-breadcrumb-link"] > a {
      color: ${n.interactive.link.active};
      transition: color 0.2s ease-in-out;
      text-decoration: underline;
      background: none !important;
    }
      
    ol > li:last-child {
      .ant-breadcrumb-link,
      span[class*="ant-breadcrumb-link"],
      span[class*="ant-breadcrumb-link"] > span,
      span[class*="ant-breadcrumb-link"] > a {
        text-decoration: none !important;
        color: ${n.onsurface["copy-light"]} !important;
        cursor: default !important;
        pointer-events: none;
      }
    }
 
    .ant-breadcrumb-separator {
      color: ${n.interactive.link.active};
    }

    .ant-dropdown-menu-item,
    .ant-dropdown-menu-item:hover {
      background: none !important;
    }

    .ant-dropdown-trigger > svg {
      display: none;
    }

    span[role="img"] {
      display: none;
    }

    .ant-dropdown-trigger:hover {
      background: none !important;
    }
  }
`,s=a.div`
  ${({$isDisabled:n})=>n&&`
      cursor: not-allowed;
      .ant-breadcrumb {
        pointer-events: none;
        opacity: 0.5;
      }
    `}
  opacity: ${({$isDisabled:n})=>n?.5:1};
`,r=a(i)`
  ${({theme:n,breakpoint:e})=>n&&t(n,e)}
`;exports.Breadcrumb=r,exports.BreadcrumbWrapper=s,exports.default=r,exports.getGenesisClass=t;
