// @gobolt/genesis/dist/components/Button/styles.cjs
"use strict";Object.defineProperty(exports,Symbol.toStringTag,{value:"Module"});const C=require("styled-components"),k=require("../../node_modules/antd/es/button/button.cjs"),y=C(k)`
  ${({theme:p,$themeType:$,$state:b,children:v,$isIconButton:u,size:x})=>((g,h,z,m,S)=>`
    &.ant-btn {
      font-family: 'Inter', sans-serif;
 
      ${(({colors:o,sizing:d,borderRadius:s,components:f},n,i,l,t)=>(o[n][i].backgroundColor,`
      &.ant-btn {
      
        color: ${o[n][i].color};
        font-size: ${d.Size4}px;
        line-height: ${d.Size6}px;
        letter-spacing: 0;
        font-weight: 400;
        border-width: 1px;
        line-height: 1 !important;

        width: ${((r,a)=>a?r==="normal"||r==="large"?"40px":r==="small"?"32px":"40px":"auto")(l,t)}; !important;
        height: ${(r=>r==="normal"||r==="large"?"40px":r==="small"?"32px":"40px")(l)}; !important;

        background-color: ${o[n][i].backgroundColor};
        padding: ${((r,a,e,c)=>!c||r!=="normal"&&r!=="large"?c&&r==="small"?`${e.Size1}px ${e.Size1}px`:a==="utility"&&r==="small"?`${e.Size1_5}px ${e.Size2}px`:a!=="utility"||r!=="normal"&&r!=="large"?r==="small"?`${e.Size1_5}px ${e.Size4}px`:r==="normal"||r==="large"?`${e.Size2_5}px ${e.Size6}px`:`${e.Size2}px ${e.Size2}px`:`${e.Size2_5}px ${e.Size3}px`:`${e.Size2}px ${e.Size2}px`)(l,n,d,t)};

        box-shadow: ${n==="utility"?"0px 1px 2px 0px #00000026":"none"};

        border-radius: ${s.BorderRadiusMd}px;
        gap: ${f.button.gap}px;
        border-style: solid;
        border-color: ${o[n][i].borderColor};
        
        transition: all 0.2s ease-in-out;

        &:hover {
          color: ${o[n].hover.color};
          background-color: ${o[n].hover.backgroundColor};
          border-color: ${o[n].hover.borderColor};
        }

        &:active {
          color: ${o[n].pressed.color};
          background-color: ${o[n].pressed.backgroundColor};
          border-color: ${o[n].pressed.borderColor};
        }

        &:focus-visible {
          outline: none;
          color: ${o[n].focussed.color};
          background-color: ${o[n].focussed.backgroundColor};
          box-shadow: 0 0 0 1px #fff, 0 0 0 4px ${o[n].focussed.ringColor};
          border-radius: ${s.BorderRadiusMd}px;
          border-color: ${o[n].focussed.borderColor};
          transition: box-shadow 0.2s ease-in-out;
        }

        &:disabled {
          color: ${o[n].disabled.color};
          background-color: ${o[n].disabled.backgroundColor};
          border-color: ${o[n].disabled.borderColor};
          cursor: not-allowed;
        }
    
      }
        
      `))(g,h,z,S,m)}
    } 
`)(p,$,b,u,x)}
`;exports.Button=y;
