// @gobolt/genesis/dist/components/Button/styles.js
import C from "styled-components";
import k from "../../node_modules/antd/es/button/button.js";
const B = C(k)`
  ${({ theme: c, $themeType: $, $state: b, children: w, $isIconButton: x, size: u }) => ((g, m, h, z, S) => `
    &.ant-btn {
      font-family: 'Inter', sans-serif;
 
      ${(({ colors: n, sizing: d, borderRadius: s, components: f }, o, i, l, t) => (n[o][i].backgroundColor, `
      &.ant-btn {
      
        color: ${n[o][i].color};
        font-size: ${d.Size4}px;
        line-height: ${d.Size6}px;
        letter-spacing: 0;
        font-weight: 400;
        border-width: 1px;
        line-height: 1 !important;

        width: ${/* @__PURE__ */ ((r, a) => a ? r === "normal" || r === "large" ? "40px" : r === "small" ? "32px" : "40px" : "auto")(l, t)}; !important;
        height: ${/* @__PURE__ */ ((r) => r === "normal" || r === "large" ? "40px" : r === "small" ? "32px" : "40px")(l)}; !important;

        background-color: ${n[o][i].backgroundColor};
        padding: ${((r, a, e, p) => !p || r !== "normal" && r !== "large" ? p && r === "small" ? `${e.Size1}px ${e.Size1}px` : a === "utility" && r === "small" ? `${e.Size1_5}px ${e.Size2}px` : a !== "utility" || r !== "normal" && r !== "large" ? r === "small" ? `${e.Size1_5}px ${e.Size4}px` : r === "normal" || r === "large" ? `${e.Size2_5}px ${e.Size6}px` : `${e.Size2}px ${e.Size2}px` : `${e.Size2_5}px ${e.Size3}px` : `${e.Size2}px ${e.Size2}px`)(l, o, d, t)};

        box-shadow: ${o === "utility" ? "0px 1px 2px 0px #00000026" : "none"};

        border-radius: ${s.BorderRadiusMd}px;
        gap: ${f.button.gap}px;
        border-style: solid;
        border-color: ${n[o][i].borderColor};
        
        transition: all 0.2s ease-in-out;

        &:hover {
          color: ${n[o].hover.color};
          background-color: ${n[o].hover.backgroundColor};
          border-color: ${n[o].hover.borderColor};
        }

        &:active {
          color: ${n[o].pressed.color};
          background-color: ${n[o].pressed.backgroundColor};
          border-color: ${n[o].pressed.borderColor};
        }

        &:focus-visible {
          outline: none;
          color: ${n[o].focussed.color};
          background-color: ${n[o].focussed.backgroundColor};
          box-shadow: 0 0 0 1px #fff, 0 0 0 4px ${n[o].focussed.ringColor};
          border-radius: ${s.BorderRadiusMd}px;
          border-color: ${n[o].focussed.borderColor};
          transition: box-shadow 0.2s ease-in-out;
        }

        &:disabled {
          color: ${n[o].disabled.color};
          background-color: ${n[o].disabled.backgroundColor};
          border-color: ${n[o].disabled.borderColor};
          cursor: not-allowed;
        }
    
      }
        
      `))(g, m, h, S, z)}
    } 
`)(c, $, b, x, u)}
`;
export {
  B as Button
};
