// @rc-component/trigger/es/hooks/useAlign.js
import Z from "../../../../@babel/runtime/helpers/esm/objectSpread2.js";
import z from "../../../../@babel/runtime/helpers/esm/slicedToArray.js";
import { isDOM as Ze } from "../../../../rc-util/es/Dom/findDOMNode.js";
import ze from "../../../../rc-util/es/Dom/isVisible.js";
import Be from "../../../../rc-util/es/hooks/useEvent.js";
import Lr from "../../../../rc-util/es/hooks/useLayoutEffect.js";
import * as I from "react";
import { collectScroller as Le, getWin as De, getVisibleArea as Dr, toNum as Jr } from "../util.js";
function Xr(e) {
  var f = "".concat(arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : 0), l = f.match(/^(.*)\%$/);
  return l ? e * (parseFloat(l[1]) / 100) : parseFloat(f);
}
function qr(e, f) {
  var l = z(f || [], 2), u = l[0], v = l[1];
  return [Xr(e.width, u), Xr(e.height, v)];
}
function Ir() {
  var e = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : "";
  return [e[0], e[1]];
}
function V(e, f) {
  var l, u = f[0], v = f[1];
  return l = u === "t" ? e.y : u === "b" ? e.y + e.height : e.y + e.height / 2, { x: v === "l" ? e.x : v === "r" ? e.x + e.width : e.x + e.width / 2, y: l };
}
function d(e, f) {
  var l = { t: "b", b: "t", l: "r", r: "l" };
  return e.map(function(u, v) {
    return v === f ? l[u] || "c" : u;
  }).join("");
}
function oo(e, f, l, u, v, Hr, ur) {
  var re = I.useState({ ready: !1, offsetX: 0, offsetY: 0, offsetR: 0, offsetB: 0, arrowX: 0, arrowY: 0, scaleX: 1, scaleY: 1, align: v[u] || {} }), hr = z(re, 2), c = hr[0], cr = hr[1], H = I.useRef(0), mr = I.useMemo(function() {
    return f ? Le(f) : [];
  }, [f]), s = I.useRef({});
  e || (s.current = {});
  var ee = Be(function() {
    if (f && l && e) {
      let y = function(G, $e) {
        var q = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : Mr, zr = n.x + G, Br = n.y + $e, ke = zr + S, Ae = Br + b, _e = Math.max(zr, q.left), je = Math.max(Br, q.top), Ce = Math.min(ke, q.right), Ge = Math.min(Ae, q.bottom);
        return Math.max(0, (Ce - _e) * (Ge - je));
      }, Zr = function() {
        U = n.y + a, L = U + b, K = n.x + t, D = K + S;
      };
      var M, rr, er, or, i, r = f, vr = r.ownerDocument, nr = De(r).getComputedStyle(r), oe = nr.width, ne = nr.height, ie = nr.position, te = r.style.left, ae = r.style.top, le = r.style.right, fe = r.style.bottom, se = r.style.overflow, R = Z(Z({}, v[u]), Hr), w = vr.createElement("div");
      if ((M = r.parentElement) === null || M === void 0 || M.appendChild(w), w.style.left = "".concat(r.offsetLeft, "px"), w.style.top = "".concat(r.offsetTop, "px"), w.style.position = ie, w.style.height = "".concat(r.offsetHeight, "px"), w.style.width = "".concat(r.offsetWidth, "px"), r.style.left = "0", r.style.top = "0", r.style.right = "auto", r.style.bottom = "auto", r.style.overflow = "hidden", Array.isArray(l)) i = { x: l[0], y: l[1], width: 0, height: 0 };
      else {
        var ir, tr, m = l.getBoundingClientRect();
        m.x = (ir = m.x) !== null && ir !== void 0 ? ir : m.left, m.y = (tr = m.y) !== null && tr !== void 0 ? tr : m.top, i = { x: m.x, y: m.y, width: m.width, height: m.height };
      }
      var n = r.getBoundingClientRect();
      n.x = (rr = n.x) !== null && rr !== void 0 ? rr : n.left, n.y = (er = n.y) !== null && er !== void 0 ? er : n.top;
      var Y = vr.documentElement, ue = Y.clientWidth, he = Y.clientHeight, ce = Y.scrollWidth, me = Y.scrollHeight, pr = Y.scrollTop, gr = Y.scrollLeft, b = n.height, S = n.width, E = i.height, N = i.width, ye = { left: 0, top: 0, right: ue, bottom: he }, ve = { left: -gr, top: -pr, right: ce - gr, bottom: me - pr }, O = R.htmlRegion, xr = "visible", dr = "visibleFirst";
      O !== "scroll" && O !== dr && (O = xr);
      var Q = O === dr, pe = Dr(ve, mr), o = Dr(ye, mr), Mr = O === xr ? o : pe, B = Q ? o : Mr;
      r.style.left = "auto", r.style.top = "auto", r.style.right = "0", r.style.bottom = "0";
      var wr = r.getBoundingClientRect();
      r.style.left = te, r.style.top = ae, r.style.right = le, r.style.bottom = fe, r.style.overflow = se, (or = r.parentElement) === null || or === void 0 || or.removeChild(w);
      var W = Jr(Math.round(S / parseFloat(oe) * 1e3) / 1e3), F = Jr(Math.round(b / parseFloat(ne) * 1e3) / 1e3);
      if (W === 0 || F === 0 || Ze(l) && !ze(l)) return;
      var ge = R.offset, xe = R.targetOffset, de = qr(n, ge), Pr = z(de, 2), p = Pr[0], g = Pr[1], Me = qr(i, xe), Rr = z(Me, 2), we = Rr[0], Pe = Rr[1];
      i.x -= we, i.y -= Pe;
      var Re = R.points || [], Yr = z(Re, 2), Ye = Yr[0], P = Ir(Yr[1]), h = Ir(Ye), br = V(i, P), Sr = V(n, h), T = Z({}, R), t = br.x - Sr.x + p, a = br.y - Sr.y + g, U, L, K, D, x = y(t, a), J = y(t, a, o), Wr = V(i, ["t", "l"]), Fr = V(n, ["t", "l"]), Tr = V(i, ["b", "r"]), Vr = V(n, ["b", "r"]), X = R.overflow || {}, be = X.adjustX, Se = X.adjustY, Er = X.shiftX, Nr = X.shiftY, Or = function(G) {
        return typeof G == "boolean" ? G : G >= 0;
      };
      Zr();
      var Qr = Or(Se), Ur = h[0] === P[0];
      if (Qr && h[0] === "t" && (L > B.bottom || s.current.bt)) {
        var $ = a;
        Ur ? $ -= b - E : $ = Wr.y - Vr.y - g;
        var Kr = y(t, $), We = y(t, $, o);
        Kr > x || Kr === x && (!Q || We >= J) ? (s.current.bt = !0, a = $, g = -g, T.points = [d(h, 0), d(P, 0)]) : s.current.bt = !1;
      }
      if (Qr && h[0] === "b" && (U < B.top || s.current.tb)) {
        var k = a;
        Ur ? k += b - E : k = Tr.y - Fr.y - g;
        var $r = y(t, k), Fe = y(t, k, o);
        $r > x || $r === x && (!Q || Fe >= J) ? (s.current.tb = !0, a = k, g = -g, T.points = [d(h, 0), d(P, 0)]) : s.current.tb = !1;
      }
      var kr = Or(be), Ar = h[1] === P[1];
      if (kr && h[1] === "l" && (D > B.right || s.current.rl)) {
        var A = t;
        Ar ? A -= S - N : A = Wr.x - Vr.x - p;
        var _r = y(A, a), Te = y(A, a, o);
        _r > x || _r === x && (!Q || Te >= J) ? (s.current.rl = !0, t = A, p = -p, T.points = [d(h, 1), d(P, 1)]) : s.current.rl = !1;
      }
      if (kr && h[1] === "r" && (K < B.left || s.current.lr)) {
        var _ = t;
        Ar ? _ += S - N : _ = Tr.x - Fr.x - p;
        var jr = y(_, a), Ve = y(_, a, o);
        jr > x || jr === x && (!Q || Ve >= J) ? (s.current.lr = !0, t = _, p = -p, T.points = [d(h, 1), d(P, 1)]) : s.current.lr = !1;
      }
      Zr();
      var j = Er === !0 ? 0 : Er;
      typeof j == "number" && (K < o.left && (t -= K - o.left - p, i.x + N < o.left + j && (t += i.x - o.left + N - j)), D > o.right && (t -= D - o.right - p, i.x > o.right - j && (t += i.x - o.right + j)));
      var C = Nr === !0 ? 0 : Nr;
      typeof C == "number" && (U < o.top && (a -= U - o.top - g, i.y + E < o.top + C && (a += i.y - o.top + E - C)), L > o.bottom && (a -= L - o.bottom - g, i.y > o.bottom - C && (a += i.y - o.bottom + C)));
      var ar = n.x + t, Ee = ar + S, lr = n.y + a, Ne = lr + b, Cr = i.x, Oe = Cr + N, Gr = i.y, Qe = Gr + E, Ue = (Math.max(ar, Cr) + Math.min(Ee, Oe)) / 2 - ar, Ke = (Math.max(lr, Gr) + Math.min(Ne, Qe)) / 2 - lr;
      ur == null || ur(f, T);
      var fr = wr.right - n.x - (t + n.width), sr = wr.bottom - n.y - (a + n.height);
      W === 1 && (t = Math.round(t), fr = Math.round(fr)), F === 1 && (a = Math.round(a), sr = Math.round(sr)), cr({ ready: !0, offsetX: t / W, offsetY: a / F, offsetR: fr / W, offsetB: sr / F, arrowX: Ue / W, arrowY: Ke / F, scaleX: W, scaleY: F, align: T });
    }
  }), yr = function() {
    cr(function(M) {
      return Z(Z({}, M), {}, { ready: !1 });
    });
  };
  return Lr(yr, [u]), Lr(function() {
    e || yr();
  }, [e]), [c.ready, c.offsetX, c.offsetY, c.offsetR, c.offsetB, c.arrowX, c.arrowY, c.scaleX, c.scaleY, c.align, function() {
    H.current += 1;
    var M = H.current;
    Promise.resolve().then(function() {
      H.current === M && ee();
    });
  }];
}
export {
  oo as default
};
