// @secondcloset/types/src/_common/user.ts
import { ServiceArea } from "./serviceArea";

export interface BaseUser {
  id: string;
  first_name: string;
  last_name: string;
  user_code: string;
  email: string;
  role: string;
  phone_number: string | null;
}

export type DriverType = "fleet_operator" | "independent_operator";

export interface User extends BaseUser {
  business_name: string | null;
  business_type: string | null;
  business_user: boolean;
  closets: number;
  confirmed: boolean;
  created_at: string;
  legacy: boolean;
  orders: number;
  organization_id: string;
  organization_role: string;
  organization_role_id: string;
  payment_issue: boolean | null;
  stripe_customer_id: string;
  subrole: string;
  updated_at: string;
  driver_user?: DriverUser;
}

export interface DriverUser {
  id: string;
  terms_and_conditions_signed: boolean;
  terms_and_conditions_signed_at?: string;
  fleet_operator?: string;
  service_areas?: ServiceArea[];
  driver_type?: DriverType;
  delivery_partner_company_name?: string;
}
