// rc-virtual-list/es/List.js
import dr from "../../@babel/runtime/helpers/esm/extends.js";
import hr from "../../@babel/runtime/helpers/esm/typeof.js";
import C from "../../@babel/runtime/helpers/esm/objectSpread2.js";
import be from "../../@babel/runtime/helpers/esm/defineProperty.js";
import w from "../../@babel/runtime/helpers/esm/slicedToArray.js";
import pr from "../../@babel/runtime/helpers/esm/objectWithoutProperties.js";
import Sr from "../../../_virtual/index.js";
import gr from "../../rc-resize-observer/es/index.js";
import Re from "../../rc-util/es/hooks/useEvent.js";
import q from "../../rc-util/es/hooks/useLayoutEffect.js";
import * as a from "react";
import { useRef as v, useState as F } from "react";
import "../../rc-util/es/ref.js";
import "../../rc-util/es/warning.js";
import { flushSync as Be } from "react-dom";
import Mr from "./Filler.js";
import yr from "./hooks/useChildren.js";
import wr from "./hooks/useDiffItem.js";
import xr from "./hooks/useFrameWheel.js";
import { useGetSize as Hr } from "./hooks/useGetSize.js";
import zr from "./hooks/useHeights.js";
import Er from "./hooks/useMobileTouchMove.js";
import br from "./hooks/useOriginScroll.js";
import Rr from "./hooks/useScrollDrag.js";
import Br from "./hooks/useScrollTo.js";
import Le from "./ScrollBar.js";
import { getSpinSize as Ce } from "./utils/scrollbarUtil.js";
var Lr = ["prefixCls", "className", "height", "itemHeight", "fullHeight", "style", "data", "children", "itemKey", "virtual", "direction", "scrollWidth", "component", "onScroll", "onVirtualScroll", "onVisibleChange", "innerProps", "extraRender", "styles", "showScrollBar"], Cr = [], Nr = { overflowY: "auto", overflowAnchor: "none" };
function Tr(o, Ne) {
  var J = o.prefixCls, x = J === void 0 ? "rc-virtual-list" : J, Te = o.className, f = o.height, g = o.itemHeight, Q = o.fullHeight, Pe = Q === void 0 || Q, Ie = o.style, D = o.data, Oe = o.children, N = o.itemKey, Ve = o.virtual, Ye = o.direction, c = o.scrollWidth, U = o.component, De = U === void 0 ? "div" : U, Z = o.onScroll, $ = o.onVirtualScroll, k = o.onVisibleChange, We = o.innerProps, ee = o.extraRender, d = o.styles, re = o.showScrollBar, te = re === void 0 ? "optional" : re, Xe = pr(o, Lr), M = a.useCallback(function(e) {
    return typeof N == "function" ? N(e) : e == null ? void 0 : e[N];
  }, [N]), Ke = zr(M), T = w(Ke, 4), je = T[0], oe = T[1], u = T[2], Ae = T[3], h = !(Ve === !1 || !f || !g), _e = a.useMemo(function() {
    return Object.values(u.maps).reduce(function(e, r) {
      return e + r;
    }, 0);
  }, [u.id, u.maps]), H = h && D && (Math.max(g * D.length, _e) > f || !!c), p = Ye === "rtl", Ge = Sr(x, be({}, "".concat(x, "-rtl"), p), Te), i = D || Cr, z = v(), ne = v(), ie = v(), qe = F(0), le = w(qe, 2), S = le[0], Fe = le[1], Je = F(0), ae = w(Je, 2), s = ae[0], P = ae[1], Qe = F(!1), ce = w(Qe, 2), Ue = ce[0], fe = ce[1], ue = function() {
    fe(!0);
  }, se = function() {
    fe(!1);
  }, Ze = { getKey: M };
  function E(e) {
    Fe(function(r) {
      var t = function(l) {
        var n = l;
        return Number.isNaN(X.current) || (n = Math.min(n, X.current)), n = Math.max(n, 0), n;
      }(typeof e == "function" ? e(r) : e);
      return z.current.scrollTop = t, t;
    });
  }
  var me = v({ start: 0, end: i.length }), $e = v(), ke = wr(i, M), er = w(ke, 1)[0];
  $e.current = er;
  var I = a.useMemo(function() {
    if (!h) return { scrollHeight: void 0, start: 0, end: i.length - 1, offset: void 0 };
    var e;
    if (!H) return { scrollHeight: ((e = ne.current) === null || e === void 0 ? void 0 : e.offsetHeight) || 0, start: 0, end: i.length - 1, offset: void 0 };
    for (var r, t, l, n = 0, sr = i.length, L = 0; L < sr; L += 1) {
      var mr = i[L], vr = M(mr), Ee = u.get(vr), G = n + (Ee === void 0 ? g : Ee);
      G >= S && r === void 0 && (r = L, t = n), G > S + f && l === void 0 && (l = L), n = G;
    }
    return r === void 0 && (r = 0, t = 0, l = Math.ceil(f / g)), l === void 0 && (l = i.length - 1), { scrollHeight: n, start: r, end: l = Math.min(l + 1, i.length - 1), offset: t };
  }, [H, h, S, i, Ae, f]), y = I.scrollHeight, b = I.start, R = I.end, ve = I.offset;
  me.current.start = b, me.current.end = R, a.useLayoutEffect(function() {
    var e = u.getRecord();
    if (e.size === 1) {
      var r = Array.from(e)[0], t = i[b];
      if (t && M(t) === r) {
        var l = u.get(r) - g;
        E(function(n) {
          return n + l;
        });
      }
    }
    u.resetRecord();
  }, [y]);
  var rr = a.useState({ width: 0, height: f }), de = w(rr, 2), m = de[0], tr = de[1], he = v(), pe = v(), or = a.useMemo(function() {
    return Ce(m.width, c);
  }, [m.width, c]), nr = a.useMemo(function() {
    return Ce(m.height, y);
  }, [m.height, y]), W = y - f, X = v(W);
  X.current = W;
  var O = S <= 0, V = S >= W, Se = s <= 0, ge = s >= c, ir = br(O, V, Se, ge), K = function() {
    return { x: p ? -s : s, y: S };
  }, j = v(K()), Y = Re(function(e) {
    if ($) {
      var r = C(C({}, K()), e);
      j.current.x === r.x && j.current.y === r.y || ($(r), j.current = r);
    }
  });
  function Me(e, r) {
    var t = e;
    r ? (Be(function() {
      P(t);
    }), Y()) : E(t);
  }
  var A = function(e) {
    var r = e, t = c ? c - m.width : 0;
    return r = Math.max(r, 0), r = Math.min(r, t);
  }, lr = Re(function(e, r) {
    r ? (Be(function() {
      P(function(t) {
        return A(t + (p ? -e : e));
      });
    }), Y()) : E(function(t) {
      return t + e;
    });
  }), ar = xr(h, O, V, Se, ge, !!c, lr), ye = w(ar, 2), _ = ye[0], we = ye[1];
  Er(h, z, function(e, r, t, l) {
    var n = l;
    return !ir(e, r, t) && (!n || !n._virtualHandled) && (n && (n._virtualHandled = !0), _({ preventDefault: function() {
    }, deltaX: e ? r : 0, deltaY: e ? 0 : r }), !0);
  }), Rr(H, z, function(e) {
    E(function(r) {
      return r + e;
    });
  }), q(function() {
    function e(t) {
      var l = O && t.detail < 0, n = V && t.detail > 0;
      !h || l || n || t.preventDefault();
    }
    var r = z.current;
    return r.addEventListener("wheel", _, { passive: !1 }), r.addEventListener("DOMMouseScroll", we, { passive: !0 }), r.addEventListener("MozMousePixelScroll", e, { passive: !1 }), function() {
      r.removeEventListener("wheel", _), r.removeEventListener("DOMMouseScroll", we), r.removeEventListener("MozMousePixelScroll", e);
    };
  }, [h, O, V]), q(function() {
    if (c) {
      var e = A(s);
      P(e), Y({ x: e });
    }
  }, [m.width, c]);
  var xe = function() {
    var e, r;
    (e = he.current) === null || e === void 0 || e.delayHidden(), (r = pe.current) === null || r === void 0 || r.delayHidden();
  }, He = Br(z, i, u, g, M, function() {
    return oe(!0);
  }, E, xe);
  a.useImperativeHandle(Ne, function() {
    return { nativeElement: ie.current, getScrollInfo: K, scrollTo: function(e) {
      var r;
      (r = e) && hr(r) === "object" && ("left" in r || "top" in r) ? (e.left !== void 0 && P(A(e.left)), He(e.top)) : He(e);
    } };
  }), q(function() {
    if (k) {
      var e = i.slice(b, R + 1);
      k(e, i);
    }
  }, [b, R, i]);
  var cr = Hr(i, M, u, g), fr = ee == null ? void 0 : ee({ start: b, end: R, virtual: H, offsetX: s, offsetY: ve, rtl: p, getSize: cr }), ur = yr(i, b, R, c, s, je, Oe, Ze), B = null;
  f && (B = C(be({}, Pe ? "height" : "maxHeight", f), Nr), h && (B.overflowY = "hidden", c && (B.overflowX = "hidden"), Ue && (B.pointerEvents = "none")));
  var ze = {};
  return p && (ze.dir = "rtl"), a.createElement("div", dr({ ref: ie, style: C(C({}, Ie), {}, { position: "relative" }), className: Ge }, ze, Xe), a.createElement(gr, { onResize: function(e) {
    tr({ width: e.offsetWidth, height: e.offsetHeight });
  } }, a.createElement(De, { className: "".concat(x, "-holder"), style: B, ref: z, onScroll: function(e) {
    var r = e.currentTarget.scrollTop;
    r !== S && E(r), Z == null || Z(e), Y();
  }, onMouseEnter: xe }, a.createElement(Mr, { prefixCls: x, height: y, offsetX: s, offsetY: ve, scrollWidth: c, onInnerResize: oe, ref: ne, innerProps: We, rtl: p, extra: fr }, ur))), H && y > f && a.createElement(Le, { ref: he, prefixCls: x, scrollOffset: S, scrollRange: y, rtl: p, onScroll: Me, onStartMove: ue, onStopMove: se, spinSize: nr, containerSize: m.height, style: d == null ? void 0 : d.verticalScrollBar, thumbStyle: d == null ? void 0 : d.verticalScrollBarThumb, showScrollBar: te }), H && c > m.width && a.createElement(Le, { ref: pe, prefixCls: x, scrollOffset: s, scrollRange: c, rtl: p, onScroll: Me, onStartMove: ue, onStopMove: se, spinSize: or, containerSize: m.width, horizontal: !0, style: d == null ? void 0 : d.horizontalScrollBar, thumbStyle: d == null ? void 0 : d.horizontalScrollBarThumb, showScrollBar: te }));
}
var Pr = a.forwardRef(Tr);
Pr.displayName = "List";
export {
  Tr as RawList,
  Pr as default
};
