// rc-tabs/es/TabNavList/index.js
import Ut from "../../../@babel/runtime/helpers/esm/extends.js";
import N from "../../../@babel/runtime/helpers/esm/defineProperty.js";
import _e from "../../../@babel/runtime/helpers/esm/toConsumableArray.js";
import J from "../../../@babel/runtime/helpers/esm/objectSpread2.js";
import l from "../../../@babel/runtime/helpers/esm/slicedToArray.js";
import ce from "../../../../_virtual/index.js";
import fe from "../../../rc-resize-observer/es/index.js";
import Ht from "../../../rc-util/es/hooks/useEvent.js";
import { useComposeRef as Pt } from "../../../rc-util/es/ref.js";
import * as a from "react";
import { useRef as b, useState as k, useEffect as Q } from "react";
import Wt from "../TabContext.js";
import It from "../hooks/useIndicator.js";
import jt from "../hooks/useOffsets.js";
import Je from "../hooks/useSyncState.js";
import qt from "../hooks/useTouchMove.js";
import Ft, { useUpdateState as Gt } from "../hooks/useUpdate.js";
import Vt from "../hooks/useVisibleRange.js";
import { stringify as Qe, getRemovable as _t, genDataNodeKey as Jt } from "../util.js";
import Qt from "./AddButton.js";
import Xe from "./ExtraContent.js";
import Xt from "./OperationNode.js";
import Yt from "./TabNode.js";
var M = function(r) {
  var E = r.current || {}, S = E.offsetWidth, K = S === void 0 ? 0 : S, O = E.offsetHeight, Y = O === void 0 ? 0 : O;
  if (r.current) {
    var g = r.current.getBoundingClientRect(), c = g.width, U = g.height;
    if (Math.abs(c - K) < 1) return [c, U];
  }
  return [K, Y];
}, X = function(r, E) {
  return r[E ? 0 : 1];
}, yn = a.forwardRef(function(r, E) {
  var S = r.className, K = r.style, O = r.id, Y = r.animated, g = r.activeKey, c = r.rtl, U = r.extra, H = r.editable, R = r.locale, ue = r.tabPosition, se = r.tabBarGutter, Ye = r.children, me = r.onTabClick, P = r.onTabScroll, Ze = r.indicator, pe = a.useContext(Wt), f = pe.prefixCls, s = pe.tabs, de = b(null), ve = b(null), he = b(null), T = b(null), W = b(null), ge = b(null), be = b(null), o = ue === "top" || ue === "bottom", $e = Je(0, function(e, t) {
    o && P && P({ direction: e > t ? "left" : "right" });
  }), we = l($e, 2), u = we[0], Z = we[1], Ae = Je(0, function(e, t) {
    !o && P && P({ direction: e > t ? "top" : "bottom" });
  }), ye = l(Ae, 2), w = ye[0], $ = ye[1], et = k([0, 0]), ke = l(et, 2), tt = ke[0], nt = ke[1], rt = k([0, 0]), Ee = l(rt, 2), xe = Ee[0], ot = Ee[1], it = k([0, 0]), Ce = l(it, 2), at = Ce[0], lt = Ce[1], ct = k([0, 0]), De = l(ct, 2), ft = De[0], ut = De[1], st = Gt(/* @__PURE__ */ new Map()), Me = l(st, 2), mt = Me[0], pt = Me[1], I = jt(s, mt, xe[0]), A = X(tt, o), j = X(xe, o), ee = X(at, o), Re = X(ft, o), Te = Math.floor(A) < Math.floor(j + ee), m = Te ? A - Re : A - ee, dt = "".concat(f, "-nav-operations-hidden"), y = 0, x = 0;
  function te(e) {
    return e < y ? y : e > x ? x : e;
  }
  o && c ? (y = 0, x = Math.max(0, j - m)) : (y = Math.min(0, m - j), x = 0);
  var ne = b(null), vt = k(), Le = l(vt, 2), q = Le[0], ze = Le[1];
  function re() {
    ze(Date.now());
  }
  function oe() {
    ne.current && clearTimeout(ne.current);
  }
  qt(T, function(e, t) {
    function n(i, d) {
      i(function(v) {
        return te(v + d);
      });
    }
    return !!Te && (o ? n(Z, e) : n($, t), oe(), re(), !0);
  }), Q(function() {
    return oe(), q && (ne.current = setTimeout(function() {
      ze(0);
    }, 100)), oe;
  }, [q]);
  var ht = Vt(I, m, o ? u : w, j, ee, Re, J(J({}, r), {}, { tabs: s })), Be = l(ht, 2), gt = Be[0], bt = Be[1], Ne = Ht(function() {
    var e = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : g, t = I.get(e) || { width: 0, height: 0, left: 0, right: 0, top: 0 };
    if (o) {
      var n = u;
      c ? t.right < u ? n = t.right : t.right + t.width > u + m && (n = t.right + t.width - m) : t.left < -u ? n = -t.left : t.left + t.width > -u + m && (n = -(t.left + t.width - m)), $(0), Z(te(n));
    } else {
      var i = w;
      t.top < -w ? i = -t.top : t.top + t.height > -w + m && (i = -(t.top + t.height - m)), Z(0), $(te(i));
    }
  }), wt = k(), Se = l(wt, 2), C = Se[0], L = Se[1], yt = k(!1), Ke = l(yt, 2), kt = Ke[0], Oe = Ke[1], p = s.filter(function(e) {
    return !e.disabled;
  }).map(function(e) {
    return e.key;
  }), D = function(e) {
    var t = p.indexOf(C || g), n = p.length, i = p[(t + e + n) % n];
    L(i);
  }, Et = function(e) {
    var t = e.code, n = c && o, i = p[0], d = p[p.length - 1];
    switch (t) {
      case "ArrowLeft":
        o && D(n ? 1 : -1);
        break;
      case "ArrowRight":
        o && D(n ? -1 : 1);
        break;
      case "ArrowUp":
        e.preventDefault(), o || D(-1);
        break;
      case "ArrowDown":
        e.preventDefault(), o || D(1);
        break;
      case "Home":
        e.preventDefault(), L(i);
        break;
      case "End":
        e.preventDefault(), L(d);
        break;
      case "Enter":
      case "Space":
        e.preventDefault(), me(C, e);
        break;
      case "Backspace":
      case "Delete":
        var v = p.indexOf(C), h = s.find(function(le) {
          return le.key === C;
        });
        _t(h == null ? void 0 : h.closable, h == null ? void 0 : h.closeIcon, H, h == null ? void 0 : h.disabled) && (e.preventDefault(), e.stopPropagation(), H.onEdit("remove", { key: C, event: e }), v === p.length - 1 ? D(-1) : D(1));
    }
  }, F = {};
  o ? F[c ? "marginRight" : "marginLeft"] = se : F.marginTop = se;
  var Ue = s.map(function(e, t) {
    var n = e.key;
    return a.createElement(Yt, { id: O, prefixCls: f, key: n, tab: e, style: t === 0 ? void 0 : F, closable: e.closable, editable: H, active: n === g, focus: n === C, renderWrapper: Ye, removeAriaLabel: R == null ? void 0 : R.removeAriaLabel, tabCount: p.length, currentPosition: t + 1, onClick: function(i) {
      me(n, i);
    }, onKeyDown: Et, onFocus: function() {
      kt || L(n), Ne(n), re(), T.current && (c || (T.current.scrollLeft = 0), T.current.scrollTop = 0);
    }, onBlur: function() {
      L(void 0);
    }, onMouseDown: function() {
      Oe(!0);
    }, onMouseUp: function() {
      Oe(!1);
    } });
  }), He = function() {
    return pt(function() {
      var e, t = /* @__PURE__ */ new Map(), n = (e = W.current) === null || e === void 0 ? void 0 : e.getBoundingClientRect();
      return s.forEach(function(i) {
        var d, v = i.key, h = (d = W.current) === null || d === void 0 ? void 0 : d.querySelector('[data-node-key="'.concat(Jt(v), '"]'));
        if (h) {
          var le = function(B, Fe) {
            var Ge = B.offsetWidth, zt = B.offsetHeight, Bt = B.offsetTop, Nt = B.offsetLeft, _ = B.getBoundingClientRect(), Ve = _.width, St = _.height, Kt = _.left, Ot = _.top;
            return Math.abs(Ve - Ge) < 1 ? [Ve, St, Kt - Fe.left, Ot - Fe.top] : [Ge, zt, Nt, Bt];
          }(h, n), V = l(le, 4), Mt = V[0], Rt = V[1], Tt = V[2], Lt = V[3];
          t.set(v, { width: Mt, height: Rt, left: Tt, top: Lt });
        }
      }), t;
    });
  };
  Q(function() {
    He();
  }, [s.map(function(e) {
    return e.key;
  }).join("_")]);
  var G = Ft(function() {
    var e = M(de), t = M(ve), n = M(he);
    nt([e[0] - t[0] - n[0], e[1] - t[1] - n[1]]);
    var i = M(be);
    lt(i);
    var d = M(ge);
    ut(d);
    var v = M(W);
    ot([v[0] - i[0], v[1] - i[1]]), He();
  }), xt = s.slice(0, gt), Ct = s.slice(bt + 1), Pe = [].concat(_e(xt), _e(Ct)), We = I.get(g), Dt = It({ activeTabOffset: We, horizontal: o, indicator: Ze, rtl: c }).style;
  Q(function() {
    Ne();
  }, [g, y, x, Qe(We), Qe(I), o]), Q(function() {
    G();
  }, [c]);
  var ie, ae, Ie, je, qe = !!Pe.length, z = "".concat(f, "-nav-wrap");
  return o ? c ? (ae = u > 0, ie = u !== x) : (ie = u < 0, ae = u !== y) : (Ie = w < 0, je = w !== y), a.createElement(fe, { onResize: G }, a.createElement("div", { ref: Pt(E, de), role: "tablist", "aria-orientation": o ? "horizontal" : "vertical", className: ce("".concat(f, "-nav"), S), style: K, onKeyDown: function() {
    re();
  } }, a.createElement(Xe, { ref: ve, position: "left", extra: U, prefixCls: f }), a.createElement(fe, { onResize: G }, a.createElement("div", { className: ce(z, N(N(N(N({}, "".concat(z, "-ping-left"), ie), "".concat(z, "-ping-right"), ae), "".concat(z, "-ping-top"), Ie), "".concat(z, "-ping-bottom"), je)), ref: T }, a.createElement(fe, { onResize: G }, a.createElement("div", { ref: W, className: "".concat(f, "-nav-list"), style: { transform: "translate(".concat(u, "px, ").concat(w, "px)"), transition: q ? "none" : void 0 } }, Ue, a.createElement(Qt, { ref: be, prefixCls: f, locale: R, editable: H, style: J(J({}, Ue.length === 0 ? void 0 : F), {}, { visibility: qe ? "hidden" : null }) }), a.createElement("div", { className: ce("".concat(f, "-ink-bar"), N({}, "".concat(f, "-ink-bar-animated"), Y.inkBar)), style: Dt }))))), a.createElement(Xt, Ut({}, r, { removeAriaLabel: R == null ? void 0 : R.removeAriaLabel, ref: ge, prefixCls: f, tabs: Pe, className: !qe && dt, tabMoving: !!q })), a.createElement(Xe, { ref: he, position: "right", extra: U, prefixCls: f })));
});
export {
  yn as default
};
