// @secondcloset/types/src/storage/appointment/appointment.ts
import { AppointmentItem } from "./appointmentItem";
import {
  AppointmentType,
  AppointmentSubtype,
  DeliveryType,
  ParcelServiceLevel,
} from "./appointmentType";
import { AppointmentStatus } from "./appointmentStatus";
import { Address } from "../address";
import { BaseImage, BaseNote, ServiceArea } from "../../_common";
import { ScServiceCode } from "../../fulfillment";

export interface Appointment {
  air_skip?: boolean;
  address: Address;
  arrived_time?: string;
  bulky_order_only: boolean;
  claim_number?: string;
  created_at: string;
  customer_segment: string;
  date: string;
  delivery_service_level?: ScServiceCode;
  driver_id?: string;
  end_time?: string;
  estimated_time: number;
  failed_notes?: string;
  failed_reason?: string;
  field_ops_notes?: BaseNote[];
  formatted_timeslot: string;
  id: string;
  immediate: true;
  incomplete_at?: string;
  issue_images?: BaseImage[];
  job_type: AppointmentType;
  job_subtype?: AppointmentSubtype;
  location?: Location;
  movers?: any;
  notes?: string;
  num_movers?: number;
  number: string;
  order_id: string;
  parcel_service_level?: ParcelServiceLevel;
  rescheduled: boolean;
  rescheduled_appointment_id?: string | null;
  signature_image?: string;
  signature_url?: string;
  signed_at?: string;
  space_order: boolean;
  start_time?: string;
  status: AppointmentStatus;
  temporary: boolean;
  timerange: AppointmentSlot;
  type: string;
  updated_at: string;
  user_code: string;
  user_email: string;
  user_firstname: string;
  user_id: string;
  user_lastname: string;
  items: AppointmentItem[];
  verification_images?: BaseImage[];
  delivery_type?: DeliveryType;
  signature_required?: boolean;
  recommended_route_user_count: number | null;
}

export type AppointmentSlot =
  | "07-10"
  | "10-13"
  | "13-16"
  | "16-19"
  | "19-22"
  | "10-22"
  | "09-13"
  | "13-17"
  | "17-21";

export interface AppointmentNote {
  id: string;
  text: string;
  created_at: string;
  updated_at: string;
}

export interface AppointmentTask {
  id: string;
  task_type: string;
  status: string;
  appointment_id: string;
  notes: string;
  assignee: string | null;
}

export interface Location {
  address: string;
  cell_phone_number: string | null;
  city: string;
  contact_name: string;
  coordinates: { lat: number; lng: number };
  country: string;
  created_at: string;
  customer_id: string;
  customer_phone_number: string;
  evening_phone_number: string | null;
  id: string;
  phone_number: string;
  postal_code: string;
  province: string;
  service_area: ServiceArea | null;
  updated_at: string;
}
