// resize-observer-polyfill/dist/ResizeObserver.es.js
var E = function() {
  if (typeof Map < "u") return Map;
  function e(t, n) {
    var r = -1;
    return t.some(function(i, o) {
      return i[0] === n && (r = o, !0);
    }), r;
  }
  return function() {
    function t() {
      this.__entries__ = [];
    }
    return Object.defineProperty(t.prototype, "size", { get: function() {
      return this.__entries__.length;
    }, enumerable: !0, configurable: !0 }), t.prototype.get = function(n) {
      var r = e(this.__entries__, n), i = this.__entries__[r];
      return i && i[1];
    }, t.prototype.set = function(n, r) {
      var i = e(this.__entries__, n);
      ~i ? this.__entries__[i][1] = r : this.__entries__.push([n, r]);
    }, t.prototype.delete = function(n) {
      var r = this.__entries__, i = e(r, n);
      ~i && r.splice(i, 1);
    }, t.prototype.has = function(n) {
      return !!~e(this.__entries__, n);
    }, t.prototype.clear = function() {
      this.__entries__.splice(0);
    }, t.prototype.forEach = function(n, r) {
      r === void 0 && (r = null);
      for (var i = 0, o = this.__entries__; i < o.length; i++) {
        var s = o[i];
        n.call(r, s[1], s[0]);
      }
    }, t;
  }();
}(), _ = typeof window < "u" && typeof document < "u" && window.document === document, l = typeof global < "u" && global.Math === Math ? global : typeof self < "u" && self.Math === Math ? self : typeof window < "u" && window.Math === Math ? window : Function("return this")(), x = typeof requestAnimationFrame == "function" ? requestAnimationFrame.bind(l) : function(e) {
  return setTimeout(function() {
    return e(Date.now());
  }, 1e3 / 60);
}, z = ["top", "right", "bottom", "left", "width", "height", "size", "weight"], k = typeof MutationObserver < "u", C = function() {
  function e() {
    this.connected_ = !1, this.mutationEventsAdded_ = !1, this.mutationsObserver_ = null, this.observers_ = [], this.onTransitionEnd_ = this.onTransitionEnd_.bind(this), this.refresh = /* @__PURE__ */ function(t, n) {
      var r = !1, i = !1, o = 0;
      function s() {
        r && (r = !1, t()), i && a();
      }
      function c() {
        x(s);
      }
      function a() {
        var h = Date.now();
        if (r) {
          if (h - o < 2) return;
          i = !0;
        } else r = !0, i = !1, setTimeout(c, n);
        o = h;
      }
      return a;
    }(this.refresh.bind(this), 20);
  }
  return e.prototype.addObserver = function(t) {
    ~this.observers_.indexOf(t) || this.observers_.push(t), this.connected_ || this.connect_();
  }, e.prototype.removeObserver = function(t) {
    var n = this.observers_, r = n.indexOf(t);
    ~r && n.splice(r, 1), !n.length && this.connected_ && this.disconnect_();
  }, e.prototype.refresh = function() {
    this.updateObservers_() && this.refresh();
  }, e.prototype.updateObservers_ = function() {
    var t = this.observers_.filter(function(n) {
      return n.gatherActive(), n.hasActive();
    });
    return t.forEach(function(n) {
      return n.broadcastActive();
    }), t.length > 0;
  }, e.prototype.connect_ = function() {
    _ && !this.connected_ && (document.addEventListener("transitionend", this.onTransitionEnd_), window.addEventListener("resize", this.refresh), k ? (this.mutationsObserver_ = new MutationObserver(this.refresh), this.mutationsObserver_.observe(document, { attributes: !0, childList: !0, characterData: !0, subtree: !0 })) : (document.addEventListener("DOMSubtreeModified", this.refresh), this.mutationEventsAdded_ = !0), this.connected_ = !0);
  }, e.prototype.disconnect_ = function() {
    _ && this.connected_ && (document.removeEventListener("transitionend", this.onTransitionEnd_), window.removeEventListener("resize", this.refresh), this.mutationsObserver_ && this.mutationsObserver_.disconnect(), this.mutationEventsAdded_ && document.removeEventListener("DOMSubtreeModified", this.refresh), this.mutationsObserver_ = null, this.mutationEventsAdded_ = !1, this.connected_ = !1);
  }, e.prototype.onTransitionEnd_ = function(t) {
    var n = t.propertyName, r = n === void 0 ? "" : n;
    z.some(function(i) {
      return !!~r.indexOf(i);
    }) && this.refresh();
  }, e.getInstance = function() {
    return this.instance_ || (this.instance_ = new e()), this.instance_;
  }, e.instance_ = null, e;
}(), w = function(e, t) {
  for (var n = 0, r = Object.keys(t); n < r.length; n++) {
    var i = r[n];
    Object.defineProperty(e, i, { value: t[i], enumerable: !1, writable: !1, configurable: !0 });
  }
  return e;
}, u = function(e) {
  return e && e.ownerDocument && e.ownerDocument.defaultView || l;
}, M = v(0, 0, 0, 0);
function d(e) {
  return parseFloat(e) || 0;
}
function O(e) {
  for (var t = [], n = 1; n < arguments.length; n++) t[n - 1] = arguments[n];
  return t.reduce(function(r, i) {
    return r + d(e["border-" + i + "-width"]);
  }, 0);
}
function D(e) {
  var t = e.clientWidth, n = e.clientHeight;
  if (!t && !n) return M;
  var r = u(e).getComputedStyle(e), i = function(f) {
    for (var m = {}, p = 0, g = ["top", "right", "bottom", "left"]; p < g.length; p++) {
      var y = g[p], T = f["padding-" + y];
      m[y] = d(T);
    }
    return m;
  }(r), o = i.left + i.right, s = i.top + i.bottom, c = d(r.width), a = d(r.height);
  if (r.boxSizing === "border-box" && (Math.round(c + o) !== t && (c -= O(r, "left", "right") + o), Math.round(a + s) !== n && (a -= O(r, "top", "bottom") + s)), !function(f) {
    return f === u(f).document.documentElement;
  }(e)) {
    var h = Math.round(c + o) - t, b = Math.round(a + s) - n;
    Math.abs(h) !== 1 && (c -= h), Math.abs(b) !== 1 && (a -= b);
  }
  return v(i.left, i.top, c, a);
}
var S = typeof SVGGraphicsElement < "u" ? function(e) {
  return e instanceof u(e).SVGGraphicsElement;
} : function(e) {
  return e instanceof u(e).SVGElement && typeof e.getBBox == "function";
};
function L(e) {
  return _ ? S(e) ? function(t) {
    var n = t.getBBox();
    return v(0, 0, n.width, n.height);
  }(e) : D(e) : M;
}
function v(e, t, n, r) {
  return { x: e, y: t, width: n, height: r };
}
var j = function() {
  function e(t) {
    this.broadcastWidth = 0, this.broadcastHeight = 0, this.contentRect_ = v(0, 0, 0, 0), this.target = t;
  }
  return e.prototype.isActive = function() {
    var t = L(this.target);
    return this.contentRect_ = t, t.width !== this.broadcastWidth || t.height !== this.broadcastHeight;
  }, e.prototype.broadcastRect = function() {
    var t = this.contentRect_;
    return this.broadcastWidth = t.width, this.broadcastHeight = t.height, t;
  }, e;
}(), W = function(e, t) {
  var n, r, i, o, s, c, a, h = (r = (n = t).x, i = n.y, o = n.width, s = n.height, c = typeof DOMRectReadOnly < "u" ? DOMRectReadOnly : Object, a = Object.create(c.prototype), w(a, { x: r, y: i, width: o, height: s, top: i, right: r + o, bottom: s + i, left: r }), a);
  w(this, { target: e, contentRect: h });
}, q = function() {
  function e(t, n, r) {
    if (this.activeObservations_ = [], this.observations_ = new E(), typeof t != "function") throw new TypeError("The callback provided as parameter 1 is not a function.");
    this.callback_ = t, this.controller_ = n, this.callbackCtx_ = r;
  }
  return e.prototype.observe = function(t) {
    if (!arguments.length) throw new TypeError("1 argument required, but only 0 present.");
    if (typeof Element < "u" && Element instanceof Object) {
      if (!(t instanceof u(t).Element)) throw new TypeError('parameter 1 is not of type "Element".');
      var n = this.observations_;
      n.has(t) || (n.set(t, new j(t)), this.controller_.addObserver(this), this.controller_.refresh());
    }
  }, e.prototype.unobserve = function(t) {
    if (!arguments.length) throw new TypeError("1 argument required, but only 0 present.");
    if (typeof Element < "u" && Element instanceof Object) {
      if (!(t instanceof u(t).Element)) throw new TypeError('parameter 1 is not of type "Element".');
      var n = this.observations_;
      n.has(t) && (n.delete(t), n.size || this.controller_.removeObserver(this));
    }
  }, e.prototype.disconnect = function() {
    this.clearActive(), this.observations_.clear(), this.controller_.removeObserver(this);
  }, e.prototype.gatherActive = function() {
    var t = this;
    this.clearActive(), this.observations_.forEach(function(n) {
      n.isActive() && t.activeObservations_.push(n);
    });
  }, e.prototype.broadcastActive = function() {
    if (this.hasActive()) {
      var t = this.callbackCtx_, n = this.activeObservations_.map(function(r) {
        return new W(r.target, r.broadcastRect());
      });
      this.callback_.call(t, n, t), this.clearActive();
    }
  }, e.prototype.clearActive = function() {
    this.activeObservations_.splice(0);
  }, e.prototype.hasActive = function() {
    return this.activeObservations_.length > 0;
  }, e;
}(), R = typeof WeakMap < "u" ? /* @__PURE__ */ new WeakMap() : new E(), A = function e(t) {
  if (!(this instanceof e)) throw new TypeError("Cannot call a class as a function.");
  if (!arguments.length) throw new TypeError("1 argument required, but only 0 present.");
  var n = C.getInstance(), r = new q(t, n, this);
  R.set(this, r);
};
["observe", "unobserve", "disconnect"].forEach(function(e) {
  A.prototype[e] = function() {
    var t;
    return (t = R.get(this))[e].apply(t, arguments);
  };
});
var B = l.ResizeObserver !== void 0 ? l.ResizeObserver : A;
export {
  B as default
};
