// @rc-component/trigger/es/index.js
import b from "../../../@babel/runtime/helpers/esm/objectSpread2.js";
import s from "../../../@babel/runtime/helpers/esm/slicedToArray.js";
import rn from "../../../@babel/runtime/helpers/esm/objectWithoutProperties.js";
import Le from "../../portal/es/Portal.js";
import G from "../../../../_virtual/index.js";
import tn from "../../../rc-resize-observer/es/index.js";
import { isDOM as Re } from "../../../rc-util/es/Dom/findDOMNode.js";
import { getShadowRoot as Te } from "../../../rc-util/es/Dom/shadow.js";
import h from "../../../rc-util/es/hooks/useEvent.js";
import un from "../../../rc-util/es/hooks/useId.js";
import P from "../../../rc-util/es/hooks/useLayoutEffect.js";
import an from "../../../rc-util/es/isMobile.js";
import * as r from "react";
import ln from "./Popup/index.js";
import sn from "./TriggerWrapper.js";
import Oe from "./context.js";
import pn from "./hooks/useAction.js";
import cn from "./hooks/useAlign.js";
import mn from "./hooks/useWatch.js";
import fn from "./hooks/useWinClick.js";
import { getAlignPopupClassName as gn, getMotion as Ve } from "./util.js";
var vn = ["prefixCls", "children", "action", "showAction", "hideAction", "popupVisible", "defaultPopupVisible", "onPopupVisibleChange", "afterPopupVisibleChange", "mouseEnterDelay", "mouseLeaveDelay", "focusDelay", "blurDelay", "mask", "maskClosable", "getPopupContainer", "forceRender", "autoDestroy", "destroyPopupOnHide", "popup", "popupClassName", "popupStyle", "popupPlacement", "builtinPlacements", "popupAlign", "zIndex", "stretch", "getPopupClassNameFromAlign", "fresh", "alignPoint", "onPopupClick", "onPopupAlign", "arrow", "popupMotion", "maskMotion", "popupTransitionName", "popupAnimation", "maskTransitionName", "maskAnimation", "className", "getTriggerDOMNode"];
function dn() {
  var ze = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : Le, K = r.forwardRef(function(o, Be) {
    var Q = o.prefixCls, y = Q === void 0 ? "rc-trigger-popup" : Q, Fe = o.children, U = o.action, He = U === void 0 ? "hover" : U, Ie = o.showAction, Xe = o.hideAction, C = o.popupVisible, Ye = o.defaultPopupVisible, Z = o.onPopupVisibleChange, $ = o.afterPopupVisibleChange, V = o.mouseEnterDelay, ee = o.mouseLeaveDelay, L = ee === void 0 ? 0.1 : ee, je = o.focusDelay, Je = o.blurDelay, oe = o.mask, ne = o.maskClosable, We = ne === void 0 || ne, _e = o.getPopupContainer, qe = o.forceRender, Ge = o.autoDestroy, Ke = o.destroyPopupOnHide, Qe = o.popup, Ue = o.popupClassName, Ze = o.popupStyle, z = o.popupPlacement, re = o.builtinPlacements, M = re === void 0 ? {} : re, te = o.popupAlign, $e = o.zIndex, ue = o.stretch, B = o.getPopupClassNameFromAlign, eo = o.fresh, k = o.alignPoint, oo = o.onPopupClick, no = o.onPopupAlign, F = o.arrow, ro = o.popupMotion, to = o.maskMotion, uo = o.popupTransitionName, io = o.popupAnimation, ao = o.maskTransitionName, lo = o.maskAnimation, ie = o.className, so = o.getTriggerDOMNode, H = rn(o, vn), po = Ge || Ke || !1, co = r.useState(!1), ae = s(co, 2), mo = ae[0], fo = ae[1];
    P(function() {
      fo(an());
    }, []);
    var le = r.useRef({}), A = r.useContext(Oe), go = r.useMemo(function() {
      return { registerSubPopup: function(e, n) {
        le.current[e] = n, A == null || A.registerSubPopup(e, n);
      } };
    }, [A]), vo = un(), ho = r.useState(null), se = s(ho, 2), p = se[0], Po = se[1], pe = r.useRef(null), yo = h(function(e) {
      pe.current = e, Re(e) && p !== e && Po(e), A == null || A.registerSubPopup(vo, e);
    }), Co = r.useState(null), ce = s(Co, 2), f = ce[0], Mo = ce[1], me = r.useRef(null), ko = h(function(e) {
      Re(e) && f !== e && (Mo(e), me.current = e);
    }), I = r.Children.only(Fe), c = (I == null ? void 0 : I.props) || {}, g = {}, Ao = h(function(e) {
      var n, i, u = f;
      return (u == null ? void 0 : u.contains(e)) || ((n = Te(u)) === null || n === void 0 ? void 0 : n.host) === e || e === u || (p == null ? void 0 : p.contains(e)) || ((i = Te(p)) === null || i === void 0 ? void 0 : i.host) === e || e === p || Object.values(le.current).some(function(t) {
        return (t == null ? void 0 : t.contains(e)) || e === t;
      });
    }), Do = Ve(y, ro, io, uo), No = Ve(y, to, lo, ao), bo = r.useState(Ye || !1), fe = s(bo, 2), Eo = fe[0], ge = fe[1], a = C ?? Eo, So = h(function(e) {
      C === void 0 && ge(e);
    });
    P(function() {
      ge(C || !1);
    }, [C]);
    var D = r.useRef(a);
    D.current = a;
    var E = r.useRef([]);
    E.current = [];
    var ve = h(function(e) {
      var n;
      So(e), ((n = E.current[E.current.length - 1]) !== null && n !== void 0 ? n : a) !== e && (E.current.push(e), Z == null || Z(e));
    }), de = r.useRef(), he = function() {
      clearTimeout(de.current);
    }, m = function(e) {
      var n = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : 0;
      he(), n === 0 ? ve(e) : de.current = setTimeout(function() {
        ve(e);
      }, 1e3 * n);
    };
    r.useEffect(function() {
      return he;
    }, []);
    var wo = r.useState(!1), Pe = s(wo, 2), X = Pe[0], ye = Pe[1];
    P(function(e) {
      e && !a || ye(!0);
    }, [a]);
    var xo = r.useState(null), Ce = s(xo, 2), Y = Ce[0], Me = Ce[1], Ro = r.useState(null), ke = s(Ro, 2), j = ke[0], To = ke[1], S = function(e) {
      To([e.clientX, e.clientY]);
    }, Oo = cn(a, p, k && j !== null ? j : f, z, M, te, no), l = s(Oo, 11), Vo = l[0], Lo = l[1], zo = l[2], Bo = l[3], Fo = l[4], Ho = l[5], Io = l[6], Xo = l[7], Yo = l[8], w = l[9], J = l[10], jo = pn(mo, He, Ie, Xe), Ae = s(jo, 2), x = Ae[0], N = Ae[1], De = x.has("click"), R = N.has("click") || N.has("contextMenu"), v = h(function() {
      X || J();
    });
    mn(a, f, p, v, function() {
      D.current && k && R && m(!1);
    }), P(function() {
      v();
    }, [j, z]), P(function() {
      !a || M != null && M[z] || v();
    }, [JSON.stringify(te)]);
    var Jo = r.useMemo(function() {
      var e = gn(M, y, w, k);
      return G(e, B == null ? void 0 : B(w));
    }, [w, B, M, y, k]);
    r.useImperativeHandle(Be, function() {
      return { nativeElement: me.current, popupElement: pe.current, forceAlign: v };
    });
    var Wo = r.useState(0), Ne = s(Wo, 2), _o = Ne[0], qo = Ne[1], Go = r.useState(0), be = s(Go, 2), Ko = be[0], Qo = be[1], Ee = function() {
      if (ue && f) {
        var e = f.getBoundingClientRect();
        qo(e.width), Qo(e.height);
      }
    };
    function d(e, n, i, u) {
      g[e] = function(t) {
        var T;
        u == null || u(t), m(n, i);
        for (var q = arguments.length, xe = new Array(q > 1 ? q - 1 : 0), O = 1; O < q; O++) xe[O - 1] = arguments[O];
        (T = c[e]) === null || T === void 0 || T.call.apply(T, [c, t].concat(xe));
      };
    }
    P(function() {
      Y && (J(), Y(), Me(null));
    }, [Y]), (De || R) && (g.onClick = function(e) {
      var n;
      D.current && R ? m(!1) : !D.current && De && (S(e), m(!0));
      for (var i = arguments.length, u = new Array(i > 1 ? i - 1 : 0), t = 1; t < i; t++) u[t - 1] = arguments[t];
      (n = c.onClick) === null || n === void 0 || n.call.apply(n, [c, e].concat(u));
    });
    var W, Se, Uo = fn(a, R, f, p, oe, We, Ao, m), Zo = x.has("hover"), $o = N.has("hover");
    Zo && (d("onMouseEnter", !0, V, function(e) {
      S(e);
    }), d("onPointerEnter", !0, V, function(e) {
      S(e);
    }), W = function(e) {
      (a || X) && p != null && p.contains(e.target) && m(!0, V);
    }, k && (g.onMouseMove = function(e) {
      var n;
      (n = c.onMouseMove) === null || n === void 0 || n.call(c, e);
    })), $o && (d("onMouseLeave", !1, L), d("onPointerLeave", !1, L), Se = function() {
      m(!1, L);
    }), x.has("focus") && d("onFocus", !0, je), N.has("focus") && d("onBlur", !1, Je), x.has("contextMenu") && (g.onContextMenu = function(e) {
      var n;
      D.current && N.has("contextMenu") ? m(!1) : (S(e), m(!0)), e.preventDefault();
      for (var i = arguments.length, u = new Array(i > 1 ? i - 1 : 0), t = 1; t < i; t++) u[t - 1] = arguments[t];
      (n = c.onContextMenu) === null || n === void 0 || n.call.apply(n, [c, e].concat(u));
    }), ie && (g.className = G(c.className, ie));
    var _ = b(b({}, c), g), we = {};
    ["onContextMenu", "onClick", "onMouseDown", "onTouchStart", "onMouseEnter", "onMouseLeave", "onFocus", "onBlur"].forEach(function(e) {
      H[e] && (we[e] = function() {
        for (var n, i = arguments.length, u = new Array(i), t = 0; t < i; t++) u[t] = arguments[t];
        (n = _[e]) === null || n === void 0 || n.call.apply(n, [_].concat(u)), H[e].apply(H, u);
      });
    });
    var en = r.cloneElement(I, b(b({}, _), we)), on = { x: Ho, y: Io }, nn = F ? b({}, F !== !0 ? F : {}) : null;
    return r.createElement(r.Fragment, null, r.createElement(tn, { disabled: !a, ref: ko, onResize: function() {
      Ee(), v();
    } }, r.createElement(sn, { getTriggerDOMNode: so }, en)), r.createElement(Oe.Provider, { value: go }, r.createElement(ln, { portal: ze, ref: yo, prefixCls: y, popup: Qe, className: G(Ue, Jo), style: Ze, target: f, onMouseEnter: W, onMouseLeave: Se, onPointerEnter: W, zIndex: $e, open: a, keepDom: X, fresh: eo, onClick: oo, onPointerDownCapture: Uo, mask: oe, motion: Do, maskMotion: No, onVisibleChanged: function(e) {
      ye(!1), J(), $ == null || $(e);
    }, onPrepare: function() {
      return new Promise(function(e) {
        Ee(), Me(function() {
          return e;
        });
      });
    }, forceRender: qe, autoDestroy: po, getPopupContainer: _e, align: w, arrow: nn, arrowPos: on, ready: Vo, offsetX: Lo, offsetY: zo, offsetR: Bo, offsetB: Fo, onAlign: v, stretch: ue, targetWidth: _o / Xo, targetHeight: Ko / Yo })));
  });
  return process.env.NODE_ENV !== "production" && (K.displayName = "Trigger"), K;
}
const Bn = dn(Le);
export {
  Bn as default,
  dn as generateTrigger
};
