// rc-table/es/Table.js
import v from "../../@babel/runtime/helpers/esm/defineProperty.js";
import T from "../../@babel/runtime/helpers/esm/extends.js";
import x from "../../@babel/runtime/helpers/esm/slicedToArray.js";
import u from "../../@babel/runtime/helpers/esm/objectSpread2.js";
import ve from "../../../_virtual/index.js";
import ot from "../../rc-resize-observer/es/index.js";
import tt from "../../rc-util/es/Dom/isVisible.js";
import { isStyleSupport as nt } from "../../rc-util/es/Dom/styleChecker.js";
import { getTargetScrollBarSize as uo } from "../../rc-util/es/getScrollBarSize.js";
import so from "../../rc-util/es/hooks/useEvent.js";
import mo from "../../rc-util/es/pickAttrs.js";
import rt from "../../rc-util/es/utils/get.js";
import { warningOnce as j } from "../../rc-util/es/warning.js";
import * as e from "react";
import at from "./Body/index.js";
import lt from "./ColGroup.js";
import { INTERNAL_HOOKS as xo, EXPAND_COLUMN as ct } from "./constant.js";
import it, { makeImmutable as ut } from "./context/TableContext.js";
import fo from "./FixedHolder/index.js";
import G, { FooterComponents as st } from "./Footer/index.js";
import mt from "./Footer/Summary.js";
import po from "./Header/Header.js";
import ft from "./hooks/useColumns/index.js";
import dt from "./hooks/useExpand.js";
import pt from "./hooks/useFixedInfo.js";
import { useLayoutState as yt, useTimeoutLock as vt } from "./hooks/useFrame.js";
import xt from "./hooks/useHover.js";
import Et from "./hooks/useSticky.js";
import ht from "./hooks/useStickyOffsets.js";
import yo from "./Panel/index.js";
import wt from "./stickyScrollBar.js";
import Ct from "./sugar/Column.js";
import Rt from "./sugar/ColumnGroup.js";
import { validNumberValue as Nt, getColumnsKey as gt, validateValue as vo } from "./utils/valueUtil.js";
import { getDOM as St } from "../../rc-util/es/Dom/findDOMNode.js";
var bt = "rc-table", kt = [], Lt = {};
function Tt() {
  return "No Data";
}
function It(J, ho) {
  var t = u({ rowKey: "key", prefixCls: bt, emptyText: Tt }, J), a = t.prefixCls, wo = t.className, xe = t.rowClassName, Co = t.style, Ee = t.data, B = t.rowKey, C = t.scroll, Q = t.tableLayout, E = t.direction, he = t.title, we = t.footer, Z = t.summary, Ce = t.caption, Ro = t.id, Re = t.showHeader, Ne = t.components, D = t.emptyText, ge = t.onRow, No = t.onHeaderRow, Se = t.onScroll, go = t.internalHooks, So = t.transformColumns, be = t.internalRefs, $ = t.tailor, ke = t.getContainerWidth, bo = t.sticky, Le = t.rowHoverable, Te = Le === void 0 || Le, p = Ee || kt, ee = !!p.length, N = go === xo;
  process.env.NODE_ENV !== "production" && (["onRowClick", "onRowDoubleClick", "onRowContextMenu", "onRowMouseEnter", "onRowMouseLeave"].forEach(function(o) {
    j(t[o] === void 0, "`".concat(o, "` is removed, please use `onRow` instead."));
  }), j(!("getBodyWrapper" in t), "`getBodyWrapper` is deprecated, please use custom `components` instead."));
  var z = e.useCallback(function(o, n) {
    return rt(Ne, o) || n;
  }, [Ne]), I = e.useMemo(function() {
    return typeof B == "function" ? B : function(o) {
      var n = o && o[B];
      return process.env.NODE_ENV !== "production" && j(n !== void 0, "Each record in table should have a unique `key` prop, or set `rowKey` to an unique primary key."), n;
    };
  }, [B]), Ie = z(["body"]), ko = xt(), oe = x(ko, 3), Oe = oe[0], Me = oe[1], He = oe[2], Lo = dt(t, p, I), g = x(Lo, 6), i = g[0], We = g[1], te = g[2], ne = g[3], Be = g[4], re = g[5], ae = C == null ? void 0 : C.x, To = e.useState(0), De = x(To, 2), K = De[0], Io = De[1], Oo = ft(u(u(u({}, t), i), {}, { expandable: !!i.expandedRowRender, columnTitle: i.columnTitle, expandedKeys: te, getRowKey: I, onTriggerExpand: re, expandIcon: ne, expandIconColumnIndex: i.expandIconColumnIndex, direction: E, scrollWidth: N && $ && typeof ae == "number" ? ae : null, clientWidth: K }), N ? So : null), _ = x(Oo, 4), O = _[0], r = _[1], ze = _[2], Mo = _[3], h = ze ?? ae, Ke = e.useMemo(function() {
    return { columns: O, flattenColumns: r };
  }, [O, r]), R = e.useRef(), le = e.useRef(), l = e.useRef(), _e = e.useRef();
  e.useImperativeHandle(ho, function() {
    return { nativeElement: R.current, scrollTo: function(o) {
      var n;
      if (l.current instanceof HTMLElement) {
        var s = o.index, c = o.top, Y = o.key;
        if (Nt(c)) {
          var m;
          (m = l.current) === null || m === void 0 || m.scrollTo({ top: c });
        } else {
          var L, H = Y ?? I(p[s]);
          (L = l.current.querySelector('[data-row-key="'.concat(H, '"]'))) === null || L === void 0 || L.scrollIntoView();
        }
      } else (n = l.current) !== null && n !== void 0 && n.scrollTo && l.current.scrollTo(o);
    } };
  });
  var ce, V, ie, Ve = e.useRef(), Ho = e.useState(!1), Fe = x(Ho, 2), Wo = Fe[0], F = Fe[1], Bo = e.useState(!1), Ae = x(Bo, 2), Do = Ae[0], A = Ae[1], zo = yt(/* @__PURE__ */ new Map()), Pe = x(zo, 2), Ko = Pe[0], _o = Pe[1], Xe = gt(r).map(function(o) {
    return Ko.get(o);
  }), qe = e.useMemo(function() {
    return Xe;
  }, [Xe.join("_")]), P = ht(qe, r, E), f = C && vo(C.y), y = C && vo(h) || !!i.fixed, S = y && r.some(function(o) {
    return o.fixed;
  }), ue = e.useRef(), b = Et(bo, a), w = b.isSticky, Vo = b.offsetHeader, Fo = b.offsetSummary, Ao = b.offsetScroll, Po = b.stickyClassName, Xo = b.container, d = e.useMemo(function() {
    return Z == null ? void 0 : Z(p);
  }, [Z, p]), M = (f || w) && e.isValidElement(d) && d.type === mt && d.props.fixed;
  f && (V = { overflowY: ee ? "scroll" : "auto", maxHeight: C.y }), y && (ce = { overflowX: "auto" }, f || (V = { overflowY: "hidden" }), ie = { width: h === !0 ? "auto" : h, minWidth: "100%" });
  var Ue = e.useCallback(function(o, n) {
    tt(R.current) && _o(function(s) {
      if (s.get(o) !== n) {
        var c = new Map(s);
        return c.set(o, n), c;
      }
      return s;
    });
  }, []), qo = vt(), Ye = x(qo, 2), Uo = Ye[0], je = Ye[1];
  function X(o, n) {
    n && (typeof n == "function" ? n(o) : n.scrollLeft !== o && (n.scrollLeft = o, n.scrollLeft !== o && setTimeout(function() {
      n.scrollLeft = o;
    }, 0)));
  }
  var k = so(function(o) {
    var n, s = o.currentTarget, c = o.scrollLeft, Y = E === "rtl", m = typeof c == "number" ? c : s.scrollLeft, L = s || Lt;
    je() && je() !== L || (Uo(L), X(m, le.current), X(m, l.current), X(m, Ve.current), X(m, (n = ue.current) === null || n === void 0 ? void 0 : n.setScrollLeft));
    var H = s || le.current;
    if (H) {
      var pe = N && $ && typeof h == "number" ? h : H.scrollWidth, ye = H.clientWidth;
      if (pe === ye) return F(!1), void A(!1);
      Y ? (F(-m < pe - ye), A(-m > 0)) : (F(m > 0), A(m < pe - ye));
    }
  }), Yo = so(function(o) {
    k(o), Se == null || Se(o);
  }), Ge = function() {
    var o;
    y && l.current ? k({ currentTarget: St(l.current), scrollLeft: (o = l.current) === null || o === void 0 ? void 0 : o.scrollLeft }) : (F(!1), A(!1));
  }, Je = e.useRef(!1);
  e.useEffect(function() {
    Je.current && Ge();
  }, [y, Ee, O.length]), e.useEffect(function() {
    Je.current = !0;
  }, []);
  var jo = e.useState(0), Qe = x(jo, 2), q = Qe[0], Ze = Qe[1], Go = e.useState(!0), $e = x(Go, 2), eo = $e[0], Jo = $e[1];
  e.useEffect(function() {
    $ && N || (l.current instanceof Element ? Ze(uo(l.current).width) : Ze(uo(_e.current).width)), Jo(nt("position", "sticky"));
  }, []), e.useEffect(function() {
    N && be && (be.body.current = l.current);
  });
  var se, Qo = e.useCallback(function(o) {
    return e.createElement(e.Fragment, null, e.createElement(po, o), M === "top" && e.createElement(G, o, d));
  }, [M, d]), Zo = e.useCallback(function(o) {
    return e.createElement(G, o, d);
  }, [d]), oo = z(["table"], "table"), U = e.useMemo(function() {
    return Q || (S ? h === "max-content" ? "auto" : "fixed" : f || w || r.some(function(o) {
      return o.ellipsis;
    }) ? "fixed" : "auto");
  }, [f, S, r, Q, w]), me = { colWidths: qe, columCount: r.length, stickyOffsets: P, onHeaderRow: No, fixHeader: f, scroll: C }, to = e.useMemo(function() {
    return ee ? null : typeof D == "function" ? D() : D;
  }, [ee, D]), no = e.createElement(at, { data: p, measureColumnWidth: f || y || w }), ro = e.createElement(lt, { colWidths: r.map(function(o) {
    return o.width;
  }), columns: r }), ao = Ce != null ? e.createElement("caption", { className: "".concat(a, "-caption") }, Ce) : void 0, $o = mo(t, { data: !0 }), lo = mo(t, { aria: !0 });
  if (f || w) {
    var fe;
    typeof Ie == "function" ? (fe = Ie(p, { scrollbarSize: q, ref: l, onScroll: k }), me.colWidths = r.map(function(o, n) {
      var s = o.width, c = n === r.length - 1 ? s - q : s;
      return typeof c != "number" || Number.isNaN(c) ? (process.env.NODE_ENV !== "production" && j(t.columns.length === 0, "When use `components.body` with render props. Each column should have a fixed `width` value."), 0) : c;
    })) : fe = e.createElement("div", { style: u(u({}, ce), V), onScroll: Yo, ref: l, className: ve("".concat(a, "-body")) }, e.createElement(oo, T({ style: u(u({}, ie), {}, { tableLayout: U }) }, lo), ao, ro, no, !M && d && e.createElement(G, { stickyOffsets: P, flattenColumns: r }, d)));
    var co = u(u(u({ noData: !p.length, maxContentScroll: y && h === "max-content" }, me), Ke), {}, { direction: E, stickyClassName: Po, onScroll: k });
    se = e.createElement(e.Fragment, null, Re !== !1 && e.createElement(fo, T({}, co, { stickyTopOffset: Vo, className: "".concat(a, "-header"), ref: le }), Qo), fe, M && M !== "top" && e.createElement(fo, T({}, co, { stickyBottomOffset: Fo, className: "".concat(a, "-summary"), ref: Ve }), Zo), w && l.current && l.current instanceof Element && e.createElement(wt, { ref: ue, offsetScroll: Ao, scrollBodyRef: l, onScroll: k, container: Xo, direction: E }));
  } else se = e.createElement("div", { style: u(u({}, ce), V), className: ve("".concat(a, "-content")), onScroll: k, ref: l }, e.createElement(oo, T({ style: u(u({}, ie), {}, { tableLayout: U }) }, lo), ao, ro, Re !== !1 && e.createElement(po, T({}, me, Ke)), no, d && e.createElement(G, { stickyOffsets: P, flattenColumns: r }, d)));
  var de = e.createElement("div", T({ className: ve(a, wo, v(v(v(v(v(v(v(v(v(v({}, "".concat(a, "-rtl"), E === "rtl"), "".concat(a, "-ping-left"), Wo), "".concat(a, "-ping-right"), Do), "".concat(a, "-layout-fixed"), Q === "fixed"), "".concat(a, "-fixed-header"), f), "".concat(a, "-fixed-column"), S), "".concat(a, "-fixed-column-gapped"), S && Mo), "".concat(a, "-scroll-horizontal"), y), "".concat(a, "-has-fix-left"), r[0] && r[0].fixed), "".concat(a, "-has-fix-right"), r[r.length - 1] && r[r.length - 1].fixed === "right")), style: Co, id: Ro, ref: R }, $o), he && e.createElement(yo, { className: "".concat(a, "-title") }, he(p)), e.createElement("div", { ref: _e, className: "".concat(a, "-container") }, se), we && e.createElement(yo, { className: "".concat(a, "-footer") }, we(p)));
  y && (de = e.createElement(ot, { onResize: function(o) {
    var n, s = o.width;
    (n = ue.current) === null || n === void 0 || n.checkScrollBarVisible();
    var c = R.current ? R.current.offsetWidth : s;
    N && ke && R.current && (c = ke(R.current, c) || c), c !== K && (Ge(), Io(c));
  } }, de));
  var io = pt(r, P, E), et = e.useMemo(function() {
    return { scrollX: h, prefixCls: a, getComponent: z, scrollbarSize: q, direction: E, fixedInfoList: io, isSticky: w, supportSticky: eo, componentWidth: K, fixHeader: f, fixColumn: S, horizonScroll: y, tableLayout: U, rowClassName: xe, expandedRowClassName: i.expandedRowClassName, expandIcon: ne, expandableType: We, expandRowByClick: i.expandRowByClick, expandedRowRender: i.expandedRowRender, onTriggerExpand: re, expandIconColumnIndex: i.expandIconColumnIndex, indentSize: i.indentSize, allColumnsFixedLeft: r.every(function(o) {
      return o.fixed === "left";
    }), emptyNode: to, columns: O, flattenColumns: r, onColumnResize: Ue, hoverStartRow: Oe, hoverEndRow: Me, onHover: He, rowExpandable: i.rowExpandable, onRow: ge, getRowKey: I, expandedKeys: te, childrenColumnName: Be, rowHoverable: Te };
  }, [h, a, z, q, E, io, w, eo, K, f, S, y, U, xe, i.expandedRowClassName, ne, We, i.expandRowByClick, i.expandedRowRender, re, i.expandIconColumnIndex, i.indentSize, to, O, r, Ue, Oe, Me, He, i.rowExpandable, ge, I, te, Be, Te]);
  return e.createElement(it.Provider, { value: et }, de);
}
var Eo = e.forwardRef(It);
function Ot(J) {
  return ut(Eo, J);
}
process.env.NODE_ENV !== "production" && (Eo.displayName = "Table");
var W = Ot();
W.EXPAND_COLUMN = ct, W.INTERNAL_HOOKS = xo, W.Column = Ct, W.ColumnGroup = Rt, W.Summary = st;
export {
  bt as DEFAULT_PREFIX,
  W as default,
  Ot as genTable
};
