// @gobolt/genesis/dist/components/Typography/styles.js
import h from "styled-components";
import $ from "../../node_modules/antd/es/typography/index.js";
const m = ({ colors: t, typography: n }, o, p, r, i, a, g) => `
    &.ant-typography {
      font-family: ${((e) => `
    ${e.includes("digits") ? "'Roboto Mono', sans-serif" : "'Inter', sans-serif"} !important;
  `)(p)};
      color: ${((e, s, l) => e || s[l].active.color)(a, t, o)} !important;
      ${((e, s, l) => `
        font-size: ${e[l][s].fontSize}px !important;
        line-height: ${e[l][s].lineHeight};
        letter-spacing: ${e[l][s].letterSpacing}px;
        font-weight: ${e[l][s].fontWeight};
      `)(n, p, i)}
      margin-bottom: 0 !important;
      width: ${g ? "100%" : "auto"};
      line-height: 1;

      &:disabled {
        color: ${t.status.disabled.default} !important;
        userSelect: none;
        cursor: not-allowed;
      }
    } 
`, f = h($.Title)`
  ${({ theme: t, $themeType: n, variant: o, state: p, breakpoint: r, color: i, $isFullWidth: a }) => m(t, n, o, 0, r, i, a)}
`, y = h($.Paragraph)`
  ${({ theme: t, $themeType: n, variant: o, state: p, breakpoint: r, color: i, $isFullWidth: a }) => m(t, n, o, 0, r, i, a)}
`, u = h($.Text)`
  ${({ theme: t, $themeType: n, variant: o, state: p, breakpoint: r, color: i, $isFullWidth: a }) => m(t, n, o, 0, r, i, a)}
`;
export {
  f as Headline,
  y as Paragraph,
  u as Text
};
