// @gobolt/genesis/dist/components/Typography/styles.cjs
"use strict";Object.defineProperty(exports,Symbol.toStringTag,{value:"Module"});const h=require("styled-components"),$=require("../../node_modules/antd/es/typography/index.cjs"),g=({colors:t,typography:n},i,p,r,o,a,c)=>`
    &.ant-typography {
      font-family: ${(e=>`
    ${e.includes("digits")?"'Roboto Mono', sans-serif":"'Inter', sans-serif"} !important;
  `)(p)};
      color: ${((e,l,s)=>e||l[s].active.color)(a,t,i)} !important;
      ${((e,l,s)=>`
        font-size: ${e[s][l].fontSize}px !important;
        line-height: ${e[s][l].lineHeight};
        letter-spacing: ${e[s][l].letterSpacing}px;
        font-weight: ${e[s][l].fontWeight};
      `)(n,p,o)}
      margin-bottom: 0 !important;
      width: ${c?"100%":"auto"};
      line-height: 1;

      &:disabled {
        color: ${t.status.disabled.default} !important;
        userSelect: none;
        cursor: not-allowed;
      }
    } 
`,d=h($.Title)`
  ${({theme:t,$themeType:n,variant:i,state:p,breakpoint:r,color:o,$isFullWidth:a})=>g(t,n,i,0,r,o,a)}
`,m=h($.Paragraph)`
  ${({theme:t,$themeType:n,variant:i,state:p,breakpoint:r,color:o,$isFullWidth:a})=>g(t,n,i,0,r,o,a)}
`,u=h($.Text)`
  ${({theme:t,$themeType:n,variant:i,state:p,breakpoint:r,color:o,$isFullWidth:a})=>g(t,n,i,0,r,o,a)}
`;exports.Headline=d,exports.Paragraph=m,exports.Text=u;
