// @ant-design/cssinjs-utils/es/util/calc/CSSCalculator.js
import h from "../../../../../@babel/runtime/helpers/esm/typeof.js";
import f from "../../../../../@babel/runtime/helpers/esm/classCallCheck.js";
import y from "../../../../../@babel/runtime/helpers/esm/createClass.js";
import n from "../../../../../@babel/runtime/helpers/esm/assertThisInitialized.js";
import m from "../../../../../@babel/runtime/helpers/esm/inherits.js";
import p from "../../../../../@babel/runtime/helpers/esm/createSuper.js";
import u from "../../../../../@babel/runtime/helpers/esm/defineProperty.js";
import v from "./calculator.js";
var l = "CALC_UNIT", g = new RegExp(l, "g");
function c(i) {
  return typeof i == "number" ? "".concat(i).concat(l) : i;
}
var V = function() {
  m(s, v);
  var i = p(s);
  function s(t, o) {
    var r;
    f(this, s), r = i.call(this), u(n(r), "result", ""), u(n(r), "unitlessCssVar", void 0), u(n(r), "lowPriority", void 0);
    var e = h(t);
    return r.unitlessCssVar = o, t instanceof s ? r.result = "(".concat(t.result, ")") : e === "number" ? r.result = c(t) : e === "string" && (r.result = t), r;
  }
  return y(s, [{ key: "add", value: function(t) {
    return t instanceof s ? this.result = "".concat(this.result, " + ").concat(t.getResult()) : typeof t != "number" && typeof t != "string" || (this.result = "".concat(this.result, " + ").concat(c(t))), this.lowPriority = !0, this;
  } }, { key: "sub", value: function(t) {
    return t instanceof s ? this.result = "".concat(this.result, " - ").concat(t.getResult()) : typeof t != "number" && typeof t != "string" || (this.result = "".concat(this.result, " - ").concat(c(t))), this.lowPriority = !0, this;
  } }, { key: "mul", value: function(t) {
    return this.lowPriority && (this.result = "(".concat(this.result, ")")), t instanceof s ? this.result = "".concat(this.result, " * ").concat(t.getResult(!0)) : typeof t != "number" && typeof t != "string" || (this.result = "".concat(this.result, " * ").concat(t)), this.lowPriority = !1, this;
  } }, { key: "div", value: function(t) {
    return this.lowPriority && (this.result = "(".concat(this.result, ")")), t instanceof s ? this.result = "".concat(this.result, " / ").concat(t.getResult(!0)) : typeof t != "number" && typeof t != "string" || (this.result = "".concat(this.result, " / ").concat(t)), this.lowPriority = !1, this;
  } }, { key: "getResult", value: function(t) {
    return this.lowPriority || t ? "(".concat(this.result, ")") : this.result;
  } }, { key: "equal", value: function(t) {
    var o = this, r = (t || {}).unit, e = !0;
    return typeof r == "boolean" ? e = r : Array.from(this.unitlessCssVar).some(function(a) {
      return o.result.includes(a);
    }) && (e = !1), this.result = this.result.replace(g, e ? "px" : ""), this.lowPriority !== void 0 ? "calc(".concat(this.result, ")") : this.result;
  } }]), s;
}();
export {
  V as default
};
