// rc-util/es/Dom/dynamicCSS.js
import y from "../../../@babel/runtime/helpers/esm/objectSpread2.js";
import L from "./canUseDom.js";
import N from "./contains.js";
var S = "data-rc-order", b = "data-rc-priority", p = /* @__PURE__ */ new Map();
function A() {
  var e = (arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : {}).mark;
  return e ? e.startsWith("data-") ? e : "data-".concat(e) : "rc-util-key";
}
function g(e) {
  return e.attachTo ? e.attachTo : document.querySelector("head") || document.body;
}
function v(e) {
  return Array.from((p.get(e) || e).children).filter(function(t) {
    return t.tagName === "STYLE";
  });
}
function C(e) {
  var t = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {};
  if (!L()) return null;
  var r = t.csp, a = t.prepend, l = t.priority, u = l === void 0 ? 0 : l, i = /* @__PURE__ */ function(c) {
    return c === "queue" ? "prependQueue" : c ? "prepend" : "append";
  }(a), s = i === "prependQueue", n = document.createElement("style");
  n.setAttribute(S, i), s && u && n.setAttribute(b, "".concat(u)), r != null && r.nonce && (n.nonce = r == null ? void 0 : r.nonce), n.innerHTML = e;
  var o = g(t), m = o.firstChild;
  if (a) {
    if (s) {
      var d = (t.styles || v(o)).filter(function(c) {
        if (!["prepend", "prependQueue"].includes(c.getAttribute(S))) return !1;
        var f = Number(c.getAttribute(b) || 0);
        return u >= f;
      });
      if (d.length) return o.insertBefore(n, d[d.length - 1].nextSibling), n;
    }
    o.insertBefore(n, m);
  } else o.appendChild(n);
  return n;
}
function T(e) {
  var t = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {}, r = g(t);
  return (t.styles || v(r)).find(function(a) {
    return a.getAttribute(A(t)) === e;
  });
}
function H(e) {
  var t = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {}, r = T(e, t);
  r && g(t).removeChild(r);
}
function Q(e, t) {
  var r = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {}, a = g(r), l = v(a), u = y(y({}, r), {}, { styles: l });
  (function(d, c) {
    var f = p.get(d);
    if (!f || !N(document, f)) {
      var h = C("", c), M = h.parentNode;
      p.set(d, M), d.removeChild(h);
    }
  })(a, u);
  var i = T(t, u);
  if (i) {
    var s, n, o;
    return (s = u.csp) !== null && s !== void 0 && s.nonce && i.nonce !== ((n = u.csp) === null || n === void 0 ? void 0 : n.nonce) && (i.nonce = (o = u.csp) === null || o === void 0 ? void 0 : o.nonce), i.innerHTML !== e && (i.innerHTML = e), i;
  }
  var m = C(e, u);
  return m.setAttribute(A(u), t), m;
}
export {
  C as injectCSS,
  H as removeCSS,
  Q as updateCSS
};
