// @ant-design/cssinjs/es/StyleContext.js
import * as s from "react";
import "../../../rc-util/es/warning.js";
import h from "./Cache.js";
var m = "data-token-hash", o = "data-css-hash", A = "data-cache-path", t = "__cssinjs_instance__";
function u() {
  var a = Math.random().toString(12).slice(2);
  if (typeof document < "u" && document.head && document.body) {
    var d = document.body.querySelectorAll("style[".concat(o, "]")) || [], i = document.head.firstChild;
    Array.from(d).forEach(function(e) {
      e[t] = e[t] || a, e[t] === a && document.head.insertBefore(e, i);
    });
    var c = {};
    Array.from(document.querySelectorAll("style[".concat(o, "]"))).forEach(function(e) {
      var r, n = e.getAttribute(o);
      c[n] ? e[t] === a && ((r = e.parentNode) === null || r === void 0 || r.removeChild(e)) : c[n] = !0;
    });
  }
  return new h(a);
}
var _ = s.createContext({ hashPriority: "low", cache: u(), defaultCache: !0 });
export {
  A as ATTR_CACHE_PATH,
  o as ATTR_MARK,
  m as ATTR_TOKEN,
  t as CSS_IN_JS_INSTANCE,
  u as createCache,
  _ as default
};
