// rc-picker/es/hooks/useTimeConfig.js
import S from "../../../@babel/runtime/helpers/esm/objectSpread2.js";
import M from "../../../@babel/runtime/helpers/esm/slicedToArray.js";
import T from "../../../@babel/runtime/helpers/esm/typeof.js";
import { getRowFormat as k, toArray as O, pickProps as V } from "../utils/miscUtil.js";
import { fillTimeFormat as y } from "./useLocale.js";
function L(r, u, f) {
  return f ?? u.some(function(h) {
    return r.includes(h);
  });
}
var A = ["showNow", "showHour", "showMinute", "showSecond", "showMillisecond", "use12Hours", "hourStep", "minuteStep", "secondStep", "millisecondStep", "hideDisabledOptions", "defaultValue", "disabledHours", "disabledMinutes", "disabledSeconds", "disabledMilliseconds", "disabledTime", "changeOnScroll", "defaultOpenValue"];
function C(r) {
  return r && typeof r == "string";
}
function g(r, u, f, h) {
  return [r, u, f, h].some(function(w) {
    return w !== void 0;
  });
}
function b(r, u, f, h, w) {
  var n = u, o = f, e = h;
  if (r || n || o || e || w) {
    if (r) {
      var a, d, i, t = [n, o, e].some(function(c) {
        return c === !1;
      }), m = [n, o, e].some(function(c) {
        return c === !0;
      }), s = !!t || !m;
      n = (a = n) !== null && a !== void 0 ? a : s, o = (d = o) !== null && d !== void 0 ? d : s, e = (i = e) !== null && i !== void 0 ? i : s;
    }
  } else n = !0, o = !0, e = !0;
  return [n, o, e, w];
}
function N(r) {
  var u = r.showTime, f = function(c) {
    var p = V(c, A), H = c.format, v = c.picker, l = null;
    return H && (l = H, Array.isArray(l) && (l = l[0]), l = T(l) === "object" ? l.format : l), v === "time" && (p.format = l), [p, l];
  }(r), h = M(f, 2), w = h[0], n = h[1], o = u && T(u) === "object" ? u : {}, e = S(S({ defaultOpenValue: o.defaultOpenValue || o.defaultValue }, w), o), a = e.showMillisecond, d = e.showHour, i = e.showMinute, t = e.showSecond, m = b(g(d, i, t, a), d, i, t, a), s = M(m, 3);
  return d = s[0], i = s[1], t = s[2], [e, S(S({}, e), {}, { showHour: d, showMinute: i, showSecond: t, showMillisecond: a }), e.format, n];
}
function R(r, u, f, h, w) {
  if (r === "datetime" || r === "time") {
    for (var n = h, o = k(r, w, null), e = [u, f], a = 0; a < e.length; a += 1) {
      var d = O(e[a])[0];
      if (C(d)) {
        o = d;
        break;
      }
    }
    var i = n.showHour, t = n.showMinute, m = n.showSecond, s = n.showMillisecond, c = L(o, ["a", "A", "LT", "LLL", "LTS"], n.use12Hours), p = g(i, t, m, s);
    p || (i = L(o, ["H", "h", "k", "LT", "LLL"]), t = L(o, ["m", "LT", "LLL"]), m = L(o, ["s", "LTS"]), s = L(o, ["SSS"]));
    var H = b(p, i, t, m, s), v = M(H, 3);
    i = v[0], t = v[1], m = v[2];
    var l = u || y(i, t, m, s, c);
    return S(S({}, n), {}, { format: l, showHour: i, showMinute: t, showSecond: m, showMillisecond: s, use12Hours: c });
  }
  return null;
}
export {
  R as fillShowTimeConfig,
  N as getTimeProps
};
