// @ant-design/cssinjs/es/hooks/useStyleRegister.js
import rt from "../../../../@babel/runtime/helpers/esm/extends.js";
import G from "../../../../@babel/runtime/helpers/esm/defineProperty.js";
import z from "../../../../@babel/runtime/helpers/esm/objectSpread2.js";
import O from "../../../../@babel/runtime/helpers/esm/slicedToArray.js";
import M from "../../../../@babel/runtime/helpers/esm/toConsumableArray.js";
import W from "../../../../@babel/runtime/helpers/esm/typeof.js";
import nt from "../../../../@emotion/hash/dist/hash.browser.esm.js";
import { removeCSS as at, updateCSS as Y } from "../../../../rc-util/es/Dom/dynamicCSS.js";
import * as K from "react";
import et from "../../../../@emotion/unitless/dist/unitless.browser.esm.js";
import ot from "../linters/contentQuotesLinter.js";
import it from "../linters/hashedAnimationLinter.js";
import "../../../../rc-util/es/warning.js";
import ct, { ATTR_TOKEN as J, ATTR_MARK as Z, CSS_IN_JS_INSTANCE as st, ATTR_CACHE_PATH as ut } from "../StyleContext.js";
import { isClientSide as U, toStyleStr as X } from "../util/index.js";
import { existPath as lt, getStyleAndHash as ft, CSS_FILE_STYLE as mt } from "../util/cacheMapUtil.js";
import pt from "./useGlobalCache.js";
import { serialize as vt, stringify as ht } from "../../../../stylis/src/Serializer.js";
import { compile as yt } from "../../../../stylis/src/Parser.js";
var B = "_skip_check_", $ = "_multi_value_";
function q(n) {
  return vt(yt(n), ht).replace(/\{%%%\:[^;];}/g, ";");
}
function dt(n, m, s) {
  if (!m) return n;
  var e = ".".concat(m), _ = s === "low" ? ":where(".concat(e, ")") : e;
  return n.split(",").map(function(I) {
    var u, a = I.trim().split(/\s+/), o = a[0] || "", E = ((u = o.match(/^\w+/)) === null || u === void 0 ? void 0 : u[0]) || "";
    return [o = "".concat(E).concat(_).concat(o.slice(E.length))].concat(M(a.slice(1))).join(" ");
  }).join(",");
}
var St = function n(m) {
  var s = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {}, e = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : { root: !0, parentSelectors: [] }, _ = e.root, I = e.injectHash, u = e.parentSelectors, a = s.hashId, o = s.layer, E = s.path, i = s.hashPriority, j = s.transformers, L = j === void 0 ? [] : j, N = s.linters, w = N === void 0 ? [] : N, l = "", H = {};
  function D(d) {
    var f = d.getName(a);
    if (!H[f]) {
      var p = n(d.style, s, { root: !1, parentSelectors: u }), t = O(p, 1)[0];
      H[f] = "@keyframes ".concat(d.getName(a)).concat(t);
    }
  }
  var Q = function d(f) {
    var p = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : [];
    return f.forEach(function(t) {
      Array.isArray(t) ? d(t, p) : t && p.push(t);
    }), p;
  }(Array.isArray(m) ? m : [m]);
  return Q.forEach(function(d) {
    var f = typeof d != "string" || _ ? d : {};
    if (typeof f == "string") l += "".concat(f, `
`);
    else if (f._keyframe) D(f);
    else {
      var p = L.reduce(function(t, r) {
        var T;
        return (r == null || (T = r.visit) === null || T === void 0 ? void 0 : T.call(r, t)) || t;
      }, f);
      Object.keys(p).forEach(function(t) {
        var r = p[t];
        if (W(r) !== "object" || !r || t === "animationName" && r._keyframe || function(h) {
          return W(h) === "object" && h && (B in h || $ in h);
        }(r)) {
          let h = function(y, g) {
            process.env.NODE_ENV === "production" || W(r) === "object" && r != null && r[B] || [ot, it].concat(M(w)).forEach(function(b) {
              return b(y, g, { path: E, hashId: a, parentSelectors: u });
            });
            var R = y.replace(/[A-Z]/g, function(b) {
              return "-".concat(b.toLowerCase());
            }), A = g;
            et[y] || typeof A != "number" || A === 0 || (A = "".concat(A, "px")), y === "animationName" && g != null && g._keyframe && (D(g), A = g.getName(a)), l += "".concat(R, ":").concat(A, ";");
          };
          var T, P = (T = r == null ? void 0 : r.value) !== null && T !== void 0 ? T : r;
          W(r) === "object" && r != null && r[$] && Array.isArray(P) ? P.forEach(function(y) {
            h(t, y);
          }) : h(t, P);
        } else {
          var F = !1, k = t.trim(), V = !1;
          (_ || I) && a ? k.startsWith("@") ? F = !0 : k = dt(k === "&" ? "" : t, a, i) : !_ || a || k !== "&" && k !== "" || (k = "", V = !0);
          var C = n(r, s, { root: V, injectHash: F, parentSelectors: [].concat(M(u), [k]) }), c = O(C, 2), v = c[0], x = c[1];
          H = z(z({}, H), x), l += "".concat(k).concat(v);
        }
      });
    }
  }), _ ? o && (l && (l = "@layer ".concat(o.name, " {").concat(l, "}")), o.dependencies && (H["@layer ".concat(o.name)] = o.dependencies.map(function(d) {
    return "@layer ".concat(d, ", ").concat(o.name, ";");
  }).join(`
`))) : l = "{".concat(l, "}"), [l, H];
};
function _t(n, m) {
  return nt("".concat(n.join("%")).concat(m));
}
function Et() {
  return null;
}
var gt = "style";
function Vt(n, m) {
  var s = n.token, e = n.path, _ = n.hashId, I = n.layer, u = n.nonce, a = n.clientOnly, o = n.order, E = o === void 0 ? 0 : o, i = K.useContext(ct), j = i.autoClear, L = i.mock, N = i.defaultCache, w = i.hashPriority, l = i.container, H = i.ssrInline, D = i.transformers, Q = i.linters, d = i.cache, f = i.layer, p = s._tokenKey, t = [p];
  f && t.push("layer"), t.push.apply(t, M(e));
  var r = U;
  process.env.NODE_ENV !== "production" && L !== void 0 && (r = L === "client");
  var T = pt(gt, t, function() {
    var C = t.join("|");
    if (lt(C)) {
      var c = ft(C), v = O(c, 2), x = v[0], h = v[1];
      if (x) return [x, p, h, {}, a, E];
    }
    var y = m(), g = St(y, { hashId: _, hashPriority: w, layer: f ? I : void 0, path: e.join("-"), transformers: D, linters: Q }), R = O(g, 2), A = R[0], b = R[1], S = q(A), tt = _t(t, S);
    return [S, p, tt, b, a, E];
  }, function(C, c) {
    var v = O(C, 3)[2];
    (c || j) && U && at(v, { mark: Z });
  }, function(C) {
    var c = O(C, 4), v = c[0];
    c[1];
    var x = c[2], h = c[3];
    if (r && v !== mt) {
      var y = { mark: Z, prepend: !f && "queue", attachTo: l, priority: E }, g = typeof u == "function" ? u() : u;
      g && (y.csp = { nonce: g });
      var R = [], A = [];
      Object.keys(h).forEach(function(S) {
        S.startsWith("@layer") ? R.push(S) : A.push(S);
      }), R.forEach(function(S) {
        Y(q(h[S]), "_layer-".concat(S), z(z({}, y), {}, { prepend: !0 }));
      });
      var b = Y(v, x, y);
      b[st] = d.instanceId, b.setAttribute(J, p), process.env.NODE_ENV !== "production" && b.setAttribute(ut, t.join("|")), A.forEach(function(S) {
        Y(q(h[S]), "_effect-".concat(S), y);
      });
    }
  }), P = O(T, 3), F = P[0], k = P[1], V = P[2];
  return function(C) {
    var c, v;
    return H && !r && N ? c = K.createElement("style", rt({}, (G(v = {}, J, k), G(v, Z, V), v), { dangerouslySetInnerHTML: { __html: F } })) : c = K.createElement(Et, null), K.createElement(K.Fragment, null, c, C);
  };
}
var Wt = function(n, m, s) {
  var e = O(n, 6), _ = e[0], I = e[1], u = e[2], a = e[3], o = e[4], E = e[5], i = (s || {}).plain;
  if (o) return null;
  var j = _, L = { "data-rc-order": "prependQueue", "data-rc-priority": "".concat(E) };
  return j = X(_, I, u, L, i), a && Object.keys(a).forEach(function(N) {
    if (!m[N]) {
      m[N] = !0;
      var w = q(a[N]), l = X(w, I, "_effect-".concat(N), L, i);
      N.startsWith("@layer") ? j = l + j : j += l;
    }
  }), [E, u, j];
};
export {
  gt as STYLE_PREFIX,
  Vt as default,
  Wt as extract,
  q as normalizeStyle,
  St as parseStyle,
  _t as uniqueHash
};
