// @secondcloset/types/src/fulfillment/appointment/appointment.ts
import {
  AppointmentType,
  AppointmentSubtype,
  DeliveryType,
  ParcelServiceLevel,
} from "./appointmentType";
import { AppointmentStatus } from "./appointmentStatus";
import { Shipment, ScServiceCode } from "../shipment";
import { Order } from "../order";
import { AppointmentSlot } from "../../storage";
import { BaseImage, BaseNote, ServiceArea } from "../../_common";

export interface Appointment {
  air_skip?: boolean;
  appointment_actions: string[];
  arrived_time: string | null;
  cancelled_at: string | null;
  completed_at: string | null;
  claim_number?: string;
  created_at: string;
  customer_segment: "business";
  date: string;
  delivery_service_level?: ScServiceCode;
  driver_id: string | null;
  end_time: string | null;
  estimated_time: number;
  facility?: {
    name: string;
    street: string;
  };
  failed_notes: string | null;
  failed_reason: string | null;
  field_ops_notes?: BaseNote[];
  formatted_timeslot: string | null;
  id: string;
  incomplete_at: string | null;
  issue_images?: BaseImage[];
  job_type: AppointmentType;
  job_subtype?: AppointmentSubtype;
  location: Location;
  movers: null;
  num_movers: null;
  number: string;
  parcel_service_level?: ParcelServiceLevel;
  range_of_days_assigned_dates?: {
    assigned_start_datetime: string;
    assigned_end_datetime: string;
  } | null;
  range_of_days_start_date?: string | null;
  range_of_days_end_date?: string | null;
  rescheduled: boolean;
  rescheduled_appointment_id: string | null;
  shipment: Partial<Shipment>;
  signature_image: any;
  signature_url: string | null;
  signed_at: string | null;
  sms: boolean;
  source: Partial<Order>;
  start_time: string | null;
  status: AppointmentStatus;
  timerange: AppointmentSlot;
  type: "fulfillment";
  updated_at: string;
  user_code: string;
  user_email: string;
  user_firstname: string;
  user_id: string;
  user_lastname: string;
  verification_images?: BaseImage[];
  delivery_type?: DeliveryType;
  shipping_labels?: LastMileShippingLabel[];
  signature_required?: boolean;
  estimated_arrival_time_window_start: string;
  estimated_arrival_time_window_end: string;
  package_weights: {
    weight_unit: "lb" | "kg";
    weight_value: number;
  }[] | null;
  delivery_notes: DeliveryNote[];
  master_tracking_number: string | null;
  tracking_numbers: string[] | null;
  package_ids: string[] | null;
  all_labels_received_at_destination: boolean;
  recommended_route_user_count: number | null;
}

export interface DeliveryNote {
  id: string;
  appointment_id: string;
  details: string | null;
  access: string | null;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

export interface Location {
  address: string;
  cell_phone_number: string | null;
  city: string;
  contact_name: string;
  coordinates: { lat: number; lng: number };
  country: string;
  created_at: string;
  customer_id: string;
  customer_phone_number: string;
  evening_phone_number: string | null;
  id: string;
  phone_number: string;
  postal_code: string;
  province: string;
  service_area: ServiceArea | null;
  updated_at: string;
}

interface LastMileShippingLabel {
  id: string;
  label_id?: string;
  tracking_number?: string;
  tracking_url?: string;
  label_download?: string;
}
