// @gobolt/genesis/dist/components/Message/styles.js
import o from "styled-components";
import { SIZE as n } from "../../constants/index.js";
const s = o.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  gap: ${({ theme: e }) => e.sizing.Size4}px;
  padding: ${({ theme: e }) => e.sizing.Size3}px;
  background-color: ${({ theme: e, state: t }) => e.colors.status[t].surface};
  width: ${({ theme: e, state: t, size: i }) => i === "standard" ? "320px" : "240px"};
  pointer-events: none;
  cursor: default;
  border-radius: ${({ theme: e }) => e.borderRadius.BorderRadiusSm}px;
`, d = o.div`
  display: flex;
  align-items: center;
  gap: ${({ theme: e }) => e.sizing.Size2}px;
  flex: 1;
  max-width: ${({ size: e, action: t }) => e !== "standard" && t ? "240px" : "320px"};

  .ant-typography {
    display: -webkit-box;
    -webkit-box-orient: vertical;
    overflow: hidden;
    font-family: ${({ theme: e }) => e.typography.fontFamily};
    line-height: ${({ theme: e, size: t }) => t === n.standard ? e.typography.lineHeight : "23px"};
    font-size: ${({ theme: e, size: t }) => t === "standard" ? e.typography.body2 : e.typography.body1};
  }
`, p = o.button`
  text-decoration: underline;
  border: none;
  background: none;
  cursor: pointer;
  pointer-events: all;
  color: ${({ theme: e }) => e.colors.interactive.link.active};

  &:hover {
    color: ${({ theme: e }) => e.colors.interactive.link.hover};
  }
`;
export {
  p as ActionButton,
  d as MessageContent,
  s as MessageRow
};
