// antd/es/date-picker/util.js
import * as n from "react";
import c from "../select/useIcons.js";
function h(e, r, a) {
  return a !== void 0 ? a : r === "year" && e.lang.yearPlaceholder ? e.lang.yearPlaceholder : r === "quarter" && e.lang.quarterPlaceholder ? e.lang.quarterPlaceholder : r === "month" && e.lang.monthPlaceholder ? e.lang.monthPlaceholder : r === "week" && e.lang.weekPlaceholder ? e.lang.weekPlaceholder : r === "time" && e.timePickerLocale.placeholder ? e.timePickerLocale.placeholder : e.lang.placeholder;
}
function d(e, r, a) {
  return a !== void 0 ? a : r === "year" && e.lang.yearPlaceholder ? e.lang.rangeYearPlaceholder : r === "quarter" && e.lang.quarterPlaceholder ? e.lang.rangeQuarterPlaceholder : r === "month" && e.lang.monthPlaceholder ? e.lang.rangeMonthPlaceholder : r === "week" && e.lang.weekPlaceholder ? e.lang.rangeWeekPlaceholder : r === "time" && e.timePickerLocale.placeholder ? e.timePickerLocale.rangePlaceholder : e.lang.rangePlaceholder;
}
function P(e, r) {
  const { allowClear: a = !0 } = e, { clearIcon: l, removeIcon: o } = c(Object.assign(Object.assign({}, e), { prefixCls: r, componentName: "DatePicker" }));
  return [n.useMemo(() => a === !1 ? !1 : Object.assign({ clearIcon: l }, a === !0 ? {} : a), [a, l]), o];
}
export {
  h as getPlaceholder,
  d as getRangePlaceholder,
  P as useIcons
};
