// rc-pagination/es/Pagination.js
import g from "../../@babel/runtime/helpers/esm/defineProperty.js";
import x from "../../@babel/runtime/helpers/esm/extends.js";
import pn from "../../@babel/runtime/helpers/esm/typeof.js";
import dn from "../../@babel/runtime/helpers/esm/objectSpread2.js";
import W from "../../@babel/runtime/helpers/esm/slicedToArray.js";
import N from "../../../_virtual/index.js";
import De from "../../rc-util/es/hooks/useMergedState.js";
import h from "../../rc-util/es/KeyCode.js";
import vn from "../../rc-util/es/pickAttrs.js";
import { warningOnce as gn } from "../../rc-util/es/warning.js";
import r, { useEffect as hn } from "react";
import fn from "./locale/zh_CN.js";
import En from "./Options.js";
import _ from "./Pager.js";
var Nn = function(n, s, o) {
  return o;
};
function Q() {
}
function Ke(n) {
  var s = Number(n);
  return typeof s == "number" && !Number.isNaN(s) && isFinite(s) && Math.floor(s) === s;
}
function y(n, s, o) {
  var D = n === void 0 ? s : n;
  return Math.floor((o - 1) / D) + 1;
}
var yn = function(n) {
  var s = n.prefixCls, o = s === void 0 ? "rc-pagination" : s, D = n.selectPrefixCls, Oe = D === void 0 ? "rc-select" : D, Te = n.className, Me = n.current, q = n.defaultCurrent, Ve = q === void 0 ? 1 : q, ee = n.total, c = ee === void 0 ? 0 : ee, Re = n.pageSize, ne = n.defaultPageSize, Ue = ne === void 0 ? 10 : ne, te = n.onChange, C = te === void 0 ? Q : te, He = n.hideOnSinglePage, X = n.align, ae = n.showPrevNextJumpers, Be = ae === void 0 || ae, Z = n.showQuickJumper, k = n.showLessItems, oe = n.showTitle, f = oe === void 0 || oe, re = n.onShowSizeChange, le = re === void 0 ? Q : re, ie = n.locale, d = ie === void 0 ? fn : ie, Ge = n.style, ce = n.totalBoundaryShowSizeChanger, Le = ce === void 0 ? 50 : ce, K = n.disabled, E = n.simple, ue = n.showTotal, me = n.showSizeChanger, Ye = me === void 0 ? c > Le : me, Xe = n.sizeChangerRender, Ze = n.pageSizeOptions, se = n.itemRender, w = se === void 0 ? Nn : se, pe = n.jumpPrevIcon, de = n.jumpNextIcon, $e = n.prevIcon, Ae = n.nextIcon, Fe = r.useRef(null), Je = De(10, { value: Re, defaultValue: Ue }), ve = W(Je, 2), m = ve[0], We = ve[1], Qe = De(1, { value: Me, defaultValue: Ve, postState: function(e) {
    return Math.max(1, Math.min(e, y(void 0, m, c)));
  } }), ge = W(Qe, 2), t = ge[0], he = ge[1], qe = r.useState(t), fe = W(qe, 2), b = fe[0], O = fe[1];
  hn(function() {
    O(t);
  }, [t]);
  var en = C !== Q, nn = "current" in n;
  process.env.NODE_ENV !== "production" && gn(!nn || en, "You provided a `current` prop to a Pagination component without an `onChange` handler. This will render a read-only component.");
  var Ee = Math.max(1, t - (k ? 3 : 5)), Ne = Math.min(y(void 0, m, c), t + (k ? 3 : 5));
  function T(e, a) {
    var l = e || r.createElement("button", { type: "button", "aria-label": a, className: "".concat(o, "-item-link") });
    return typeof e == "function" && (l = r.createElement(e, dn({}, n))), l;
  }
  function ye(e) {
    var a = e.target.value, l = y(void 0, m, c);
    return a === "" ? a : Number.isNaN(Number(a)) ? b : a >= l ? l : Number(a);
  }
  var tn = c > m && Z;
  function be(e) {
    var a = ye(e);
    switch (a !== b && O(a), e.keyCode) {
      case h.ENTER:
        p(a);
        break;
      case h.UP:
        p(a - 1);
        break;
      case h.DOWN:
        p(a + 1);
    }
  }
  function p(e) {
    if (function(z) {
      return Ke(z) && z !== t && Ke(c) && c > 0;
    }(e) && !K) {
      var a = y(void 0, m, c), l = e;
      return e > a ? l = a : e < 1 && (l = 1), l !== b && O(l), he(l), C == null || C(l, m), l;
    }
    return t;
  }
  var M = t > 1, V = t < y(void 0, m, c);
  function xe() {
    M && p(t - 1);
  }
  function Ce() {
    V && p(t + 1);
  }
  function ke() {
    p(Ee);
  }
  function we() {
    p(Ne);
  }
  function P(e, a) {
    if (e.key === "Enter" || e.charCode === h.ENTER || e.keyCode === h.ENTER) {
      for (var l = arguments.length, z = new Array(l > 2 ? l - 2 : 0), Y = 2; Y < l; Y++) z[Y - 2] = arguments[Y];
      a.apply(void 0, z);
    }
  }
  function R(e) {
    e.type !== "click" && e.keyCode !== h.ENTER || p(b);
  }
  var Pe = null, an = vn(n, { aria: !0, data: !0 }), on = ue && r.createElement("li", { className: "".concat(o, "-total-text") }, ue(c, [c === 0 ? 0 : (t - 1) * m + 1, t * m > c ? c : t * m])), je = null, i = y(void 0, m, c);
  if (He && c <= m) return null;
  var u = [], j = { rootPrefixCls: o, onClick: p, onKeyPress: P, showTitle: f, itemRender: w, page: -1 }, rn = t - 1 > 0 ? t - 1 : 0, ln = t + 1 < i ? t + 1 : i, U = Z && Z.goButton, cn = pn(E) === "object" ? E.readOnly : !E, H = U, Se = null;
  E && (U && (H = typeof U == "boolean" ? r.createElement("button", { type: "button", onClick: R, onKeyUp: R }, d.jump_to_confirm) : r.createElement("span", { onClick: R, onKeyUp: R }, U), H = r.createElement("li", { title: f ? "".concat(d.jump_to).concat(t, "/").concat(i) : null, className: "".concat(o, "-simple-pager") }, H)), Se = r.createElement("li", { title: f ? "".concat(t, "/").concat(i) : null, className: "".concat(o, "-simple-pager") }, cn ? b : r.createElement("input", { type: "text", "aria-label": d.jump_to, value: b, disabled: K, onKeyDown: function(e) {
    e.keyCode !== h.UP && e.keyCode !== h.DOWN || e.preventDefault();
  }, onKeyUp: be, onChange: be, onBlur: function(e) {
    p(ye(e));
  }, size: 3 }), r.createElement("span", { className: "".concat(o, "-slash") }, "/"), i));
  var v = k ? 1 : 2;
  if (i <= 3 + 2 * v) {
    i || u.push(r.createElement(_, x({}, j, { key: "noPager", page: 1, className: "".concat(o, "-item-disabled") })));
    for (var S = 1; S <= i; S += 1) u.push(r.createElement(_, x({}, j, { key: S, page: S, active: t === S })));
  } else {
    var un = k ? d.prev_3 : d.prev_5, mn = k ? d.next_3 : d.next_5, Ie = w(Ee, "jump-prev", T(pe, "prev page")), ze = w(Ne, "jump-next", T(de, "next page"));
    Be && (Pe = Ie ? r.createElement("li", { title: f ? un : null, key: "prev", onClick: ke, tabIndex: 0, onKeyDown: function(e) {
      P(e, ke);
    }, className: N("".concat(o, "-jump-prev"), g({}, "".concat(o, "-jump-prev-custom-icon"), !!pe)) }, Ie) : null, je = ze ? r.createElement("li", { title: f ? mn : null, key: "next", onClick: we, tabIndex: 0, onKeyDown: function(e) {
      P(e, we);
    }, className: N("".concat(o, "-jump-next"), g({}, "".concat(o, "-jump-next-custom-icon"), !!de)) }, ze) : null);
    var $ = Math.max(1, t - v), A = Math.min(t + v, i);
    t - 1 <= v && (A = 1 + 2 * v), i - t <= v && ($ = i - 2 * v);
    for (var I = $; I <= A; I += 1) u.push(r.createElement(_, x({}, j, { key: I, page: I, active: t === I })));
    if (t - 1 >= 2 * v && t !== 3 && (u[0] = r.cloneElement(u[0], { className: N("".concat(o, "-item-after-jump-prev"), u[0].props.className) }), u.unshift(Pe)), i - t >= 2 * v && t !== i - 2) {
      var _e = u[u.length - 1];
      u[u.length - 1] = r.cloneElement(_e, { className: N("".concat(o, "-item-before-jump-next"), _e.props.className) }), u.push(je);
    }
    $ !== 1 && u.unshift(r.createElement(_, x({}, j, { key: 1, page: 1 }))), A !== i && u.push(r.createElement(_, x({}, j, { key: i, page: i })));
  }
  var B = function(e) {
    var a = w(e, "prev", T($e, "prev page"));
    return r.isValidElement(a) ? r.cloneElement(a, { disabled: !M }) : a;
  }(rn);
  if (B) {
    var F = !M || !i;
    B = r.createElement("li", { title: f ? d.prev_page : null, onClick: xe, tabIndex: F ? null : 0, onKeyDown: function(e) {
      P(e, xe);
    }, className: N("".concat(o, "-prev"), g({}, "".concat(o, "-disabled"), F)), "aria-disabled": F }, B);
  }
  var G, J, L = function(e) {
    var a = w(e, "next", T(Ae, "next page"));
    return r.isValidElement(a) ? r.cloneElement(a, { disabled: !V }) : a;
  }(ln);
  L && (E ? (G = !V, J = M ? 0 : null) : J = (G = !V || !i) ? null : 0, L = r.createElement("li", { title: f ? d.next_page : null, onClick: Ce, tabIndex: J, onKeyDown: function(e) {
    P(e, Ce);
  }, className: N("".concat(o, "-next"), g({}, "".concat(o, "-disabled"), G)), "aria-disabled": G }, L));
  var sn = N(o, Te, g(g(g(g(g({}, "".concat(o, "-start"), X === "start"), "".concat(o, "-center"), X === "center"), "".concat(o, "-end"), X === "end"), "".concat(o, "-simple"), E), "".concat(o, "-disabled"), K));
  return r.createElement("ul", x({ className: sn, style: Ge, ref: Fe }, an), on, B, E ? Se : u, L, r.createElement(En, { locale: d, rootPrefixCls: o, disabled: K, selectPrefixCls: Oe, changeSize: function(e) {
    var a = y(e, m, c), l = t > a && a !== 0 ? a : t;
    We(e), O(l), le == null || le(t, e), he(l), C == null || C(l, e);
  }, pageSize: m, pageSizeOptions: Ze, quickGo: tn ? p : null, goButton: H, showSizeChanger: Ye, sizeChangerRender: Xe }));
};
process.env.NODE_ENV !== "production" && (yn.displayName = "Pagination");
export {
  yn as default
};
