// @secondcloset/types/src/fulfillment/product/product.ts
import { Stock } from "./stock";
import { BaseImage } from "../../_common";

export interface BaseProduct {
  id: string;
  scid: string;
  name: string;
  sku: string;
  upc: string;
  type: "base" | "kit";
  active: boolean;
  height: number;
  length: number;
  width: number;
  weight: number;
}

export interface Product extends Omit<BaseProduct, "type"> {
  organization_id: string;
  description: string;
  platform: string;
  fragile: boolean;
  dangerous: boolean;
  liquid: boolean;
  digital: boolean;
  tracks_lot_numbers: boolean;
  contains_battery: boolean;
  external_product_id: string | null;
  weight_unit: string;
  usd_value: number;
  value: number;
  cost: number;
  wholesale_price: number;
  retail_price: number;
  hs_tariff_code: string;
  customs_description: string | null;
  images: BaseImage[];
  packaging_criteria: string;
  product_type: BaseProduct["type"];
  stocks?: Stock[];
  deleted_at: string;
  created_at: string;
  updated_at: string;
  virtual_kits?: {
    id: string;
    organization_id: string;
    scid: string;
    sku: string;
    name: string;
    quantity_per_kit: number;
    deleted_at: string | null;
    created_at: string;
    updated_at: string;
    product_type: string;
  }[];
  required_parts?: { [productID: string]: number };
  packaging_levels?: PackagingLevel[];
  return_days?: number;
  insurance_required: boolean;
  insurance_value: number | null;
  has_errors: boolean;
  under_conversion: boolean;
  base_products?: Product[];
}

export interface PackagingLevel {
  id?: string;
  product_id?: string;
  product: BaseProduct & {
    upc: string;
    tracks_lot_numbers: boolean;
    tracks_expiry_date: boolean;
  };
  level: "single_item" | "inner_pack" | "master_pack" | "pallet";
  base_product_quantity: number;
  sku: string;
  name?: string;
  description: string;
  weight_value: string;
  weight_unit: "lb" | "kg" | "g" | "mg" | "oz";
  length_value: string;
  length_unit: "m" | "in" | "ft" | "yd" | "mi";
  height_value: string;
  height_unit: "m" | "in" | "ft" | "yd" | "mi";
  width_value: string;
  width_unit: "m" | "in" | "ft" | "yd" | "mi";
  created_at: string;
  updated_at: string;
}
