// @gobolt/genesis/dist/components/Tabs/styles.cjs
"use strict";Object.defineProperty(exports,Symbol.toStringTag,{value:"Module"});const a=require("styled-components"),o=require("../../node_modules/antd/es/tabs/index.cjs"),r=({colors:n,borderRadius:t,sizing:e},s="primary",b="active")=>`
  &.ant-tabs {
    background: #F4F4F4;

    .ant-tabs-nav {
      margin-bottom: 0;
    }
    
    /* Fix for tab navigation list to ensure no gaps */
    .ant-tabs-nav-list {
      overflow: hidden;
      border-radius: 0;
      display: flex;
      box-sizing: border-box;
    }
    
    .ant-tabs-tab {
      background: #EEEEEE;
      margin: 0; /* Important: no margin between tabs */
      padding: 8px 16px;
      border-radius: 0 !important;
      float: none; /* Prevent default float behavior */
      position: relative;
      
      /* Use single side borders to prevent double borders */
      border-left: 1px solid ${n.surface.border};

      margin-right: -2px;
      user-select: none;
      color: ${n.onsurface.copy};
      font-family: Inter;
      font-size: 14px;
      font-style: normal;
      font-weight: 500;
      line-height: 14px;
      display: flex;
      
      &:hover {
        background: ${n.interactive.utility.hover};
      }

      border-top: 0px;
      
      &:first-child {
        border-radius: 0 !important;
        border-left: none;
      }
      
      /* Only apply right border to the last tab */
      &:last-child, &.ant-tabs-tab-btn:last-child {
        border-right: 3px solid ${n.surface.border};
      }

      .ant-tabs-tab-btn {
        color: #3E3E3E;
        user-select: none;
        background: transparent;

        &:hover {
          background: ${n.interactive.utility.hover};
        }

        &:focus {
          background: transparent;
        }

        &:active {
          background: transparent;
        }
      }
    
      &.ant-tabs-tab-active {
        background: ${n.interactive.utility.active}; 
        user-select: none;
        
        .ant-tabs-tab-btn {
          color: #222222;
        }

        &:hover {
          background: ${n.interactive.utility.hover};
        }

        &:focus {
          background: ${n.interactive.utility.active};
        }

        &:active {
          background: ${n.interactive.utility.active};
        }
      }
    }
  }
`,i=a(o)`
  ${({theme:n,type:t,state:e})=>r(n,t,e)}
`;exports.Tabs=i,exports.getGenesisClass=r;
