// rc-util/es/React/render.js
import u from "../../../@babel/runtime/helpers/esm/regeneratorRuntime.js";
import p from "../../../@babel/runtime/helpers/esm/asyncToGenerator.js";
import l from "../../../@babel/runtime/helpers/esm/typeof.js";
import d from "../../../@babel/runtime/helpers/esm/objectSpread2.js";
import * as v from "react-dom";
var a, o = d({}, v), g = o.version, m = o.render, h = o.unmountComponentAtNode;
try {
  Number((g || "").split(".")[0]) >= 18 && (a = o.createRoot);
} catch {
}
function s(r) {
  var t = o.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
  t && l(t) === "object" && (t.usingClientEntryPoint = r);
}
var i = "__rc_react_root__";
function w(r, t) {
  a ? function(n, e) {
    s(!0);
    var c = e[i] || a(e);
    s(!1), c.render(n), e[i] = c;
  }(r, t) : function(n, e) {
    m == null || m(n, e);
  }(r, t);
}
function y(r) {
  return f.apply(this, arguments);
}
function f() {
  return (f = p(u().mark(function r(t) {
    return u().wrap(function(n) {
      for (; ; ) switch (n.prev = n.next) {
        case 0:
          return n.abrupt("return", Promise.resolve().then(function() {
            var e;
            (e = t[i]) === null || e === void 0 || e.unmount(), delete t[i];
          }));
        case 1:
        case "end":
          return n.stop();
      }
    }, r);
  }))).apply(this, arguments);
}
function E(r) {
  h(r);
}
function x(r) {
  return _.apply(this, arguments);
}
function _() {
  return (_ = p(u().mark(function r(t) {
    return u().wrap(function(n) {
      for (; ; ) switch (n.prev = n.next) {
        case 0:
          if (a === void 0) {
            n.next = 2;
            break;
          }
          return n.abrupt("return", y(t));
        case 2:
          E(t);
        case 3:
        case "end":
          return n.stop();
      }
    }, r);
  }))).apply(this, arguments);
}
export {
  w as render,
  x as unmount
};
