// dayjs/plugin/customParseFormat.js
import { __module as b } from "../../../_virtual/customParseFormat2.js";
var B;
b.exports;
function ot() {
  return B || (B = 1, b.exports = function() {
    var G = { LTS: "h:mm:ss A", LT: "h:mm A", L: "MM/DD/YYYY", LL: "MMMM D, YYYY", LLL: "MMMM D, YYYY h:mm A", LLLL: "dddd, MMMM D, YYYY h:mm A" }, I = /(\[[^[]*\])|([-_:/.,()\s]+)|(A|a|Q|YYYY|YY?|ww?|MM?M?M?|Do|DD?|hh?|HH?|mm?|ss?|S{1,3}|z|ZZ?)/g, q = /\d/, L = /\d\d/, i = /\d\d?/, y = /\d*[^-_:/,()\s\d]+/, m = {}, E = function(t) {
      return (t = +t) + (t > 68 ? 1900 : 2e3);
    }, o = function(t) {
      return function(r) {
        this[t] = +r;
      };
    }, Q = [/[+-]\d\d:?(\d\d)?|Z/, function(t) {
      (this.zone || (this.zone = {})).offset = function(r) {
        if (!r || r === "Z") return 0;
        var e = r.match(/([+-]|\d\d)/g), n = 60 * e[1] + (+e[2] || 0);
        return n === 0 ? 0 : e[0] === "+" ? -n : n;
      }(t);
    }], _ = function(t) {
      var r = m[t];
      return r && (r.indexOf ? r : r.s.concat(r.f));
    }, X = function(t, r) {
      var e, n = m.meridiem;
      if (n) {
        for (var h = 1; h <= 24; h += 1) if (t.indexOf(n(h, 0, r)) > -1) {
          e = h > 12;
          break;
        }
      } else e = t === (r ? "pm" : "PM");
      return e;
    }, J = { A: [y, function(t) {
      this.afternoon = X(t, !1);
    }], a: [y, function(t) {
      this.afternoon = X(t, !0);
    }], Q: [q, function(t) {
      this.month = 3 * (t - 1) + 1;
    }], S: [q, function(t) {
      this.milliseconds = 100 * +t;
    }], SS: [L, function(t) {
      this.milliseconds = 10 * +t;
    }], SSS: [/\d{3}/, function(t) {
      this.milliseconds = +t;
    }], s: [i, o("seconds")], ss: [i, o("seconds")], m: [i, o("minutes")], mm: [i, o("minutes")], H: [i, o("hours")], h: [i, o("hours")], HH: [i, o("hours")], hh: [i, o("hours")], D: [i, o("day")], DD: [L, o("day")], Do: [y, function(t) {
      var r = m.ordinal, e = t.match(/\d+/);
      if (this.day = e[0], r) for (var n = 1; n <= 31; n += 1) r(n).replace(/\[|\]/g, "") === t && (this.day = n);
    }], w: [i, o("week")], ww: [L, o("week")], M: [i, o("month")], MM: [L, o("month")], MMM: [y, function(t) {
      var r = _("months"), e = (_("monthsShort") || r.map(function(n) {
        return n.slice(0, 3);
      })).indexOf(t) + 1;
      if (e < 1) throw new Error();
      this.month = e % 12 || e;
    }], MMMM: [y, function(t) {
      var r = _("months").indexOf(t) + 1;
      if (r < 1) throw new Error();
      this.month = r % 12 || r;
    }], Y: [/[+-]?\d+/, o("year")], YY: [L, function(t) {
      this.year = E(t);
    }], YYYY: [/\d{4}/, o("year")], Z: Q, ZZ: Q };
    function K(t) {
      var r, e;
      r = t, e = m && m.formats;
      for (var n = (t = r.replace(/(\[[^\]]+])|(LTS?|l{1,4}|L{1,4})/g, function(p, l, f) {
        var s = f && f.toUpperCase();
        return l || e[f] || G[f] || e[s].replace(/(\[[^\]]+])|(MMMM|MM|DD|dddd)/g, function(M, v, D) {
          return v || D.slice(1);
        });
      })).match(I), h = n.length, u = 0; u < h; u += 1) {
        var x = n[u], Y = J[x], c = Y && Y[0], d = Y && Y[1];
        n[u] = d ? { regex: c, parser: d } : x.replace(/^\[|\]$/g, "");
      }
      return function(p) {
        for (var l = {}, f = 0, s = 0; f < h; f += 1) {
          var M = n[f];
          if (typeof M == "string") s += M.length;
          else {
            var v = M.regex, D = M.parser, $ = p.slice(s), g = v.exec($)[0];
            D.call(l, g), p = p.replace(g, "");
          }
        }
        return function(w) {
          var S = w.afternoon;
          if (S !== void 0) {
            var a = w.hours;
            S ? a < 12 && (w.hours += 12) : a === 12 && (w.hours = 0), delete w.afternoon;
          }
        }(l), l;
      };
    }
    return function(t, r, e) {
      e.p.customParseFormat = !0, t && t.parseTwoDigitYear && (E = t.parseTwoDigitYear);
      var n = r.prototype, h = n.parse;
      n.parse = function(u) {
        var x = u.date, Y = u.utc, c = u.args;
        this.$u = Y;
        var d = c[1];
        if (typeof d == "string") {
          var p = c[2] === !0, l = c[3] === !0, f = p || l, s = c[2];
          l && (s = c[2]), m = this.$locale(), !p && s && (m = e.Ls[s]), this.$d = function($, g, w, S) {
            try {
              if (["x", "X"].indexOf(g) > -1) return new Date((g === "X" ? 1e3 : 1) * $);
              var a = K(g)($), A = a.year, T = a.month, N = a.day, R = a.hours, W = a.minutes, tt = a.seconds, rt = a.milliseconds, V = a.zone, j = a.week, H = /* @__PURE__ */ new Date(), O = N || (A || T ? 1 : H.getDate()), z = A || H.getFullYear(), Z = 0;
              A && !T || (Z = T > 0 ? T - 1 : H.getMonth());
              var k, C = R || 0, F = W || 0, P = tt || 0, U = rt || 0;
              return V ? new Date(Date.UTC(z, Z, O, C, F, P, U + 60 * V.offset * 1e3)) : w ? new Date(Date.UTC(z, Z, O, C, F, P, U)) : (k = new Date(z, Z, O, C, F, P, U), j && (k = S(k).week(j).toDate()), k);
            } catch {
              return /* @__PURE__ */ new Date("");
            }
          }(x, d, Y, e), this.init(), s && s !== !0 && (this.$L = this.locale(s).$L), f && x != this.format(d) && (this.$d = /* @__PURE__ */ new Date("")), m = {};
        } else if (d instanceof Array) for (var M = d.length, v = 1; v <= M; v += 1) {
          c[1] = d[v - 1];
          var D = e.apply(this, c);
          if (D.isValid()) {
            this.$d = D.$d, this.$L = D.$L, this.init();
            break;
          }
          v === M && (this.$d = /* @__PURE__ */ new Date(""));
        }
        else h.call(this, u);
      };
    };
  }()), b.exports;
}
export {
  ot as __require
};
