// rc-motion/es/util/motion.js
import p from "../../../@babel/runtime/helpers/esm/typeof.js";
import c from "../../../rc-util/es/Dom/canUseDom.js";
function r(n, t) {
  var i = {};
  return i[n.toLowerCase()] = t.toLowerCase(), i["Webkit".concat(n)] = "webkit".concat(t), i["Moz".concat(n)] = "moz".concat(t), i["ms".concat(n)] = "MS".concat(t), i["O".concat(n)] = "o".concat(t.toLowerCase()), i;
}
function v(n, t) {
  var i = { animationend: r("Animation", "AnimationEnd"), transitionend: r("Transition", "TransitionEnd") };
  return n && ("AnimationEvent" in t || delete i.animationend.animation, "TransitionEvent" in t || delete i.transitionend.transition), i;
}
var l = v(c(), typeof window < "u" ? window : {}), s = {};
if (c()) {
  var w = document.createElement("div");
  s = w.style;
}
var a = {};
function m(n) {
  if (a[n]) return a[n];
  var t = l[n];
  if (t) for (var i = Object.keys(t), u = i.length, o = 0; o < u; o += 1) {
    var e = i[o];
    if (Object.prototype.hasOwnProperty.call(t, e) && e in s) return a[n] = t[e], a[n];
  }
  return "";
}
var d = m("animationend"), f = m("transitionend"), g = !(!d || !f), y = d || "animationend", T = f || "transitionend";
function k(n, t) {
  return n ? p(n) === "object" ? n[t.replace(/-\w/g, function(i) {
    return i[1].toUpperCase();
  })] : "".concat(n, "-").concat(t) : null;
}
export {
  y as animationEndName,
  k as getTransitionName,
  m as getVendorPrefixedEventName,
  v as getVendorPrefixes,
  g as supportTransition,
  T as transitionEndName
};
