// @gobolt/genesis/dist/components/Notification/NotificationBadge/styles.cjs
"use strict";Object.defineProperty(exports,Symbol.toStringTag,{value:"Module"});const o=require("styled-components"),t=o.div`
  position: relative;
  display: flex;
  position: absolute;
  justify-content: center;
  align-items: center;
  top: -16px;
  right: -16px;
  box-shadow: -2px 2px white;
  width: 20px;
  height: 20px;
  border: 2px solid ${({color:e})=>e};
  border-radius: 50%;
  box-sizing: border-box;
  background-color: white;

  &::before {
    content: "";
    position: absolute;
    top: 0px;
    left: 0px;
    right: 0px;
    bottom: 0px;
    border: 2px solid white;
    width: 12px;
    height: 12px;
    border-radius: 50%;
  }
`,i=o.div`
  display: flex;
  position: relative;
`,r=o.div`
  display: flex;
  justify-content: center;
  align-items: center;
  position: relative;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  border: 1px solid ${({theme:e})=>e.colors.surface.border};
  background-color: ${({theme:e})=>e.colors.surface.default};
`;exports.Badge=i,exports.BadgeState=t,exports.BadgeWrapper=r;
