// @ant-design/cssinjs-utils/es/_util/hooks/useUniqueMemo.js
import c from "../../../../../@babel/runtime/helpers/esm/typeof.js";
import r from "../../../../../@babel/runtime/helpers/esm/classCallCheck.js";
import u from "../../../../../@babel/runtime/helpers/esm/createClass.js";
import i from "../../../../../@babel/runtime/helpers/esm/defineProperty.js";
import h from "react";
var o = new (function() {
  function a() {
    r(this, a), i(this, "map", /* @__PURE__ */ new Map()), i(this, "objectIDMap", /* @__PURE__ */ new WeakMap()), i(this, "nextID", 0), i(this, "lastAccessBeat", /* @__PURE__ */ new Map()), i(this, "accessBeat", 0);
  }
  return u(a, [{ key: "set", value: function(t, s) {
    this.clear();
    var e = this.getCompositeKey(t);
    this.map.set(e, s), this.lastAccessBeat.set(e, Date.now());
  } }, { key: "get", value: function(t) {
    var s = this.getCompositeKey(t), e = this.map.get(s);
    return this.lastAccessBeat.set(s, Date.now()), this.accessBeat += 1, e;
  } }, { key: "getCompositeKey", value: function(t) {
    var s = this;
    return t.map(function(e) {
      return e && c(e) === "object" ? "obj_".concat(s.getObjectID(e)) : "".concat(c(e), "_").concat(e);
    }).join("|");
  } }, { key: "getObjectID", value: function(t) {
    if (this.objectIDMap.has(t)) return this.objectIDMap.get(t);
    var s = this.nextID;
    return this.objectIDMap.set(t, s), this.nextID += 1, s;
  } }, { key: "clear", value: function() {
    var t = this;
    if (this.accessBeat > 1e4) {
      var s = Date.now();
      this.lastAccessBeat.forEach(function(e, n) {
        s - e > 6e5 && (t.map.delete(n), t.lastAccessBeat.delete(n));
      }), this.accessBeat = 0;
    }
  } }]), a;
}())();
function D(a, t) {
  return h.useMemo(function() {
    var s = o.get(t);
    if (s) return s;
    var e = a();
    return o.set(t, e), e;
  }, t);
}
export {
  D as default
};
