// rc-overflow/es/Overflow.js
import x from "../../@babel/runtime/helpers/esm/extends.js";
import I from "../../@babel/runtime/helpers/esm/objectSpread2.js";
import m from "../../@babel/runtime/helpers/esm/slicedToArray.js";
import Be from "../../@babel/runtime/helpers/esm/objectWithoutProperties.js";
import * as a from "react";
import { useState as W, useMemo as ue, useCallback as fe } from "react";
import Le from "../../../_virtual/index.js";
import qe from "../../rc-resize-observer/es/index.js";
import He from "../../rc-util/es/hooks/useLayoutEffect.js";
import _ from "./Item.js";
import g, { useBatcher as Je } from "./hooks/useEffectState.js";
import Qe from "./RawItem.js";
import { OverflowContext as ce } from "./context.js";
var Ue = ["prefixCls", "data", "renderItem", "renderRawItem", "itemKey", "itemWidth", "ssr", "style", "className", "maxCount", "renderRest", "renderRawRest", "suffix", "component", "itemComponent", "onVisibleChange"], de = "responsive", ve = "invalidate";
function Ye(r) {
  return "+ ".concat(r.length, " ...");
}
function Ze(r, pe) {
  var D = r.prefixCls, P = D === void 0 ? "rc-overflow" : D, T = r.data, o = T === void 0 ? [] : T, F = r.renderItem, G = r.renderRawItem, d = r.itemKey, X = r.itemWidth, j = X === void 0 ? 10 : X, ye = r.ssr, ge = r.style, he = r.className, s = r.maxCount, Ee = r.renderRest, B = r.renderRawRest, S = r.suffix, L = r.component, Re = L === void 0 ? "div" : L, be = r.itemComponent, q = r.onVisibleChange, Ce = Be(r, Ue), M = ye === "full", v = Je(), we = g(v, null), H = m(we, 2), h = H[0], xe = H[1], u = h || 0, Ie = g(v, /* @__PURE__ */ new Map()), J = m(Ie, 2), Q = J[0], Ne = J[1], Se = g(v, 0), U = m(Se, 2), Me = U[0], ke = U[1], ze = g(v, 0), Y = m(ze, 2), E = Y[0], Ae = Y[1], Oe = g(v, 0), Z = m(Oe, 2), R = Z[0], Ke = Z[1], Ve = W(null), $ = m(Ve, 2), k = $[0], ee = $[1], We = W(null), te = m(We, 2), b = te[0], _e = te[1], c = a.useMemo(function() {
    return b === null && M ? Number.MAX_SAFE_INTEGER : b || 0;
  }, [b, h]), De = W(!1), re = m(De, 2), Pe = re[0], Te = re[1], z = "".concat(P, "-item"), ne = Math.max(Me, E), A = s === de, l = o.length && A, oe = s === ve, Fe = l || typeof s == "number" && o.length > s, f = ue(function() {
    var e = o;
    return l ? e = h === null && M ? o : o.slice(0, Math.min(o.length, u / j)) : typeof s == "number" && (e = o.slice(0, s)), e;
  }, [o, j, h, s, l]), O = ue(function() {
    return l ? o.slice(c + 1) : o.slice(f.length);
  }, [o, f, l, c]), C = fe(function(e, t) {
    var n;
    return typeof d == "function" ? d(e) : (n = d && (e == null ? void 0 : e[d])) !== null && n !== void 0 ? n : t;
  }, [d]), Ge = fe(F || function(e) {
    return e;
  }, [F]);
  function w(e, t, n) {
    (b !== e || t !== void 0 && t !== k) && (_e(e), n || (Te(e < o.length - 1), q == null || q(e)), t !== void 0 && ee(t));
  }
  function ie(e, t) {
    Ne(function(n) {
      var i = new Map(n);
      return t === null ? i.delete(e) : i.set(e, t), i;
    });
  }
  function K(e) {
    return Q.get(C(f[e], e));
  }
  He(function() {
    if (u && typeof ne == "number" && f) {
      var e = R, t = f.length, n = t - 1;
      if (!t) return void w(0, null);
      for (var i = 0; i < t; i += 1) {
        var y = K(i);
        if (M && (y = y || 0), y === void 0) {
          w(i - 1, void 0, !0);
          break;
        }
        if (e += y, n === 0 && e <= u || i === n - 1 && e + K(n) <= u) {
          w(n, null);
          break;
        }
        if (e + ne > u) {
          w(i - 1, e - y - R + E);
          break;
        }
      }
      S && K(0) + R > u && ee(null);
    }
  }, [u, Q, E, R, C, f]);
  var le = Pe && !!O.length, ae = {};
  k !== null && l && (ae = { position: "absolute", left: k, top: 0 });
  var p = { prefixCls: z, responsive: l, component: be, invalidate: oe }, Xe = G ? function(e, t) {
    var n = C(e, t);
    return a.createElement(ce.Provider, { key: n, value: I(I({}, p), {}, { order: t, item: e, itemKey: n, registerSize: ie, display: t <= c }) }, G(e, t));
  } : function(e, t) {
    var n = C(e, t);
    return a.createElement(_, x({}, p, { order: t, key: n, item: e, renderItem: Ge, itemKey: n, registerSize: ie, display: t <= c }));
  }, me = { order: le ? c : Number.MAX_SAFE_INTEGER, className: "".concat(z, "-rest"), registerSize: function(e, t) {
    Ae(t), ke(E);
  }, display: le }, V = Ee || Ye, je = B ? a.createElement(ce.Provider, { value: I(I({}, p), me) }, B(O)) : a.createElement(_, x({}, p, me), typeof V == "function" ? V(O) : V), se = a.createElement(Re, x({ className: Le(!oe && P, he), style: ge, ref: pe }, Ce), f.map(Xe), Fe ? je : null, S && a.createElement(_, x({}, p, { responsive: A, responsiveDisabled: !l, order: c, className: "".concat(z, "-suffix"), registerSize: function(e, t) {
    Ke(t);
  }, display: !0, style: ae }), S));
  return A ? a.createElement(qe, { onResize: function(e, t) {
    xe(t.clientWidth);
  }, disabled: !l }, se) : se;
}
var N = a.forwardRef(Ze);
N.displayName = "Overflow", N.Item = Qe, N.RESPONSIVE = de, N.INVALIDATE = ve;
export {
  ce as OverflowContext,
  N as default
};
