// @secondcloset/types/src/warehouse/carrierTransferPackage.ts
import {
  ExternalCarrierShippingLabel,
  Package,
  Shipment,
} from "../fulfillment";
import { Organization } from "../_common";
import { CarrierTransferPallet } from "./carrierTransferPallet";

export type CarrierTransferPackageStatus = "at_facility" | "with_carrier";

export interface CarrierTransferPackage {
  id: string;
  status: CarrierTransferPackageStatus;
  carrier_name?: string;
  tracking_number?: string;
  carrier_transfer_pallet_id?: string;
  carrier_transfer_pallet?: CarrierTransferPallet;
  handover_by?: string;
  handover_at?: string;
  created_by: string;
  created_at: string;
  destination_country?: string;
}

interface PackageCarrier {
  id: string;
  carrier_code?: string;
  name?: string;
  external_platform?: string;
  external_id?: string;
  enabled_service_codes?: string[];
  created_at: string;
  updated_at: string;
  deleted_at?: string;
  organization_id?: string;
  disabled?: string;
}

interface ShippingLabel extends Partial<ExternalCarrierShippingLabel> {
  logistics_external_carrier_shipment_id?: string;
  external_platform_shipment_id?: string;
  external_platform_service_code?: string;
  external_platform_void_label_error?: string;
  created_at: string;
  updated_at: string;
}
interface PackageShipment extends Partial<Shipment<"V1">> {
  user_id: string;
  shipping_method_id?: string;
  customer_scheduling_token?: string;
  customer_scheduling_token_expired?: string;
  latest_booking_email_sent_at?: string;
  logistics_carrier_id?: string;
  facility_id?: string;
}
export interface PackageInfo extends Partial<Package> {
  tracking_number?: string;
  package_id_type: string;
}

export interface PackageDetailsInfo extends PackageInfo {
  shipment?: PackageShipment;
  organization: Organization;
  carrier?: PackageCarrier;
  shipping_label?: ShippingLabel;
}
