// @secondcloset/types/src/warehouse/organization.ts
import { FacilityCode } from "../_common";
import { LotCode } from "./lotCode";
import { Product } from "./product";
import { ShipmentProductsTracking } from "./shipmentProductsTracking";
import { Supply } from "./supply";

export interface Organization {
  id: string;
  name: string;
  products?: Product[];
  supplies?: Supply[];
  shipment_products_trackings?: ShipmentProductsTracking[];
  lot_codes?: LotCode[];
}

export enum OrganizationConfigKey {
  ENABLE_SCAN_ITEMS = "ENABLE_SCAN_ITEMS",
  OPENED_GOOD_CONDITION_SHIP_TO_ADDRESS = "OPENED_GOOD_CONDITION_SHIP_TO_ADDRESS",
  SHIP_TO_ME = "SHIP_TO_ME",
  OPENED_GOOD_CONDITION_INSTRUCTION = "OPENED_GOOD_CONDITION_INSTRUCTION",
  DAMAGED_CONDITION_INSTRUCTION = "DAMAGED_CONDITION_INSTRUCTION",
  GOOD_CONDITION_DAMAGED_PACKAGING_INSTRUCTION = "GOOD_CONDITION_DAMAGED_PACKAGING_INSTRUCTION",
  RTS_INSTRUCTION = "RTS_INSTRUCTION",
  ENABLE_SEND_EMAIL_TO_CUSTOMERS = "ENABLE_SEND_EMAIL_TO_CUSTOMERS",
  DEFAULT_RATE_FETCHING_V1 = "DEFAULT_RATE_FETCHING_V1",
  DEFAULT_REQUIRE_SIGNATURE = "DEFAULT_REQUIRE_SIGNATURE",
}
export interface OrganizationConfig {
  id: string;
  key: OrganizationConfigKey;
  value: string;
  organization_id: string;
  organization?: Organization;
  created_at: string;
  updated_at: string;
  metadata?: Record<string, any>;
}

export interface FacilityOrganization {
  id: string;
  organization_id: string;
  facility: FacilityCode;
  organization?: Organization;
}
