// @secondcloset/types/src/warehouse/picksheet.ts
import { FacilityCode } from "../_common";
import { Organization } from "./organization";

export type PicksheetOrganization = {
  name: string;
  id: string;
};

export type PicksheetFailedReasons = {
  order_number?: string;
  external_order_number?: string;
  failed_reason: string;
};

export type PicksheetStats = {
  organization?: PicksheetOrganization;
  ready_to_fulfill_count: number;
  processing_items_count: number;
  is_pushed_shipments_enabled: boolean;
};

export enum PicksheetStatus {
  IN_PROGRESS = "IN_PROGRESS",
  COMPLETED = "COMPLETED",
}

export enum InflationStatus {
  INFLATING = "INFLATING",
  INFLATED = "INFLATED",
  FAILED = "FAILED",
}

export enum PicksheetRequestType {
  DISCRETE = "DISCRETE",
  WAVE = "WAVE",
}

export enum PicksheetType {
  PARCEL = "PARCEL",
  LAST_MILE = "LAST_MILE",
}

export type Picksheet = {
  id: string;
  request_id: string;
  external_order_id: string;
  external_order_number?: string;
  inflation_status?: InflationStatus;
  order_id?: string;
  shipment_id?: string;
  shipment_number?: string;
  failed_reason?: string;
  created_at: string;
  updated_at: string;
  wavesheet_id?: string;
  request?: PicksheetRequest;
  picksheet_items?: PicksheetItem[];
  type: PicksheetType;
  organization?: Organization;
  organization_id: string;
  facility?: FacilityCode;
};

export type PicksheetItem = {
  id: string;
  picksheet_id: string;
  picksheet: Picksheet;
  sku: string;
  upc: string;
  name: string;
  quantity: number;
  picked_quantity: number;
  manual_item_code?: string;
  manual_item_customer_name?: string;
};

export type PicksheetRequest = {
  id: string;
  organization_id: string;
  organization?: Organization;
  facility: FacilityCode;
  status: PicksheetStatus;
  number_of_picksheets: number;
  requested_by: string;
  created_at: string;
  updated_at: string;
  type: PicksheetRequestType;
  picksheets?: Picksheet[];
};
