// @secondcloset/types/src/warehouse/productQuantum.ts
import { Product } from "./product";

enum ProductQuantumTypeEnum {
  EACH = "EACH",
  CASE = "CASE",
  INNER_PACK = "INNER_PACK",
  MASTER_PACK = "MASTER_PACK",
}

enum ContainsTypeEnum {
  EACH = "EACH",
  INNER_PACK = "INNER_PACK",
}

enum WeightUnitEnum {
  LB = "LB",
  OZ = "OZ",
  G = "G",
  KG = "KG",
}

enum DimensionsUnitEnum {
  IN = "IN",
  FT = "FT",
  CM = "CM",
  M = "M",
}

export interface ProductQuantum {
  id: string;
  type: ProductQuantumTypeEnum;
  quantity: number;
  sku: string;
  upc?: string;
  contains_type: ContainsTypeEnum | null;
  weight: number;
  height: number;
  length: number;
  width: number;
  weight_unit: WeightUnitEnum;
  dimensions_unit: DimensionsUnitEnum;
  product_id: string;
  product: Product;
}

export interface ProductQuantumUpdateBody {
  type: ProductQuantumTypeEnum;
  contains_type: ContainsTypeEnum;
  quantity: number;
  sku: string;
  upc: string;
  length: number;
  height: number;
  weight: number;
  width: number;
  dimensions_unit: DimensionsUnitEnum;
  weight_unit: WeightUnitEnum;
  product_id: string;
}

export interface CreateProductQuantumBody {
  type: ProductQuantumTypeEnum;
  upc: string;
  sku: string;
  product_id: string;
  contains_type: ProductQuantumTypeEnum;
  quantity: number;
  weight: number;
  weight_unit: WeightUnitEnum;
  height: number;
  width: number;
  length: number;
  dimensions_unit: DimensionsUnitEnum;
}
