// @secondcloset/types/src/warehouse/specialProject.ts
import { Organization } from "./organization";
import { FacilityCode } from "../_common";
import { Supply } from ".";

export enum SpecialProjectType {
  RECEIVING_SUPPORT = "RECEIVING_SUPPORT",
  INSPECTIONS = "INSPECTIONS",
  KITTING = "KITTING",
  PRODUCT_ADJUSTMENT = "PRODUCT_ADJUSTMENT",
  LABELLING = "LABELLING",
  INVENTORY_COUNT = "INVENTORY_COUNT",
  OTHER = "OTHER",
}

export interface SpecialProjectSupply {
  id: string;
  quantity: number;
  created_at: string;
  special_project_id: string;
  supply_id: string;
  supply: Supply;
}

export interface SpecialProject {
  id: string;
  organization_id: string;
  organization?: Organization;
  user_names: string;
  start_time: string;
  end_time: string;
  type: SpecialProjectType;
  facility: FacilityCode;
  details?: string;
  reference_number?: string;
  created_at: string;
  updated_at: string;
  special_project_supplies?: SpecialProjectSupply[];
}

export interface SpecialProjectLog {
  id: string;
  user_email: string;
  before?: string;
  after: string;
  special_project_id: string;
  created_at: string;
}
