// @secondcloset/types/src/_common/facility.ts
import { Version } from ".";
import { Common } from "..";
import { ServiceArea } from "./serviceArea";

export type FacilityCode =
  | "yyz1"
  | "yyz2"
  | "yyz3"
  | "yyz4"
  | "yyz5"
  | "yvr1"
  | "yvr2"
  | "yow1"
  | "yul1"
  | "yul2"
  | "lax1"
  | "lax2"
  | "nyc1"
  | "mia1"
  | "hou1"
  | "yyc1"
  | "atl1"
  | "atl2"
  | "unknown";

export type DeprecatedFacilityCode =
  | "yyz1"
  | "yyz2"
  | "yyz3"
  | "yyz4"
  | "yvr1"
  | "atl2";

export type FacilityCity =
  | "Toronto"
  | "Mississauga"
  | "Markham"
  | "Vancouver"
  | "Ottawa"
  | "Montreal"
  | "Buena Park"
  | "Santa Fe Springs"
  | "Carlstadt"
  | "Medley"
  | "Houston"
  | "Tucker"
  | "Newnan"
  | "Calgary";

export type Facility<V = Version> = V extends "V1"
  ? {
      id: number;
      name: FacilityCode;
      phone: string;
      email: string;
      address: {
        table: {
          street: string;
          city: FacilityCity;
          province: string;
          country: string;
          postal_code: string;
        };
      };
      service_area: ServiceArea;
    }
  : {
      id: string;
      name: FacilityCode;
      enabled: boolean;
      service_area: ServiceArea;
      address: {
        street: string;
        city: string;
        province: string;
        country: string;
        postal_code: string;
      };
      created_at: string;
      updated_at: string;
    };
