// @secondcloset/types/src/warehouse/return.ts
import { FacilityCode } from "../_common";
import { Organization } from "./organization";
import { RmaContainer } from "./rmaContainer";

export enum ReturnItemCondition {
  RESELLABLE = "resellable",
  UNSELLABLE = "unsellable",
  OTHER = "other",
}

export enum ReturnType {
  RTS = "RTS",
  REGULAR = "REGULAR",
}

export enum ReturnStatus {
  READY_TO_PROCESS = "READY_TO_PROCESS",
  NEEDS_REVIEW = "NEEDS_REVIEW",
  ARRIVED = "ARRIVED",
  PROCESSING = "PROCESSING",
  REQUESTED = "REQUESTED",
  ON_HOLD = "ON_HOLD",
  COMPLETED = "COMPLETED",
  CANCELLED = "CANCELLED",
  TO_BE_RESHIPPED = "TO_BE_RESHIPPED",
}

export interface ReturnPhoto {
  id: string;
  return_id?: string;
  photo_link: string;
}

export interface ReturnItem {
  id: string;
  name: string;
  external_product_id: string;
  sku?: string;
  upc?: string;
  condition?: string;
  condition_notes?: string;
  reason?: string;
  reason_notes?: string;
  rma_container_id?: string;
  rma_container?: RmaContainer;
}

export interface ReturnAddress {
  name: string;
  phone_number: string;
  address_1: string;
  address_2?: string;
  city: string;
  province: string;
  postal_code: string;
  country: string;
}

export interface Return {
  id: string;
  rma?: string;
  tracking_number?: string;
  tracking_url?: string;
  label_url?: string;
  carrier_logo_url?: string;
  external_order_number: string;
  reship_external_order_number?: string;
  status: ReturnStatus;
  type: ReturnType;
  shipping_from?: ReturnAddress;
  carrier?: string;
  organization: Organization;
  photos?: ReturnPhoto[];
  return_items?: ReturnItem[];
  reason?: string;
  reason_notes?: string;
  external_notes?: string;
  internal_notes?: string;
  order_date?: string;
  created_at: string;
  updated_at: string;
  processed_at_facility?: FacilityCode;
}
