// @gobolt/genesis/dist/components/Badge/styles.js
import d from "styled-components";
const t = (o, n, a, e) => `
  display: flex;
  gap: 4px;
  justify-content: center;
  align-items: center;
  position: relative;
  padding-top: ${o.components.badge["ver-padding"]}px;
  padding-bottom: ${o.components.badge["ver-padding"]}px;
  padding-left: ${o.components.badge["left-padding"]}px;
  padding-right: ${o.components.badge["right-padding"]}px;
  border-radius: ${o.components.badge.radius}px;
  border: 1px solid ${((r, s) => s === "hollow" ? r.colors.inputs.surface.border : r.colors.status[s].surface)(o, n)};
  background-color: ${o.colors.status[n].surface};

  &:not(:disabled):hover {
    background-color: ${o.colors.status[n]["surface-hover"]};
  }

  ${e ? `
    background-color: ${o.colors.status.disabled.surface} !important;
    border-color: ${o.colors.status.disabled} !important;
    border: none;
    cursor: not-allowed;

    &:hover {
      cursor: not-allowed;
    }
  ` : ""}
`, l = d.div`
  ${({ theme: o, state: n, label: a, disabled: e }) => t(o, n, 0, e)}
`;
export {
  l as Badge,
  t as getGenesisClass
};
