// rc-menu/es/hooks/useAccessibility.js
import i from "../../../@babel/runtime/helpers/esm/defineProperty.js";
import { getFocusNodeList as ue } from "../../../rc-util/es/Dom/focus.js";
import m from "../../../rc-util/es/KeyCode.js";
import O from "../../../rc-util/es/raf.js";
import * as P from "react";
import { getMenuId as le } from "../context/IdContext.js";
var U = m.LEFT, W = m.RIGHT, z = m.UP, w = m.DOWN, k = m.ENTER, I = m.ESC, p = m.HOME, S = m.END, C = [z, w, U, W];
function te(u, a) {
  return ue(u, !0).filter(function(o) {
    return a.has(o);
  });
}
function ee(u, a, o) {
  var c = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : 1;
  if (!u) return null;
  var s = te(u, a), l = s.length, e = s.findIndex(function(x) {
    return o === x;
  });
  return c < 0 ? e === -1 ? e = l - 1 : e -= 1 : c > 0 && (e += 1), s[e = (e + l) % l];
}
var ne = function(u, a) {
  var o = /* @__PURE__ */ new Set(), c = /* @__PURE__ */ new Map(), s = /* @__PURE__ */ new Map();
  return u.forEach(function(l) {
    var e = document.querySelector("[data-menu-id='".concat(le(a, l), "']"));
    e && (o.add(e), s.set(e, l), c.set(l, e));
  }), { elements: o, key2element: c, element2key: s };
};
function me(u, a, o, c, s, l, e, x, F, J) {
  var M = P.useRef(), K = P.useRef();
  K.current = a;
  var T = function() {
    O.cancel(M.current);
  };
  return P.useEffect(function() {
    return function() {
      T();
    };
  }, []), function(A) {
    var f = A.which;
    if ([].concat(C, [k, I, p, S]).includes(f)) {
      var Q = l(), D = ne(Q, c), L = D, R = L.elements, V = L.key2element, X = L.element2key, g = function(r, n) {
        for (var t = r || document.activeElement; t; ) {
          if (n.has(t)) return t;
          t = t.parentElement;
        }
        return null;
      }(V.get(a), R), N = X.get(g), v = function(r, n, t, h) {
        var G, E = "prev", b = "next", d = "children", y = "parent";
        if (r === "inline" && h === k) return { inlineTrigger: !0 };
        var j = i(i({}, z, E), w, b), ie = i(i(i(i({}, U, t ? b : E), W, t ? E : b), w, d), k, d), H = i(i(i(i(i(i({}, z, E), w, b), k, d), I, y), U, t ? d : y), W, t ? y : d);
        switch ((G = { inline: j, horizontal: ie, vertical: H, inlineSub: j, horizontalSub: H, verticalSub: H }["".concat(r).concat(n ? "" : "Sub")]) === null || G === void 0 ? void 0 : G[h]) {
          case E:
            return { offset: -1, sibling: !0 };
          case b:
            return { offset: 1, sibling: !0 };
          case y:
            return { offset: -1, sibling: !1 };
          case d:
            return { offset: 1, sibling: !1 };
          default:
            return null;
        }
      }(u, e(N, !0).length === 1, o, f);
      if (!v && f !== p && f !== S) return;
      (C.includes(f) || [p, S].includes(f)) && A.preventDefault();
      var q = function(r) {
        if (r) {
          var n = r, t = r.querySelector("a");
          t != null && t.getAttribute("href") && (n = t);
          var h = X.get(r);
          x(h), T(), M.current = O(function() {
            K.current === h && n.focus();
          });
        }
      };
      if ([p, S].includes(f) || v.sibling || !g) {
        var Y, Z, B = te(Y = g && u !== "inline" ? function(r) {
          for (var n = r; n; ) {
            if (n.getAttribute("data-menu-list")) return n;
            n = n.parentElement;
          }
          return null;
        }(g) : s.current, R);
        Z = f === p ? B[0] : f === S ? B[B.length - 1] : ee(Y, R, g, v.offset), q(Z);
      } else if (v.inlineTrigger) F(N);
      else if (v.offset > 0) F(N, !0), T(), M.current = O(function() {
        D = ne(Q, c);
        var r = g.getAttribute("aria-controls"), n = ee(document.getElementById(r), D.elements);
        q(n);
      }, 5);
      else if (v.offset < 0) {
        var _ = e(N, !0), $ = _[_.length - 2], re = V.get($);
        F($, !1), q(re);
      }
    }
    J == null || J(A);
  };
}
export {
  te as getFocusableElements,
  ne as refreshElements,
  me as useAccessibility
};
