// @gobolt/genesis/dist/components/Checkbox/styles.cjs
"use strict";Object.defineProperty(exports,Symbol.toStringTag,{value:"Module"});const a=require("styled-components"),b=require("../../node_modules/antd/es/checkbox/index.cjs"),c=({colors:n,borderRadius:e},o="primary",r="active")=>`
  .ant-checkbox-inner {
    background-color: #FFFFFF;
    border-color: #9A9A9A;
    border-radius: 2px;
    transition: all 0.2s ease-in-out;

    &::after {
      border-color: #9A9A9A;
      border: 2px solid ${n[o][r].color};
      border-top: 0;
      border-left: 0;
      content: "";
      display: table;
      height: 9px;
      left: 6px;
      opacity: 1;
      position: absolute;
      top: 2px;
      transform: rotate(45deg) scale(1);
      transition: all 0.2s cubic-bezier(0.12, 0.4, 0.29, 1.46) 0.1s;
      width: 5px;
    }
  }

  &:hover .ant-checkbox-inner::after {
    border-color: #6C6C6C;
  }

  &:hover .ant-checkbox-checked .ant-checkbox-inner::after {
    border-color: #6C6C6C;
  }

  .ant-checkbox-disabled .ant-checkbox-inner::after {
    border-color: #9A9A9A;
    background-color: #F4F4F4;
  }

  .ant-checkbox-disabled.ant-checkbox-checked .ant-checkbox-inner::after {
    border-color: #6C6C6C;
  }

  &:hover .ant-checkbox-disabled .ant-checkbox-inner::after {
    border-color: #6C6C6C;
  }

  .ant-checkbox-disabled.ant-checkbox-checked .ant-checkbox-inner {
    background-color: #F4F4F4;
    border-color: #9A9A9A;
  }

  .ant-checkbox-disabled.ant-checkbox-checked .ant-checkbox-inner::after {
    border-color: #9A9A9A;
  }

  &:hover .ant-checkbox-disabled.ant-checkbox-checked .ant-checkbox-inner {
    background-color: #F4F4F4;
    border-color: #9A9A9A;
  }

  &:hover .ant-checkbox-disabled.ant-checkbox-checked .ant-checkbox-inner::after {
    border-color: #9A9A9A;
  }

  .ant-checkbox-checked .ant-checkbox-inner {
    background-color: #fff;
    border-color: #9A9A9A;
  }

  .ant-checkbox-checked .ant-checkbox-inner::after {
    border-color: ${n[o].active.color};
  }

  &:hover .ant-checkbox-inner {
    border-color: #9A9A9A;

    
  }

  &:hover .ant-checkbox-checked .ant-checkbox-inner {
    background-color: #9A9A9A;
    border-color: #FFF;
    
  }

  &:hover .ant-checkbox-checked .ant-checkbox-inner::after {
    border-color: #fff;
  }

  .ant-checkbox-checked .ant-checkbox-inner {
    background-color:${n[o].focussed.ringColor};
    border-color: #6C6C6C;
  }

  .ant-checkbox-checked .ant-checkbox-inner {
    background-color: ${n.inputs.surface.checked};
    border-color: ${n.inputs.surface.checked};
  }

  &:hover .ant-checkbox-checked .ant-checkbox-inner {
    background-color: ${n.inputs.surface.checked};
    border-color: ${n.inputs.surface.checked};
  }

  &:hover .ant-checkbox-checked .ant-checkbox-inner::after {
    border-color: #fff;
  }

  &:hover .ant-checkbox-inner {
    border-color: #9A9A9A;
  }

  &:hover .ant-checkbox-checked .ant-checkbox-inner {
    background-color: ${n[o].focussed.ringColor};
    border-color: #6C6C6C;
  }

  .ant-checkbox:focus-visible .ant-checkbox-inner,
  .ant-checkbox-input:focus-visible + .ant-checkbox-inner {
    outline: none;
    box-shadow: 0 0 0 1px #fff, 0 0 0 3px ${n[o].focussed.ringColor};
    border-radius: ${e.BorderRadiusXs}px;
    border-color: #9A9A9A;
    transition: box-shadow 0.2s ease-in-out;
  }

  .ant-checkbox-input:focus + .ant-checkbox-inner {
    border-color: ${n[o].hover.borderColor};
  }

  &:focus-visible {
    outline: none;
    color: ${n[o].focussed.color};
    background-color: ${n[o].focussed.backgroundColor};
    box-shadow: 0 0 0 1px #fff, 0 0 0 3px ${n[o].focussed.ringColor};
    border-radius: ${e.BorderRadiusXs}px;
    border-color: #9A9A9A;
    transition: box-shadow 0.2s ease-in-out;
  }
`,t=a(b)`
  ${({theme:n,type:e,state:o})=>c(n,e,o)}
`;exports.StyledCheckbox=t,exports.getGenesisClass=c;
