// @ant-design/cssinjs/es/util/cacheMapUtil.js
import u from "../../../../@babel/runtime/helpers/esm/slicedToArray.js";
import l from "../../../../rc-util/es/Dom/canUseDom.js";
import { ATTR_MARK as f } from "../StyleContext.js";
var n, r = "data-ant-cssinjs-cache-path", v = "_FILE_STYLE__", c = !0;
function y() {
  if (!n && (n = {}, l())) {
    var e = document.createElement("div");
    e.className = r, e.style.position = "fixed", e.style.visibility = "hidden", e.style.top = "-9999px", document.body.appendChild(e);
    var i = getComputedStyle(e).content || "";
    (i = i.replace(/^"/, "").replace(/"$/, "")).split(";").forEach(function(d) {
      var s = d.split(":"), a = u(s, 2), p = a[0], m = a[1];
      n[p] = m;
    });
    var t, o = document.querySelector("style[".concat(r, "]"));
    o && (c = !1, (t = o.parentNode) === null || t === void 0 || t.removeChild(o)), document.body.removeChild(e);
  }
}
function C(e) {
  return y(), !!n[e];
}
function E(e) {
  var i = n[e], t = null;
  if (i && l()) if (c) t = v;
  else {
    var o = document.querySelector("style[".concat(f, '="').concat(n[e], '"]'));
    o ? t = o.innerHTML : delete n[e];
  }
  return [t, i];
}
export {
  r as ATTR_CACHE_MAP,
  v as CSS_FILE_STYLE,
  C as existPath,
  E as getStyleAndHash,
  y as prepare
};
