// @ant-design/cssinjs/es/util/index.js
import m from "../../../../@babel/runtime/helpers/esm/defineProperty.js";
import u from "../../../../@babel/runtime/helpers/esm/objectSpread2.js";
import l from "../../../../@babel/runtime/helpers/esm/typeof.js";
import s from "../../../../@emotion/hash/dist/hash.browser.esm.js";
import v from "../../../../rc-util/es/Dom/canUseDom.js";
import { ATTR_TOKEN as k, ATTR_MARK as y } from "../StyleContext.js";
import "../theme/createTheme.js";
import h from "../theme/Theme.js";
import "../theme/ThemeCache.js";
var T = /* @__PURE__ */ new WeakMap(), i = {};
function W(n, t) {
  for (var e = T, r = 0; r < t.length; r += 1) {
    var o = t[r];
    e.has(o) || e.set(o, /* @__PURE__ */ new WeakMap()), e = e.get(o);
  }
  return e.has(i) || e.set(i, n()), e.get(i);
}
var f = /* @__PURE__ */ new WeakMap();
function p(n) {
  var t = f.get(n) || "";
  return t || (Object.keys(n).forEach(function(e) {
    var r = n[e];
    t += e, r instanceof h ? t += r.id : r && l(r) === "object" ? t += p(r) : t += r;
  }), t = s(t), f.set(n, t)), t;
}
function _(n, t) {
  return s("".concat(t, "_").concat(p(n)));
}
var x = v();
function E(n) {
  return typeof n == "number" ? "".concat(n, "px") : n;
}
function K(n, t, e) {
  var r;
  if (arguments.length > 4 && arguments[4] !== void 0 && arguments[4]) return n;
  var o = u(u({}, arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : {}), {}, (m(r = {}, k, t), m(r, y, e), r)), g = Object.keys(o).map(function(a) {
    var c = o[a];
    return c ? "".concat(a, '="').concat(c, '"') : null;
  }).filter(function(a) {
    return a;
  }).join(" ");
  return "<style ".concat(g, ">").concat(n, "</style>");
}
export {
  p as flattenToken,
  x as isClientSide,
  W as memoResult,
  K as toStyleStr,
  _ as token2key,
  E as unit
};
