// rc-field-form/es/utils/validateUtil.js
import P from "../../../@babel/runtime/helpers/esm/toConsumableArray.js";
import E from "../../../@babel/runtime/helpers/esm/defineProperty.js";
import h from "../../../@babel/runtime/helpers/esm/regeneratorRuntime.js";
import x from "../../../@babel/runtime/helpers/esm/objectSpread2.js";
import _ from "../../../@babel/runtime/helpers/esm/asyncToGenerator.js";
import D from "../../../@rc-component/async-validator/es/index.js";
import * as I from "react";
import { warningOnce as j } from "../../../rc-util/es/warning.js";
import { defaultValidateMessages as G } from "./messages.js";
import { merge as L } from "../../../rc-util/es/utils/set.js";
var M = D;
function W(p, o) {
  return p.replace(/\\?\$\{\w+\}/g, function(u) {
    if (u.startsWith("\\")) return u.slice(1);
    var i = u.slice(2, -1);
    return o[i];
  });
}
var A = "CODE_LOGIC_ERROR";
function O(p, o, u, i, s) {
  return F.apply(this, arguments);
}
function F() {
  return F = _(h().mark(function p(o, u, i, s, g) {
    var n, b, d, w, t, c, l, m, f;
    return h().wrap(function(r) {
      for (; ; ) switch (r.prev = r.next) {
        case 0:
          return delete (n = x({}, i)).ruleIndex, M.warning = function() {
          }, n.validator && (b = n.validator, n.validator = function() {
            try {
              return b.apply(void 0, arguments);
            } catch (a) {
              return console.error(a), Promise.reject(A);
            }
          }), d = null, n && n.type === "array" && n.defaultField && (d = n.defaultField, delete n.defaultField), w = new M(E({}, o, [n])), t = L(G, s.validateMessages), w.messages(t), c = [], r.prev = 10, r.next = 13, Promise.resolve(w.validate(E({}, o, u), x({}, s)));
        case 13:
          r.next = 18;
          break;
        case 15:
          r.prev = 15, r.t0 = r.catch(10), r.t0.errors && (c = r.t0.errors.map(function(a, e) {
            var y = a.message, v = y === A ? t.default : y;
            return I.isValidElement(v) ? I.cloneElement(v, { key: "error_".concat(e) }) : v;
          }));
        case 18:
          if (c.length || !d) {
            r.next = 23;
            break;
          }
          return r.next = 21, Promise.all(u.map(function(a, e) {
            return O("".concat(o, ".").concat(e), a, d, s, g);
          }));
        case 21:
          return l = r.sent, r.abrupt("return", l.reduce(function(a, e) {
            return [].concat(P(a), P(e));
          }, []));
        case 23:
          return m = x(x({}, i), {}, { name: o, enum: (i.enum || []).join(", ") }, g), f = c.map(function(a) {
            return typeof a == "string" ? W(a, m) : a;
          }), r.abrupt("return", f);
        case 26:
        case "end":
          return r.stop();
      }
    }, p, null, [[10, 15]]);
  })), F.apply(this, arguments);
}
function Q(p, o, u, i, s, g) {
  var n, b = p.join("."), d = u.map(function(t, c) {
    var l = t.validator, m = x(x({}, t), {}, { ruleIndex: c });
    return l && (m.validator = function(f, r, a) {
      var e = !1, y = l(f, r, function() {
        for (var v = arguments.length, R = new Array(v), k = 0; k < v; k++) R[k] = arguments[k];
        Promise.resolve().then(function() {
          j(!e, "Your validator function has already return a promise. `callback` will be ignored."), e || a.apply(void 0, R);
        });
      });
      e = y && typeof y.then == "function" && typeof y.catch == "function", j(e, "`callback` is deprecated. Please return a promise instead."), e && y.then(function() {
        a();
      }).catch(function(v) {
        a(v || " ");
      });
    }), m;
  }).sort(function(t, c) {
    var l = t.warningOnly, m = t.ruleIndex, f = c.warningOnly, r = c.ruleIndex;
    return !!l == !!f ? m - r : l ? 1 : -1;
  });
  if (s === !0) n = new Promise(function() {
    var t = _(h().mark(function c(l, m) {
      var f, r, a;
      return h().wrap(function(e) {
        for (; ; ) switch (e.prev = e.next) {
          case 0:
            f = 0;
          case 1:
            if (!(f < d.length)) {
              e.next = 12;
              break;
            }
            return r = d[f], e.next = 5, O(b, o, r, i, g);
          case 5:
            if (!(a = e.sent).length) {
              e.next = 9;
              break;
            }
            return m([{ errors: a, rule: r }]), e.abrupt("return");
          case 9:
            f += 1, e.next = 1;
            break;
          case 12:
            l([]);
          case 13:
          case "end":
            return e.stop();
        }
      }, c);
    }));
    return function(c, l) {
      return t.apply(this, arguments);
    };
  }());
  else {
    var w = d.map(function(t) {
      return O(b, o, t, i, g).then(function(c) {
        return { errors: c, rule: t };
      });
    });
    n = (s ? function(t) {
      return V.apply(this, arguments);
    }(w) : function(t) {
      return C.apply(this, arguments);
    }(w)).then(function(t) {
      return Promise.reject(t);
    });
  }
  return n.catch(function(t) {
    return t;
  }), n;
}
function C() {
  return (C = _(h().mark(function p(o) {
    return h().wrap(function(u) {
      for (; ; ) switch (u.prev = u.next) {
        case 0:
          return u.abrupt("return", Promise.all(o).then(function(i) {
            var s;
            return (s = []).concat.apply(s, P(i));
          }));
        case 1:
        case "end":
          return u.stop();
      }
    }, p);
  }))).apply(this, arguments);
}
function V() {
  return (V = _(h().mark(function p(o) {
    var u;
    return h().wrap(function(i) {
      for (; ; ) switch (i.prev = i.next) {
        case 0:
          return u = 0, i.abrupt("return", new Promise(function(s) {
            o.forEach(function(g) {
              g.then(function(n) {
                n.errors.length && s([n]), (u += 1) === o.length && s([]);
              });
            });
          }));
        case 2:
        case "end":
          return i.stop();
      }
    }, p);
  }))).apply(this, arguments);
}
export {
  Q as validateRules
};
