// stylis/src/Parser.js
import { DECLARATION as z, COMMENT as K, RULESET as J } from "./Enum.js";
import { strlen as T, append as j, charat as V, replace as P, substr as O, indexof as D, abs as C, sizeof as L, trim as U, from as G } from "./Utility.js";
import { dealloc as y, alloc as H, next as W, peek as S, delimit as X, identifier as q, commenter as M, token as _, escaping as A, whitespace as F, node as Z, char as I, prev as N, caret as Y } from "./Tokenizer.js";
function ra(e) {
  return y(Q("", null, null, null, [""], e = H(e), 0, [0], e));
}
function Q(e, u, n, c, i, v, R, f, b) {
  for (var d = 0, o = 0, r = R, g = 0, k = 0, m = 0, t = 1, E = 1, p = 1, s = 0, h = "", x = i, w = v, l = c, a = h; E; ) switch (m = s, s = W()) {
    case 40:
      if (m != 108 && V(a, r - 1) == 58) {
        D(a += P(X(s), "&", "&\f"), "&\f", C(d ? f[d - 1] : 0)) != -1 && (p = -1);
        break;
      }
    case 34:
    case 39:
    case 91:
      a += X(s);
      break;
    case 9:
    case 10:
    case 13:
    case 32:
      a += F(m);
      break;
    case 92:
      a += A(Y() - 1, 7);
      continue;
    case 47:
      switch (S()) {
        case 42:
        case 47:
          j(aa(M(W(), Y()), u, n, b), b), _(m || 1) != 5 && _(S() || 1) != 5 || !T(a) || O(a, -1, void 0) === " " || (a += " ");
          break;
        default:
          a += "/";
      }
      break;
    case 123 * t:
      f[d++] = T(a) * p;
    case 125 * t:
    case 59:
    case 0:
      switch (s) {
        case 0:
        case 125:
          E = 0;
        case 59 + o:
          p == -1 && (a = P(a, /\f/g, "")), k > 0 && (T(a) - r || t === 0 && m === 47) && j(k > 32 ? B(a + ";", c, n, r - 1, b) : B(P(a, " ", "") + ";", c, n, r - 2, b), b);
          break;
        case 59:
          a += ";";
        default:
          if (j(l = $(a, u, n, d, o, i, f, h, x = [], w = [], r, v), v), s === 123) if (o === 0) Q(a, u, l, l, x, v, r, f, w);
          else {
            switch (g) {
              case 99:
                if (V(a, 3) === 110) break;
              case 108:
                if (V(a, 2) === 97) break;
              default:
                o = 0;
              case 100:
              case 109:
              case 115:
            }
            o ? Q(e, l, l, c && j($(e, l, l, 0, 0, i, f, h, i, x = [], r, w), w), i, w, r, f, c ? x : w) : Q(a, l, l, l, [""], w, 0, f, w);
          }
      }
      d = o = k = 0, t = p = 1, h = a = "", r = R;
      break;
    case 58:
      r = 1 + T(a), k = m;
    default:
      if (t < 1) {
        if (s == 123) --t;
        else if (s == 125 && t++ == 0 && N() == 125) continue;
      }
      switch (a += G(s), s * t) {
        case 38:
          p = o > 0 ? 1 : (a += "\f", -1);
          break;
        case 44:
          f[d++] = (T(a) - 1) * p, p = 1;
          break;
        case 64:
          S() === 45 && (a += X(W())), g = S(), o = r = T(h = a += q(Y())), s++;
          break;
        case 45:
          m === 45 && T(a) == 2 && (t = 0);
      }
  }
  return v;
}
function $(e, u, n, c, i, v, R, f, b, d, o, r) {
  for (var g = i - 1, k = i === 0 ? v : [""], m = L(k), t = 0, E = 0, p = 0; t < c; ++t) for (var s = 0, h = O(e, g + 1, g = C(E = R[t])), x = e; s < m; ++s) (x = U(E > 0 ? k[s] + " " + h : P(h, /&\f/g, k[s]))) && (b[p++] = x);
  return Z(e, u, n, i === 0 ? J : f, b, d, o, r);
}
function aa(e, u, n, c) {
  return Z(e, u, n, K, G(I()), O(e, 2, -2), 0, c);
}
function B(e, u, n, c, i) {
  return Z(e, u, n, z, O(e, 0, c), O(e, c + 1, -1), c, i);
}
export {
  aa as comment,
  ra as compile,
  B as declaration,
  Q as parse,
  $ as ruleset
};
