// rc-field-form/es/utils/NameMap.js
import s from "../../../@babel/runtime/helpers/esm/slicedToArray.js";
import k from "../../../@babel/runtime/helpers/esm/toConsumableArray.js";
import h from "../../../@babel/runtime/helpers/esm/classCallCheck.js";
import y from "../../../@babel/runtime/helpers/esm/createClass.js";
import d from "../../../@babel/runtime/helpers/esm/defineProperty.js";
import _ from "../../../@babel/runtime/helpers/esm/typeof.js";
var m = "__@field_split__";
function u(r) {
  return r.map(function(t) {
    return "".concat(_(t), ":").concat(t);
  }).join(m);
}
var x = function() {
  function r() {
    h(this, r), d(this, "kvs", /* @__PURE__ */ new Map());
  }
  return y(r, [{ key: "set", value: function(t, e) {
    this.kvs.set(u(t), e);
  } }, { key: "get", value: function(t) {
    return this.kvs.get(u(t));
  } }, { key: "update", value: function(t, e) {
    var n = e(this.get(t));
    n ? this.set(t, n) : this.delete(t);
  } }, { key: "delete", value: function(t) {
    this.kvs.delete(u(t));
  } }, { key: "map", value: function(t) {
    return k(this.kvs.entries()).map(function(e) {
      var n = s(e, 2), i = n[0], f = n[1], v = i.split(m);
      return t({ key: v.map(function(c) {
        var l = c.match(/^([^:]*):(.*)$/), a = s(l, 3), p = a[1], o = a[2];
        return p === "number" ? Number(o) : o;
      }), value: f });
    });
  } }, { key: "toJSON", value: function() {
    var t = {};
    return this.map(function(e) {
      var n = e.key, i = e.value;
      return t[n.join(".")] = i, null;
    }), t;
  } }]), r;
}();
export {
  x as default
};
