// @gobolt/genesis/dist/components/Notification/styles.cjs
"use strict";Object.defineProperty(exports,Symbol.toStringTag,{value:"Module"});const i=require("styled-components"),o=require("../../constants/index.cjs"),n=i.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  gap: ${({theme:e})=>e.sizing.Size4}px;
  padding: ${({theme:e})=>e.sizing.Size3}px;
  pointer-events: none;
  cursor: default;
  border-radius: ${({theme:e})=>e.borderRadius.BorderRadiusSm}px;
  height: 85px;
  width: 500px;
  background-color: ${({theme:e,$isUnread:t})=>t?"#F4F4F4":e.colors.surface.default};
`,r=i.div`
  display: flex;
  align-items: center;
  gap: ${({theme:e})=>e.sizing.Size2}px;
  flex: 1;
  max-width: 500px;

  .ant-typography {
    display: -webkit-box;
    -webkit-line-clamp: 3;
    -webkit-box-orient: vertical;
    overflow: hidden;
    text-overflow: ellipsis;
    font-family: ${({theme:e})=>e.typography.fontFamily};
    line-height: ${({theme:e,size:t})=>t===o.SIZE.standard?e.typography.lineHeight:"16px"};
  }
`,d=i.div`
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: ${({theme:e})=>e.colors.status.info.default};
  border-radius: 50%;
  height: 8px;
  width: 8px;
  aspect-ratio: 1;
  min-height: 8px;
`;i.button`
  border: ${({theme:e})=>(t=>`1px solid ${t.colors.interactive.utility.border}`)(e)};
  background: "none";
  cursor: "pointer";
  pointer-events: "all";
  color: ${({theme:e})=>e.colors.interactive.utility.onsurface};
  border-radius: ${({theme:e})=>e.borderRadius.BorderRadiusSm};
  padding-top: "2px";
  padding-bottom: "2px";
  padding-left: ${({theme:e})=>e.components.utilityButton.padding.horPadding};
  padding-right: ${({theme:e})=>e.components.utilityButton.padding.horPadding};
`;const a=i.button`
  text-decoration: underline;
  border: none;
  background: none;
  cursor: pointer;
  pointer-events: all;
`;exports.ChevronButton=a,exports.NotificationContent=r,exports.NotificationRow=n,exports.NotificationUnread=d;
