// rc-virtual-list/es/hooks/useScrollDrag.js
import w from "../../../rc-util/es/raf.js";
import * as h from "react";
function E(e) {
  return Math.floor(Math.pow(e, 0.5));
}
function p(e, r) {
  return ("touches" in e ? e.touches[0] : e)[r ? "pageX" : "pageY"] - window[r ? "scrollX" : "scrollY"];
}
function X(e, r, L) {
  h.useEffect(function() {
    var n = r.current;
    if (e && n) {
      var f, i, a = !1, u = function() {
        w.cancel(f);
      }, c = function o() {
        u(), f = w(function() {
          L(i), o();
        });
      }, m = function(o) {
        if (!o.target.draggable) {
          var t = o;
          t._virtualHandled || (t._virtualHandled = !0, a = !0);
        }
      }, s = function() {
        a = !1, u();
      }, d = function(o) {
        if (a) {
          var t = p(o, !1), v = n.getBoundingClientRect(), l = v.top, g = v.bottom;
          t <= l ? (i = -E(l - t), c()) : t >= g ? (i = E(t - g), c()) : u();
        }
      };
      return n.addEventListener("mousedown", m), n.ownerDocument.addEventListener("mouseup", s), n.ownerDocument.addEventListener("mousemove", d), function() {
        n.removeEventListener("mousedown", m), n.ownerDocument.removeEventListener("mouseup", s), n.ownerDocument.removeEventListener("mousemove", d), u();
      };
    }
  }, [e]);
}
export {
  X as default,
  p as getPageXY
};
