// @secondcloset/types/src/fulfillment/order/order.ts
import { Shipment, ScServiceCode } from "../shipment/shipment";
import { Customer } from "./customer";
import {
  FacilityCode,
  BaseAddress,
  BaseUser,
  Organization,
  File,
  Version,
} from "../../_common";
import { OrderItem } from "./orderItem";
import { Channel } from "../inventory";

export type ShipmentTag =
  | "ready_to_fulfill"
  | "unfulfilled_items"
  | "processing_items"
  | "items_fulfilled"
  | "processing_return_items"
  | "items_returned"
  | "returned_to_sender"
  | "pending_return_to_sender"
  | "items_removed"
  | "pending_inventory_receiving"
  | "items_partially_fulfilled"
  | "customer_booking"
  | "pending_customer_pickup"
  | "items_picked_up_by_customer";

export type OrderPlatform =
  | "manual"
  | "csv"
  | "shopify"
  | "centiro"
  | "woo_commerce"
  | "shipstaion"
  | "external_api";

export interface FulfillmentOrderAddress extends BaseAddress {
  customer_id: string;
  contact_name: string;
  customer_phone_number: string | null;
  evening_phone_number: string | null;
  cell_phone_number: string | null;
  created_at: string;
  updated_at: string;
}

export type Order<V extends Version = "V1"> = V extends "V1"
  ? {
      id: string;
      order_id: string;
      number: string;
      external_order_number: string;
      external_order_id: string | null;
      notes: string;
      internal_notes?: string;
      platform: OrderPlatform;
      external_platform_version: number | null;
      parcel_type: string;
      cancelled: boolean;
      fulfilled_from: FacilityCode;
      client_type: string | null;
      insurance_value: number;
      shipping_service: string | null;
      signature_required?: boolean;
      customer: Customer;
      address: FulfillmentOrderAddress;
      created_at: string;
      updated_at: string;
      organization: Organization;
      overdue: boolean;
      user: BaseUser;
      status?: string;
      shipment_tags: ShipmentTag[];
      shipments: Shipment<V>[];
      items: OrderItem<V>[];
      files: File[];
      stock_status:
        | "in_stock"
        | "partially_in_stock"
        | "out_of_stock"
        | "unknown_stock";
      on_hold: boolean;
      on_hold_at: string;
      requested_service_code?: ScServiceCode;
      has_errors: boolean;
      sales_channel?: Channel["name"];
    }
  : Omit<Order<"V1">, "items" | "shipments">;
