// @secondcloset/types/src/fulfillment/shipment/package.ts
import { PackageIssue } from "./packageIssue";

export type PackageStatus =
  | "unattempted"
  | "failed"
  | "completed"
  | "cancelled"
  | "not_delivered"
  | "redelivered"
  | "label_generated"
  | "awaiting_carrier"
  | "with_carrier";

export type PackageType = "box" | "standard_pallet" | "double_pallet";

export interface Container {
  id: string;
  name: string;
  container_type?: string;
}

export interface Package {
  id: string;
  package_id: null | string;
  logistics_shipment_id: string;
  weight_value: string;
  weight_unit: string;
  length_value: string;
  length_unit: string;
  height_value: string;
  height_unit: string;
  width_value: string;
  width_unit: string;
  logistics_external_carrier_shipping_label_id: string;
  logistics_bolt_shipping_label_id?: string;
  shipment_item_ids: string[];
  status: PackageStatus;
  tracking_number: null | string;
  failed_reason: null | string;
  failed_notes: null | string;
  package_type?: PackageType;
  package_issues?: PackageIssue[];
  container?: Container;
  zone_code: null | string;
}
