// @gobolt/genesis/dist/components/Badge/styles.cjs
"use strict";Object.defineProperty(exports,Symbol.toStringTag,{value:"Module"});const d=require("styled-components"),t=(n,e,a,o)=>`
  display: flex;
  gap: 4px;
  justify-content: center;
  align-items: center;
  position: relative;
  padding-top: ${n.components.badge["ver-padding"]}px;
  padding-bottom: ${n.components.badge["ver-padding"]}px;
  padding-left: ${n.components.badge["left-padding"]}px;
  padding-right: ${n.components.badge["right-padding"]}px;
  border-radius: ${n.components.badge.radius}px;
  border: 1px solid ${((r,s)=>s==="hollow"?r.colors.inputs.surface.border:r.colors.status[s].surface)(n,e)};
  background-color: ${n.colors.status[e].surface};

  &:not(:disabled):hover {
    background-color: ${n.colors.status[e]["surface-hover"]};
  }

  ${o?`
    background-color: ${n.colors.status.disabled.surface} !important;
    border-color: ${n.colors.status.disabled} !important;
    border: none;
    cursor: not-allowed;

    &:hover {
      cursor: not-allowed;
    }
  `:""}
`,i=d.div`
  ${({theme:n,state:e,label:a,disabled:o})=>t(n,e,0,o)}
`;exports.Badge=i,exports.getGenesisClass=t;
