// @ant-design/cssinjs/es/theme/ThemeCache.js
import v from "../../../../@babel/runtime/helpers/esm/slicedToArray.js";
import f from "../../../../@babel/runtime/helpers/esm/classCallCheck.js";
import m from "../../../../@babel/runtime/helpers/esm/createClass.js";
import o from "../../../../@babel/runtime/helpers/esm/defineProperty.js";
function d(l, t) {
  if (l.length !== t.length) return !1;
  for (var e = 0; e < l.length; e++) if (l[e] !== t[e]) return !1;
  return !0;
}
var c = function() {
  function l() {
    f(this, l), o(this, "cache", void 0), o(this, "keys", void 0), o(this, "cacheCallTimes", void 0), this.cache = /* @__PURE__ */ new Map(), this.keys = [], this.cacheCallTimes = 0;
  }
  return m(l, [{ key: "size", value: function() {
    return this.keys.length;
  } }, { key: "internalGet", value: function(t) {
    var e, a, n = arguments.length > 1 && arguments[1] !== void 0 && arguments[1], r = { map: this.cache };
    return t.forEach(function(s) {
      var i;
      r ? r = (i = r) === null || i === void 0 || (i = i.map) === null || i === void 0 ? void 0 : i.get(s) : r = void 0;
    }), (e = r) !== null && e !== void 0 && e.value && n && (r.value[1] = this.cacheCallTimes++), (a = r) === null || a === void 0 ? void 0 : a.value;
  } }, { key: "get", value: function(t) {
    var e;
    return (e = this.internalGet(t, !0)) === null || e === void 0 ? void 0 : e[0];
  } }, { key: "has", value: function(t) {
    return !!this.internalGet(t);
  } }, { key: "set", value: function(t, e) {
    var a = this;
    if (!this.has(t)) {
      if (this.size() + 1 > l.MAX_CACHE_SIZE + l.MAX_CACHE_OFFSET) {
        var n = this.keys.reduce(function(i, h) {
          var u = v(i, 2)[1];
          return a.internalGet(h)[1] < u ? [h, a.internalGet(h)[1]] : i;
        }, [this.keys[0], this.cacheCallTimes]), r = v(n, 1)[0];
        this.delete(r);
      }
      this.keys.push(t);
    }
    var s = this.cache;
    t.forEach(function(i, h) {
      if (h === t.length - 1) s.set(i, { value: [e, a.cacheCallTimes++] });
      else {
        var u = s.get(i);
        u ? u.map || (u.map = /* @__PURE__ */ new Map()) : s.set(i, { map: /* @__PURE__ */ new Map() }), s = s.get(i).map;
      }
    });
  } }, { key: "deleteByPath", value: function(t, e) {
    var a, n = t.get(e[0]);
    if (e.length === 1) return n.map ? t.set(e[0], { map: n.map }) : t.delete(e[0]), (a = n.value) === null || a === void 0 ? void 0 : a[0];
    var r = this.deleteByPath(n.map, e.slice(1));
    return n.map && n.map.size !== 0 || n.value || t.delete(e[0]), r;
  } }, { key: "delete", value: function(t) {
    if (this.has(t)) return this.keys = this.keys.filter(function(e) {
      return !d(e, t);
    }), this.deleteByPath(this.cache, t);
  } }]), l;
}();
o(c, "MAX_CACHE_SIZE", 20), o(c, "MAX_CACHE_OFFSET", 5);
export {
  c as default,
  d as sameDerivativeOption
};
