// @ant-design/cssinjs-utils/es/util/calc/NumCalculator.js
import s from "../../../../../@babel/runtime/helpers/esm/classCallCheck.js";
import i from "../../../../../@babel/runtime/helpers/esm/createClass.js";
import n from "../../../../../@babel/runtime/helpers/esm/assertThisInitialized.js";
import o from "../../../../../@babel/runtime/helpers/esm/inherits.js";
import l from "../../../../../@babel/runtime/helpers/esm/createSuper.js";
import f from "../../../../../@babel/runtime/helpers/esm/defineProperty.js";
import a from "./calculator.js";
var k = function() {
  o(r, a);
  var e = l(r);
  function r(t) {
    var u;
    return s(this, r), u = e.call(this), f(n(u), "result", 0), t instanceof r ? u.result = t.result : typeof t == "number" && (u.result = t), u;
  }
  return i(r, [{ key: "add", value: function(t) {
    return t instanceof r ? this.result += t.result : typeof t == "number" && (this.result += t), this;
  } }, { key: "sub", value: function(t) {
    return t instanceof r ? this.result -= t.result : typeof t == "number" && (this.result -= t), this;
  } }, { key: "mul", value: function(t) {
    return t instanceof r ? this.result *= t.result : typeof t == "number" && (this.result *= t), this;
  } }, { key: "div", value: function(t) {
    return t instanceof r ? this.result /= t.result : typeof t == "number" && (this.result /= t), this;
  } }, { key: "equal", value: function() {
    return this.result;
  } }]), r;
}();
export {
  k as default
};
