// @gobolt/genesis/dist/components/Notification/styles.js
import i from "styled-components";
import { SIZE as o } from "../../constants/index.js";
const a = i.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  gap: ${({ theme: e }) => e.sizing.Size4}px;
  padding: ${({ theme: e }) => e.sizing.Size3}px;
  pointer-events: none;
  cursor: default;
  border-radius: ${({ theme: e }) => e.borderRadius.BorderRadiusSm}px;
  height: 85px;
  width: 500px;
  background-color: ${({ theme: e, $isUnread: t }) => t ? "#F4F4F4" : e.colors.surface.default};
`, d = i.div`
  display: flex;
  align-items: center;
  gap: ${({ theme: e }) => e.sizing.Size2}px;
  flex: 1;
  max-width: 500px;

  .ant-typography {
    display: -webkit-box;
    -webkit-line-clamp: 3;
    -webkit-box-orient: vertical;
    overflow: hidden;
    text-overflow: ellipsis;
    font-family: ${({ theme: e }) => e.typography.fontFamily};
    line-height: ${({ theme: e, size: t }) => t === o.standard ? e.typography.lineHeight : "16px"};
  }
`, s = i.div`
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: ${({ theme: e }) => e.colors.status.info.default};
  border-radius: 50%;
  height: 8px;
  width: 8px;
  aspect-ratio: 1;
  min-height: 8px;
`;
i.button`
  border: ${({ theme: e }) => ((t) => `1px solid ${t.colors.interactive.utility.border}`)(e)};
  background: "none";
  cursor: "pointer";
  pointer-events: "all";
  color: ${({ theme: e }) => e.colors.interactive.utility.onsurface};
  border-radius: ${({ theme: e }) => e.borderRadius.BorderRadiusSm};
  padding-top: "2px";
  padding-bottom: "2px";
  padding-left: ${({ theme: e }) => e.components.utilityButton.padding.horPadding};
  padding-right: ${({ theme: e }) => e.components.utilityButton.padding.horPadding};
`;
const p = i.button`
  text-decoration: underline;
  border: none;
  background: none;
  cursor: pointer;
  pointer-events: all;
`;
export {
  p as ChevronButton,
  d as NotificationContent,
  a as NotificationRow,
  s as NotificationUnread
};
