// @secondcloset/types/src/fulfillment/asn/asn.ts
import { FacilityCode } from "../../_common/facility";
import { Channel } from "../inventory";
import { ASNItem } from "./asnItem";

export type ASNStatus =
  | "draft"
  | "awaiting"
  | "arrived"
  | "on_hold"
  | "completed";

export type ASNShipmentType =
  | "parcel"
  | "pallet"
  | "container"
  | "floor_loaded"
  | "drop_off";

export type ASNPalletPackingOption =
  | "one_sku_per_pallet"
  | "multiple_sku_per_pallet";

export type ASNBoxPackingOption = "one_sku_per_box" | "multiple_sku_per_box";

export interface AsnBol {
  access: "external" | "internal";
  created_at: string;
  id: string;
  updated_at: string;
  url: string;
}

export interface TimelineEvent {
  user_name: string;
  user_is_admin: boolean;
  event: string;
  description: string;
  created_at: string;
}

export interface ASN {
  id: string;
  status: ASNStatus;
  number: string;
  facility?: FacilityCode;
  submitted_date?: string;
  expected_arrival_date?: string;
  actual_arrival_date?: string;
  reference_number?: string;
  po_number?: string;
  container_size?: string;
  box_packing_option?: ASNBoxPackingOption;
  pallet_packing_option?: ASNPalletPackingOption;
  shipment_type?: ASNShipmentType;
  internal_notes?: string;
  qty_under_received?: boolean;
  qty_over_received?: boolean;
  box_expected_quantity?: number;
  box_received_quantity?: number;
  pallet_expected_quantity?: number;
  pallet_received_quantity?: number;
  bols_attributes: AsnBol[];
  current_step?: string;
  created_at: string;
  updated_at: string;
  deleted_at?: string;
  items_attributes: ASNItem[];
  timeline?: TimelineEvent[];
  organization_id?: string;
  inventory_channel?: Channel["name"];
  platform?: string;
  external_id?: string;
  external_number?: string;
}
