// @secondcloset/types/src/fulfillment/order/orderItem.ts
import { Version } from "../../_common";
import { BaseProduct, Product } from "../product";
import { ShipmentActionType } from "../shipment";

export type OrderItem<V extends Version = "V1"> = V extends "V1"
  ? {
      id: string;
      product: Product;
      quantity: number;
      attributes: string[];
      return_reasons: string[];
      available_actions: ShipmentActionType[];
      actions_history: ShipmentActionType[];
      virtual_kit_id: null | string;
      available_quantity: number;
      removed: boolean;
      removed_reason: string;
    }
  : {
      product: BaseProduct & {
        available_quantity: OrderItem<"V1">["available_quantity"];
        base_products?: OrderItem<"V2">["product"][];
        required_parts?: Record<string, number>;
      };
      available_actions: {
        inventory_pick_up_item_ids: OrderItem<"V1">["id"][];
        inventory_warehouse_receiving_item_ids: OrderItem<"V1">["id"][];
        delivery_item_ids: OrderItem<"V1">["id"][];
        reverse_logistics_item_ids: OrderItem<"V1">["id"][];
        setup_item_ids: OrderItem<"V1">["id"][];
      };
      removed_items: {
        removed_reason: string;
        quantity: number;
      }[];
    };
